# Automatically generated by openapi-generator (https://openapi-generator.tech)
# Please update as you see appropriate

context("Test Variant")

model.instance <- Variant$new()

test_that("additionalInfo", {
  # tests for the property `additionalInfo` (map(character))
  # Additional arbitrary info

  # uncomment below to test the property 
  #expect_equal(model.instance$`additionalInfo`, "EXPECTED_RESULT")
})

test_that("alternate_bases", {
  # tests for the property `alternate_bases` (array[character])
  # The bases that appear instead of the reference bases. Multiple alternate alleles are possible.

  # uncomment below to test the property 
  #expect_equal(model.instance$`alternate_bases`, "EXPECTED_RESULT")
})

test_that("ciend", {
  # tests for the property `ciend` (array[integer])
  # Similar to \&quot;cipos\&quot;, but for the variant&#39;s end position (which is derived from start + svlen).

  # uncomment below to test the property 
  #expect_equal(model.instance$`ciend`, "EXPECTED_RESULT")
})

test_that("cipos", {
  # tests for the property `cipos` (array[integer])
  # In the case of structural variants, start and end of the variant may not be known with an exact base position. \&quot;cipos\&quot; provides an interval with high confidence for the start position. The interval is provided by 0 or 2 signed integers which are added to the start position. Based on the use in VCF v4.2

  # uncomment below to test the property 
  #expect_equal(model.instance$`cipos`, "EXPECTED_RESULT")
})

test_that("created", {
  # tests for the property `created` (character)
  # The timestamp when this variant was created.

  # uncomment below to test the property 
  #expect_equal(model.instance$`created`, "EXPECTED_RESULT")
})

test_that("end", {
  # tests for the property `end` (integer)
  # This field is optional and may be ignored if there is no relevant map or reference to be associated with.  The end position (exclusive), resulting in [start, end) closed-open interval. This is typically calculated  by &#x60;start + referenceBases.length&#x60;.

  # uncomment below to test the property 
  #expect_equal(model.instance$`end`, "EXPECTED_RESULT")
})

test_that("filtersApplied", {
  # tests for the property `filtersApplied` (character)
  # True if filters were applied for this variant. VCF column 7 \&quot;FILTER\&quot; any value other than the missing value.

  # uncomment below to test the property 
  #expect_equal(model.instance$`filtersApplied`, "EXPECTED_RESULT")
})

test_that("filtersFailed", {
  # tests for the property `filtersFailed` (array[character])
  # Zero or more filters that failed for this variant. VCF column 7 \&quot;FILTER\&quot; shared across all alleles in the same VCF record.

  # uncomment below to test the property 
  #expect_equal(model.instance$`filtersFailed`, "EXPECTED_RESULT")
})

test_that("filtersPassed", {
  # tests for the property `filtersPassed` (character)
  # True if all filters for this variant passed. VCF column 7 \&quot;FILTER\&quot; value PASS.

  # uncomment below to test the property 
  #expect_equal(model.instance$`filtersPassed`, "EXPECTED_RESULT")
})

test_that("referenceBases", {
  # tests for the property `referenceBases` (character)
  # The reference bases for this variant. They start at the given start position.

  # uncomment below to test the property 
  #expect_equal(model.instance$`referenceBases`, "EXPECTED_RESULT")
})

test_that("referenceName", {
  # tests for the property `referenceName` (character)
  # The reference on which this variant occurs. (e.g. &#x60;chr_20&#x60; or &#x60;X&#x60;)

  # uncomment below to test the property 
  #expect_equal(model.instance$`referenceName`, "EXPECTED_RESULT")
})

test_that("start", {
  # tests for the property `start` (integer)
  # This field is optional and may be ignored if there is no relevant map or reference to be associated with.  The start position at which this variant occurs (0-based). This corresponds to the first base of the string  of reference bases. Genomic positions are non-negative integers less than reference length. Variants spanning  the join of circular genomes are represented as two variants one on each side of the join (position 0).

  # uncomment below to test the property 
  #expect_equal(model.instance$`start`, "EXPECTED_RESULT")
})

test_that("svlen", {
  # tests for the property `svlen` (integer)
  # Length of the - if labeled as such in variant_type - structural variation. Based on the use in VCF v4.2

  # uncomment below to test the property 
  #expect_equal(model.instance$`svlen`, "EXPECTED_RESULT")
})

test_that("updated", {
  # tests for the property `updated` (character)
  # The time at which this variant was last updated.

  # uncomment below to test the property 
  #expect_equal(model.instance$`updated`, "EXPECTED_RESULT")
})

test_that("variantDbId", {
  # tests for the property `variantDbId` (character)
  # The variant ID.

  # uncomment below to test the property 
  #expect_equal(model.instance$`variantDbId`, "EXPECTED_RESULT")
})

test_that("variantNames", {
  # tests for the property `variantNames` (array[character])
  # Names for the variant, for example a RefSNP ID.

  # uncomment below to test the property 
  #expect_equal(model.instance$`variantNames`, "EXPECTED_RESULT")
})

test_that("variantSetDbId", {
  # tests for the property `variantSetDbId` (array[character])
  # An array of &#x60;VariantSet&#x60; IDs this variant belongs to. This also defines the &#x60;ReferenceSet&#x60; against which the &#x60;Variant&#x60; is to be interpreted.

  # uncomment below to test the property 
  #expect_equal(model.instance$`variantSetDbId`, "EXPECTED_RESULT")
})

test_that("variantType", {
  # tests for the property `variantType` (character)
  # The \&quot;variant_type\&quot; is used to denote e.g. structural variants. Examples:   DUP  : duplication of sequence following \&quot;start\&quot;   DEL  : deletion of sequence following \&quot;start\&quot;

  # uncomment below to test the property 
  #expect_equal(model.instance$`variantType`, "EXPECTED_RESULT")
})

