# BrAPI-Core
#
# The Breeding API (BrAPI) is a Standardized REST ful Web Service API Specification for communicating Plant Breeding Data. BrAPI allows for easy data sharing between databases and tools involved in plant breeding. <div class=\"brapi-section\"> <h2 class=\"brapi-section-title\">General Reference Documentation</h2> <div class=\"gen-info-link\"><a href=\"https://github.com/plantbreeding/API/blob/master/Specification/GeneralInfo/URL_Structure.md\">URL Structure</a></div> <div class=\"gen-info-link\"><a href=\"https://github.com/plantbreeding/API/blob/master/Specification/GeneralInfo/Response_Structure.md\">Response Structure</a></div> <div class=\"gen-info-link\"><a href=\"https://github.com/plantbreeding/API/blob/master/Specification/GeneralInfo/Date_Time_Encoding.md\">Date/Time Encoding</a></div> <div class=\"gen-info-link\"><a href=\"https://github.com/plantbreeding/API/blob/master/Specification/GeneralInfo/Location_Encoding.md\">Location Encoding</a></div> <div class=\"gen-info-link\"><a href=\"https://github.com/plantbreeding/API/blob/master/Specification/GeneralInfo/Error_Handling.md\">Error Handling</a></div> <div class=\"gen-info-link\"><a href=\"https://github.com/plantbreeding/API/blob/master/Specification/GeneralInfo/Search_Services.md\">Search Services</a></div> </div>  <div class=\"current-brapi-section brapi-section\"> <h2 class=\"brapi-section-title\">BrAPI Core</h2> <div class=\"brapi-section-description\">The BrAPI Core module contains high level entities used for organization and management. This includes Programs, Trials, Studies, Locations, People, and Lists</div> <div class=\"version-number\">V2.0</div> <div class=\"link-btn\"><a href=\"https://github.com/plantbreeding/API/tree/master/Specification/BrAPI-Core\">GitHub</a></div> <div class=\"link-btn\"><a href=\"https://app.swaggerhub.com/apis/PlantBreedingAPI/BrAPI-Core\">SwaggerHub</a></div> <div class=\"link-btn\"><a href=\"https://brapicore.docs.apiary.io\">Apiary</a></div> <div class=\"stop-float\"></div> </div>  <div class=\"brapi-section\"> <h2 class=\"brapi-section-title\">BrAPI Phenotyping</h2> <div class=\"brapi-section-description\">The BrAPI Phenotyping module contains entities related to phenotypic observations. This includes Observation Units, Observations, Observation Variables, Traits, Scales, Methods, and Images</div> <div class=\"version-number\">V2.0</div> <div class=\"link-btn\"><a href=\"https://github.com/plantbreeding/API/tree/master/Specification/BrAPI-Phenotyping\">GitHub</a></div> <div class=\"link-btn\"><a href=\"https://app.swaggerhub.com/apis/PlantBreedingAPI/BrAPI-Phenotyping\">SwaggerHub</a></div> <div class=\"link-btn\"><a href=\"https://brapiphenotyping.docs.apiary.io\">Apiary</a></div> <div class=\"stop-float\"></div> </div>  <div class=\"brapi-section\"> <h2 class=\"brapi-section-title\">BrAPI Genotyping</h2> <div class=\"brapi-section-description\">The BrAPI Genotyping module contains entities related to genotyping analysis. This includes Samples, Markers, Variant Sets, Variants, Call Sets, Calls, References, Reads, and Vendor Orders</div> <div class=\"version-number\">V2.0</div> <div class=\"link-btn\"><a href=\"https://github.com/plantbreeding/API/tree/master/Specification/BrAPI-Genotyping\">GitHub</a></div> <div class=\"link-btn\"><a href=\"https://app.swaggerhub.com/apis/PlantBreedingAPI/BrAPI-Genotyping\">SwaggerHub</a></div> <div class=\"link-btn\"><a href=\"https://brapigenotyping.docs.apiary.io\">Apiary</a></div> <div class=\"stop-float\"></div> </div>  <div class=\"brapi-section\"> <h2 class=\"brapi-section-title\">BrAPI Germplasm</h2> <div class=\"brapi-section-description\">The BrAPI Germplasm module contains entities related to germplasm management. This includes Germplasm, Germplasm Attributes, Seed Lots, Crosses, Pedigree, and Progeny</div> <div class=\"version-number\">V2.0</div> <div class=\"link-btn\"><a href=\"https://github.com/plantbreeding/API/tree/master/Specification/BrAPI-Germplasm\">GitHub</a></div> <div class=\"link-btn\"><a href=\"https://app.swaggerhub.com/apis/PlantBreedingAPI/BrAPI-Germplasm\">SwaggerHub</a></div> <div class=\"link-btn\"><a href=\"https://brapigermplasm.docs.apiary.io\">Apiary</a></div> <div class=\"stop-float\"></div> </div>  <style> .link-btn{ float: left;  margin: 2px 10px 0 0;  padding: 0 5px;  border-radius: 5px;  background-color: #ddd; } .stop-float{   clear: both; } .version-number{   float: left;    margin: 5px 10px 0 5px; } .brapi-section-title{   margin: 0 10px 0 0;   font-size: 20px; } .current-brapi-section{   font-weight: bolder;   border-radius: 5px;    background-color: #ddd; } .brapi-section{   padding: 5px 5px;  } .brapi-section-description{   margin: 5px 0 0 5px; } </style>
#
# The version of the OpenAPI document: 2.0
# 
# Generated by: https://openapi-generator.tech

#' @docType class
#' @title Call
#'
#' @description Call Class
#'
#' @format An \code{R6Class} generator object
#'
#' @field additionalInfo  named list( character ) [optional]
#'
#' @field callSetDbId  character [optional]
#'
#' @field callSetName  character [optional]
#'
#' @field genotype  \link{ListValue} [optional]
#'
#' @field genotype_likelihood  list( numeric ) [optional]
#'
#' @field phaseSet  character [optional]
#'
#' @field variantDbId  character [optional]
#'
#' @field variantName  character [optional]
#'
#' @importFrom R6 R6Class
#' @importFrom jsonlite fromJSON toJSON
#' @export
Call <- R6::R6Class(
  'Call',
  public = list(
    `additionalInfo` = NULL,
    `callSetDbId` = NULL,
    `callSetName` = NULL,
    `genotype` = NULL,
    `genotype_likelihood` = NULL,
    `phaseSet` = NULL,
    `variantDbId` = NULL,
    `variantName` = NULL,
    initialize = function(
        `additionalInfo`=NULL, `callSetDbId`=NULL, `callSetName`=NULL, `genotype`=NULL, `genotype_likelihood`=NULL, `phaseSet`=NULL, `variantDbId`=NULL, `variantName`=NULL, ...
    ) {
      local.optional.var <- list(...)
      if (!is.null(`additionalInfo`)) {
        stopifnot(is.vector(`additionalInfo`), length(`additionalInfo`) != 0)
        sapply(`additionalInfo`, function(x) stopifnot(is.character(x)))
        self$`additionalInfo` <- `additionalInfo`
      }
      if (!is.null(`callSetDbId`)) {
        stopifnot(is.character(`callSetDbId`), length(`callSetDbId`) == 1)
        self$`callSetDbId` <- `callSetDbId`
      }
      if (!is.null(`callSetName`)) {
        stopifnot(is.character(`callSetName`), length(`callSetName`) == 1)
        self$`callSetName` <- `callSetName`
      }
      if (!is.null(`genotype`)) {
        stopifnot(R6::is.R6(`genotype`))
        self$`genotype` <- `genotype`
      }
      if (!is.null(`genotype_likelihood`)) {
        stopifnot(is.vector(`genotype_likelihood`), length(`genotype_likelihood`) != 0)
        sapply(`genotype_likelihood`, function(x) stopifnot(is.character(x)))
        self$`genotype_likelihood` <- `genotype_likelihood`
      }
      if (!is.null(`phaseSet`)) {
        stopifnot(is.character(`phaseSet`), length(`phaseSet`) == 1)
        self$`phaseSet` <- `phaseSet`
      }
      if (!is.null(`variantDbId`)) {
        stopifnot(is.character(`variantDbId`), length(`variantDbId`) == 1)
        self$`variantDbId` <- `variantDbId`
      }
      if (!is.null(`variantName`)) {
        stopifnot(is.character(`variantName`), length(`variantName`) == 1)
        self$`variantName` <- `variantName`
      }
    },
    toJSON = function() {
      CallObject <- list()
      if (!is.null(self$`additionalInfo`)) {
        CallObject[['additionalInfo']] <-
          self$`additionalInfo`
      }
      if (!is.null(self$`callSetDbId`)) {
        CallObject[['callSetDbId']] <-
          self$`callSetDbId`
      }
      if (!is.null(self$`callSetName`)) {
        CallObject[['callSetName']] <-
          self$`callSetName`
      }
      if (!is.null(self$`genotype`)) {
        CallObject[['genotype']] <-
          self$`genotype`$toJSON()
      }
      if (!is.null(self$`genotype_likelihood`)) {
        CallObject[['genotype_likelihood']] <-
          self$`genotype_likelihood`
      }
      if (!is.null(self$`phaseSet`)) {
        CallObject[['phaseSet']] <-
          self$`phaseSet`
      }
      if (!is.null(self$`variantDbId`)) {
        CallObject[['variantDbId']] <-
          self$`variantDbId`
      }
      if (!is.null(self$`variantName`)) {
        CallObject[['variantName']] <-
          self$`variantName`
      }

      CallObject
    },
    fromJSON = function(CallJson) {
      CallObject <- jsonlite::fromJSON(CallJson)
      if (!is.null(CallObject$`additionalInfo`)) {
        self$`additionalInfo` <- ApiClient$new()$deserializeObj(CallObject$`additionalInfo`, "map(character)", loadNamespace("openapi"))
      }
      if (!is.null(CallObject$`callSetDbId`)) {
        self$`callSetDbId` <- CallObject$`callSetDbId`
      }
      if (!is.null(CallObject$`callSetName`)) {
        self$`callSetName` <- CallObject$`callSetName`
      }
      if (!is.null(CallObject$`genotype`)) {
        genotypeObject <- ListValue$new()
        genotypeObject$fromJSON(jsonlite::toJSON(CallObject$genotype, auto_unbox = TRUE, digits = NA))
        self$`genotype` <- genotypeObject
      }
      if (!is.null(CallObject$`genotype_likelihood`)) {
        self$`genotype_likelihood` <- ApiClient$new()$deserializeObj(CallObject$`genotype_likelihood`, "array[numeric]", loadNamespace("openapi"))
      }
      if (!is.null(CallObject$`phaseSet`)) {
        self$`phaseSet` <- CallObject$`phaseSet`
      }
      if (!is.null(CallObject$`variantDbId`)) {
        self$`variantDbId` <- CallObject$`variantDbId`
      }
      if (!is.null(CallObject$`variantName`)) {
        self$`variantName` <- CallObject$`variantName`
      }
      self
    },
    toJSONString = function() {
      jsoncontent <- c(
        if (!is.null(self$`additionalInfo`)) {
        sprintf(
        '"additionalInfo":
          "%s"
        ',
        jsonlite::toJSON(lapply(self$`additionalInfo`, function(x){ x }), auto_unbox = TRUE, digits=NA)
        )},
        if (!is.null(self$`callSetDbId`)) {
        sprintf(
        '"callSetDbId":
          "%s"
                ',
        self$`callSetDbId`
        )},
        if (!is.null(self$`callSetName`)) {
        sprintf(
        '"callSetName":
          "%s"
                ',
        self$`callSetName`
        )},
        if (!is.null(self$`genotype`)) {
        sprintf(
        '"genotype":
        %s
        ',
        jsonlite::toJSON(self$`genotype`$toJSON(), auto_unbox=TRUE, digits = NA)
        )},
        if (!is.null(self$`genotype_likelihood`)) {
        sprintf(
        '"genotype_likelihood":
           [%s]
        ',
        paste(unlist(lapply(self$`genotype_likelihood`, function(x) paste0('"', x, '"'))), collapse=",")
        )},
        if (!is.null(self$`phaseSet`)) {
        sprintf(
        '"phaseSet":
          "%s"
                ',
        self$`phaseSet`
        )},
        if (!is.null(self$`variantDbId`)) {
        sprintf(
        '"variantDbId":
          "%s"
                ',
        self$`variantDbId`
        )},
        if (!is.null(self$`variantName`)) {
        sprintf(
        '"variantName":
          "%s"
                ',
        self$`variantName`
        )}
      )
      jsoncontent <- paste(jsoncontent, collapse = ",")
      paste('{', jsoncontent, '}', sep = "")
    },
    fromJSONString = function(CallJson) {
      CallObject <- jsonlite::fromJSON(CallJson)
      self$`additionalInfo` <- ApiClient$new()$deserializeObj(CallObject$`additionalInfo`, "map(character)", loadNamespace("openapi"))
      self$`callSetDbId` <- CallObject$`callSetDbId`
      self$`callSetName` <- CallObject$`callSetName`
      self$`genotype` <- ListValue$new()$fromJSON(jsonlite::toJSON(CallObject$genotype, auto_unbox = TRUE, digits = NA))
      self$`genotype_likelihood` <- ApiClient$new()$deserializeObj(CallObject$`genotype_likelihood`, "array[numeric]", loadNamespace("openapi"))
      self$`phaseSet` <- CallObject$`phaseSet`
      self$`variantDbId` <- CallObject$`variantDbId`
      self$`variantName` <- CallObject$`variantName`
      self
    }
  )
)

