# BrAPI-Core
#
# The Breeding API (BrAPI) is a Standardized REST ful Web Service API Specification for communicating Plant Breeding Data. BrAPI allows for easy data sharing between databases and tools involved in plant breeding. <div class=\"brapi-section\"> <h2 class=\"brapi-section-title\">General Reference Documentation</h2> <div class=\"gen-info-link\"><a href=\"https://github.com/plantbreeding/API/blob/master/Specification/GeneralInfo/URL_Structure.md\">URL Structure</a></div> <div class=\"gen-info-link\"><a href=\"https://github.com/plantbreeding/API/blob/master/Specification/GeneralInfo/Response_Structure.md\">Response Structure</a></div> <div class=\"gen-info-link\"><a href=\"https://github.com/plantbreeding/API/blob/master/Specification/GeneralInfo/Date_Time_Encoding.md\">Date/Time Encoding</a></div> <div class=\"gen-info-link\"><a href=\"https://github.com/plantbreeding/API/blob/master/Specification/GeneralInfo/Location_Encoding.md\">Location Encoding</a></div> <div class=\"gen-info-link\"><a href=\"https://github.com/plantbreeding/API/blob/master/Specification/GeneralInfo/Error_Handling.md\">Error Handling</a></div> <div class=\"gen-info-link\"><a href=\"https://github.com/plantbreeding/API/blob/master/Specification/GeneralInfo/Search_Services.md\">Search Services</a></div> </div>  <div class=\"current-brapi-section brapi-section\"> <h2 class=\"brapi-section-title\">BrAPI Core</h2> <div class=\"brapi-section-description\">The BrAPI Core module contains high level entities used for organization and management. This includes Programs, Trials, Studies, Locations, People, and Lists</div> <div class=\"version-number\">V2.0</div> <div class=\"link-btn\"><a href=\"https://github.com/plantbreeding/API/tree/master/Specification/BrAPI-Core\">GitHub</a></div> <div class=\"link-btn\"><a href=\"https://app.swaggerhub.com/apis/PlantBreedingAPI/BrAPI-Core\">SwaggerHub</a></div> <div class=\"link-btn\"><a href=\"https://brapicore.docs.apiary.io\">Apiary</a></div> <div class=\"stop-float\"></div> </div>  <div class=\"brapi-section\"> <h2 class=\"brapi-section-title\">BrAPI Phenotyping</h2> <div class=\"brapi-section-description\">The BrAPI Phenotyping module contains entities related to phenotypic observations. This includes Observation Units, Observations, Observation Variables, Traits, Scales, Methods, and Images</div> <div class=\"version-number\">V2.0</div> <div class=\"link-btn\"><a href=\"https://github.com/plantbreeding/API/tree/master/Specification/BrAPI-Phenotyping\">GitHub</a></div> <div class=\"link-btn\"><a href=\"https://app.swaggerhub.com/apis/PlantBreedingAPI/BrAPI-Phenotyping\">SwaggerHub</a></div> <div class=\"link-btn\"><a href=\"https://brapiphenotyping.docs.apiary.io\">Apiary</a></div> <div class=\"stop-float\"></div> </div>  <div class=\"brapi-section\"> <h2 class=\"brapi-section-title\">BrAPI Genotyping</h2> <div class=\"brapi-section-description\">The BrAPI Genotyping module contains entities related to genotyping analysis. This includes Samples, Markers, Variant Sets, Variants, Call Sets, Calls, References, Reads, and Vendor Orders</div> <div class=\"version-number\">V2.0</div> <div class=\"link-btn\"><a href=\"https://github.com/plantbreeding/API/tree/master/Specification/BrAPI-Genotyping\">GitHub</a></div> <div class=\"link-btn\"><a href=\"https://app.swaggerhub.com/apis/PlantBreedingAPI/BrAPI-Genotyping\">SwaggerHub</a></div> <div class=\"link-btn\"><a href=\"https://brapigenotyping.docs.apiary.io\">Apiary</a></div> <div class=\"stop-float\"></div> </div>  <div class=\"brapi-section\"> <h2 class=\"brapi-section-title\">BrAPI Germplasm</h2> <div class=\"brapi-section-description\">The BrAPI Germplasm module contains entities related to germplasm management. This includes Germplasm, Germplasm Attributes, Seed Lots, Crosses, Pedigree, and Progeny</div> <div class=\"version-number\">V2.0</div> <div class=\"link-btn\"><a href=\"https://github.com/plantbreeding/API/tree/master/Specification/BrAPI-Germplasm\">GitHub</a></div> <div class=\"link-btn\"><a href=\"https://app.swaggerhub.com/apis/PlantBreedingAPI/BrAPI-Germplasm\">SwaggerHub</a></div> <div class=\"link-btn\"><a href=\"https://brapigermplasm.docs.apiary.io\">Apiary</a></div> <div class=\"stop-float\"></div> </div>  <style> .link-btn{ float: left;  margin: 2px 10px 0 0;  padding: 0 5px;  border-radius: 5px;  background-color: #ddd; } .stop-float{   clear: both; } .version-number{   float: left;    margin: 5px 10px 0 5px; } .brapi-section-title{   margin: 0 10px 0 0;   font-size: 20px; } .current-brapi-section{   font-weight: bolder;   border-radius: 5px;    background-color: #ddd; } .brapi-section{   padding: 5px 5px;  } .brapi-section-description{   margin: 5px 0 0 5px; } </style>
#
# The version of the OpenAPI document: 2.0
# 
# Generated by: https://openapi-generator.tech

#' @docType class
#' @title CrossParent
#'
#' @description CrossParent Class
#'
#' @format An \code{R6Class} generator object
#'
#' @field germplasmDbId  character [optional]
#'
#' @field germplasmName  character [optional]
#'
#' @field observationUnitDbId  character [optional]
#'
#' @field observationUnitName  character [optional]
#'
#' @field parentType  character [optional]
#'
#' @importFrom R6 R6Class
#' @importFrom jsonlite fromJSON toJSON
#' @export
CrossParent <- R6::R6Class(
  'CrossParent',
  public = list(
    `germplasmDbId` = NULL,
    `germplasmName` = NULL,
    `observationUnitDbId` = NULL,
    `observationUnitName` = NULL,
    `parentType` = NULL,
    initialize = function(
        `germplasmDbId`=NULL, `germplasmName`=NULL, `observationUnitDbId`=NULL, `observationUnitName`=NULL, `parentType`=NULL, ...
    ) {
      local.optional.var <- list(...)
      if (!is.null(`germplasmDbId`)) {
        stopifnot(is.character(`germplasmDbId`), length(`germplasmDbId`) == 1)
        self$`germplasmDbId` <- `germplasmDbId`
      }
      if (!is.null(`germplasmName`)) {
        stopifnot(is.character(`germplasmName`), length(`germplasmName`) == 1)
        self$`germplasmName` <- `germplasmName`
      }
      if (!is.null(`observationUnitDbId`)) {
        stopifnot(is.character(`observationUnitDbId`), length(`observationUnitDbId`) == 1)
        self$`observationUnitDbId` <- `observationUnitDbId`
      }
      if (!is.null(`observationUnitName`)) {
        stopifnot(is.character(`observationUnitName`), length(`observationUnitName`) == 1)
        self$`observationUnitName` <- `observationUnitName`
      }
      if (!is.null(`parentType`)) {
        stopifnot(is.character(`parentType`), length(`parentType`) == 1)
        self$`parentType` <- `parentType`
      }
    },
    toJSON = function() {
      CrossParentObject <- list()
      if (!is.null(self$`germplasmDbId`)) {
        CrossParentObject[['germplasmDbId']] <-
          self$`germplasmDbId`
      }
      if (!is.null(self$`germplasmName`)) {
        CrossParentObject[['germplasmName']] <-
          self$`germplasmName`
      }
      if (!is.null(self$`observationUnitDbId`)) {
        CrossParentObject[['observationUnitDbId']] <-
          self$`observationUnitDbId`
      }
      if (!is.null(self$`observationUnitName`)) {
        CrossParentObject[['observationUnitName']] <-
          self$`observationUnitName`
      }
      if (!is.null(self$`parentType`)) {
        CrossParentObject[['parentType']] <-
          self$`parentType`
      }

      CrossParentObject
    },
    fromJSON = function(CrossParentJson) {
      CrossParentObject <- jsonlite::fromJSON(CrossParentJson)
      if (!is.null(CrossParentObject$`germplasmDbId`)) {
        self$`germplasmDbId` <- CrossParentObject$`germplasmDbId`
      }
      if (!is.null(CrossParentObject$`germplasmName`)) {
        self$`germplasmName` <- CrossParentObject$`germplasmName`
      }
      if (!is.null(CrossParentObject$`observationUnitDbId`)) {
        self$`observationUnitDbId` <- CrossParentObject$`observationUnitDbId`
      }
      if (!is.null(CrossParentObject$`observationUnitName`)) {
        self$`observationUnitName` <- CrossParentObject$`observationUnitName`
      }
      if (!is.null(CrossParentObject$`parentType`)) {
        self$`parentType` <- CrossParentObject$`parentType`
      }
      self
    },
    toJSONString = function() {
      jsoncontent <- c(
        if (!is.null(self$`germplasmDbId`)) {
        sprintf(
        '"germplasmDbId":
          "%s"
                ',
        self$`germplasmDbId`
        )},
        if (!is.null(self$`germplasmName`)) {
        sprintf(
        '"germplasmName":
          "%s"
                ',
        self$`germplasmName`
        )},
        if (!is.null(self$`observationUnitDbId`)) {
        sprintf(
        '"observationUnitDbId":
          "%s"
                ',
        self$`observationUnitDbId`
        )},
        if (!is.null(self$`observationUnitName`)) {
        sprintf(
        '"observationUnitName":
          "%s"
                ',
        self$`observationUnitName`
        )},
        if (!is.null(self$`parentType`)) {
        sprintf(
        '"parentType":
          "%s"
                ',
        self$`parentType`
        )}
      )
      jsoncontent <- paste(jsoncontent, collapse = ",")
      paste('{', jsoncontent, '}', sep = "")
    },
    fromJSONString = function(CrossParentJson) {
      CrossParentObject <- jsonlite::fromJSON(CrossParentJson)
      self$`germplasmDbId` <- CrossParentObject$`germplasmDbId`
      self$`germplasmName` <- CrossParentObject$`germplasmName`
      self$`observationUnitDbId` <- CrossParentObject$`observationUnitDbId`
      self$`observationUnitName` <- CrossParentObject$`observationUnitName`
      self$`parentType` <- CrossParentObject$`parentType`
      self
    }
  )
)

