# BrAPI-Core
#
# The Breeding API (BrAPI) is a Standardized REST ful Web Service API Specification for communicating Plant Breeding Data. BrAPI allows for easy data sharing between databases and tools involved in plant breeding. <div class=\"brapi-section\"> <h2 class=\"brapi-section-title\">General Reference Documentation</h2> <div class=\"gen-info-link\"><a href=\"https://github.com/plantbreeding/API/blob/master/Specification/GeneralInfo/URL_Structure.md\">URL Structure</a></div> <div class=\"gen-info-link\"><a href=\"https://github.com/plantbreeding/API/blob/master/Specification/GeneralInfo/Response_Structure.md\">Response Structure</a></div> <div class=\"gen-info-link\"><a href=\"https://github.com/plantbreeding/API/blob/master/Specification/GeneralInfo/Date_Time_Encoding.md\">Date/Time Encoding</a></div> <div class=\"gen-info-link\"><a href=\"https://github.com/plantbreeding/API/blob/master/Specification/GeneralInfo/Location_Encoding.md\">Location Encoding</a></div> <div class=\"gen-info-link\"><a href=\"https://github.com/plantbreeding/API/blob/master/Specification/GeneralInfo/Error_Handling.md\">Error Handling</a></div> <div class=\"gen-info-link\"><a href=\"https://github.com/plantbreeding/API/blob/master/Specification/GeneralInfo/Search_Services.md\">Search Services</a></div> </div>  <div class=\"current-brapi-section brapi-section\"> <h2 class=\"brapi-section-title\">BrAPI Core</h2> <div class=\"brapi-section-description\">The BrAPI Core module contains high level entities used for organization and management. This includes Programs, Trials, Studies, Locations, People, and Lists</div> <div class=\"version-number\">V2.0</div> <div class=\"link-btn\"><a href=\"https://github.com/plantbreeding/API/tree/master/Specification/BrAPI-Core\">GitHub</a></div> <div class=\"link-btn\"><a href=\"https://app.swaggerhub.com/apis/PlantBreedingAPI/BrAPI-Core\">SwaggerHub</a></div> <div class=\"link-btn\"><a href=\"https://brapicore.docs.apiary.io\">Apiary</a></div> <div class=\"stop-float\"></div> </div>  <div class=\"brapi-section\"> <h2 class=\"brapi-section-title\">BrAPI Phenotyping</h2> <div class=\"brapi-section-description\">The BrAPI Phenotyping module contains entities related to phenotypic observations. This includes Observation Units, Observations, Observation Variables, Traits, Scales, Methods, and Images</div> <div class=\"version-number\">V2.0</div> <div class=\"link-btn\"><a href=\"https://github.com/plantbreeding/API/tree/master/Specification/BrAPI-Phenotyping\">GitHub</a></div> <div class=\"link-btn\"><a href=\"https://app.swaggerhub.com/apis/PlantBreedingAPI/BrAPI-Phenotyping\">SwaggerHub</a></div> <div class=\"link-btn\"><a href=\"https://brapiphenotyping.docs.apiary.io\">Apiary</a></div> <div class=\"stop-float\"></div> </div>  <div class=\"brapi-section\"> <h2 class=\"brapi-section-title\">BrAPI Genotyping</h2> <div class=\"brapi-section-description\">The BrAPI Genotyping module contains entities related to genotyping analysis. This includes Samples, Markers, Variant Sets, Variants, Call Sets, Calls, References, Reads, and Vendor Orders</div> <div class=\"version-number\">V2.0</div> <div class=\"link-btn\"><a href=\"https://github.com/plantbreeding/API/tree/master/Specification/BrAPI-Genotyping\">GitHub</a></div> <div class=\"link-btn\"><a href=\"https://app.swaggerhub.com/apis/PlantBreedingAPI/BrAPI-Genotyping\">SwaggerHub</a></div> <div class=\"link-btn\"><a href=\"https://brapigenotyping.docs.apiary.io\">Apiary</a></div> <div class=\"stop-float\"></div> </div>  <div class=\"brapi-section\"> <h2 class=\"brapi-section-title\">BrAPI Germplasm</h2> <div class=\"brapi-section-description\">The BrAPI Germplasm module contains entities related to germplasm management. This includes Germplasm, Germplasm Attributes, Seed Lots, Crosses, Pedigree, and Progeny</div> <div class=\"version-number\">V2.0</div> <div class=\"link-btn\"><a href=\"https://github.com/plantbreeding/API/tree/master/Specification/BrAPI-Germplasm\">GitHub</a></div> <div class=\"link-btn\"><a href=\"https://app.swaggerhub.com/apis/PlantBreedingAPI/BrAPI-Germplasm\">SwaggerHub</a></div> <div class=\"link-btn\"><a href=\"https://brapigermplasm.docs.apiary.io\">Apiary</a></div> <div class=\"stop-float\"></div> </div>  <style> .link-btn{ float: left;  margin: 2px 10px 0 0;  padding: 0 5px;  border-radius: 5px;  background-color: #ddd; } .stop-float{   clear: both; } .version-number{   float: left;    margin: 5px 10px 0 5px; } .brapi-section-title{   margin: 0 10px 0 0;   font-size: 20px; } .current-brapi-section{   font-weight: bolder;   border-radius: 5px;    background-color: #ddd; } .brapi-section{   padding: 5px 5px;  } .brapi-section-description{   margin: 5px 0 0 5px; } </style>
#
# The version of the OpenAPI document: 2.0
# 
# Generated by: https://openapi-generator.tech

#' @docType class
#' @title DataLink
#'
#' @description DataLink Class
#'
#' @format An \code{R6Class} generator object
#'
#' @field dataFormat  character [optional]
#'
#' @field description  character [optional]
#'
#' @field fileFormat  character [optional]
#'
#' @field name  character [optional]
#'
#' @field provenance  character [optional]
#'
#' @field scientificType  character [optional]
#'
#' @field url  character [optional]
#'
#' @field version  character [optional]
#'
#' @importFrom R6 R6Class
#' @importFrom jsonlite fromJSON toJSON
#' @export
DataLink <- R6::R6Class(
  'DataLink',
  public = list(
    `dataFormat` = NULL,
    `description` = NULL,
    `fileFormat` = NULL,
    `name` = NULL,
    `provenance` = NULL,
    `scientificType` = NULL,
    `url` = NULL,
    `version` = NULL,
    initialize = function(
        `dataFormat`=NULL, `description`=NULL, `fileFormat`=NULL, `name`=NULL, `provenance`=NULL, `scientificType`=NULL, `url`=NULL, `version`=NULL, ...
    ) {
      local.optional.var <- list(...)
      if (!is.null(`dataFormat`)) {
        stopifnot(is.character(`dataFormat`), length(`dataFormat`) == 1)
        self$`dataFormat` <- `dataFormat`
      }
      if (!is.null(`description`)) {
        stopifnot(is.character(`description`), length(`description`) == 1)
        self$`description` <- `description`
      }
      if (!is.null(`fileFormat`)) {
        stopifnot(is.character(`fileFormat`), length(`fileFormat`) == 1)
        self$`fileFormat` <- `fileFormat`
      }
      if (!is.null(`name`)) {
        stopifnot(is.character(`name`), length(`name`) == 1)
        self$`name` <- `name`
      }
      if (!is.null(`provenance`)) {
        stopifnot(is.character(`provenance`), length(`provenance`) == 1)
        self$`provenance` <- `provenance`
      }
      if (!is.null(`scientificType`)) {
        stopifnot(is.character(`scientificType`), length(`scientificType`) == 1)
        self$`scientificType` <- `scientificType`
      }
      if (!is.null(`url`)) {
        stopifnot(is.character(`url`), length(`url`) == 1)
        self$`url` <- `url`
      }
      if (!is.null(`version`)) {
        stopifnot(is.character(`version`), length(`version`) == 1)
        self$`version` <- `version`
      }
    },
    toJSON = function() {
      DataLinkObject <- list()
      if (!is.null(self$`dataFormat`)) {
        DataLinkObject[['dataFormat']] <-
          self$`dataFormat`
      }
      if (!is.null(self$`description`)) {
        DataLinkObject[['description']] <-
          self$`description`
      }
      if (!is.null(self$`fileFormat`)) {
        DataLinkObject[['fileFormat']] <-
          self$`fileFormat`
      }
      if (!is.null(self$`name`)) {
        DataLinkObject[['name']] <-
          self$`name`
      }
      if (!is.null(self$`provenance`)) {
        DataLinkObject[['provenance']] <-
          self$`provenance`
      }
      if (!is.null(self$`scientificType`)) {
        DataLinkObject[['scientificType']] <-
          self$`scientificType`
      }
      if (!is.null(self$`url`)) {
        DataLinkObject[['url']] <-
          self$`url`
      }
      if (!is.null(self$`version`)) {
        DataLinkObject[['version']] <-
          self$`version`
      }

      DataLinkObject
    },
    fromJSON = function(DataLinkJson) {
      DataLinkObject <- jsonlite::fromJSON(DataLinkJson)
      if (!is.null(DataLinkObject$`dataFormat`)) {
        self$`dataFormat` <- DataLinkObject$`dataFormat`
      }
      if (!is.null(DataLinkObject$`description`)) {
        self$`description` <- DataLinkObject$`description`
      }
      if (!is.null(DataLinkObject$`fileFormat`)) {
        self$`fileFormat` <- DataLinkObject$`fileFormat`
      }
      if (!is.null(DataLinkObject$`name`)) {
        self$`name` <- DataLinkObject$`name`
      }
      if (!is.null(DataLinkObject$`provenance`)) {
        self$`provenance` <- DataLinkObject$`provenance`
      }
      if (!is.null(DataLinkObject$`scientificType`)) {
        self$`scientificType` <- DataLinkObject$`scientificType`
      }
      if (!is.null(DataLinkObject$`url`)) {
        self$`url` <- DataLinkObject$`url`
      }
      if (!is.null(DataLinkObject$`version`)) {
        self$`version` <- DataLinkObject$`version`
      }
      self
    },
    toJSONString = function() {
      jsoncontent <- c(
        if (!is.null(self$`dataFormat`)) {
        sprintf(
        '"dataFormat":
          "%s"
                ',
        self$`dataFormat`
        )},
        if (!is.null(self$`description`)) {
        sprintf(
        '"description":
          "%s"
                ',
        self$`description`
        )},
        if (!is.null(self$`fileFormat`)) {
        sprintf(
        '"fileFormat":
          "%s"
                ',
        self$`fileFormat`
        )},
        if (!is.null(self$`name`)) {
        sprintf(
        '"name":
          "%s"
                ',
        self$`name`
        )},
        if (!is.null(self$`provenance`)) {
        sprintf(
        '"provenance":
          "%s"
                ',
        self$`provenance`
        )},
        if (!is.null(self$`scientificType`)) {
        sprintf(
        '"scientificType":
          "%s"
                ',
        self$`scientificType`
        )},
        if (!is.null(self$`url`)) {
        sprintf(
        '"url":
          "%s"
                ',
        self$`url`
        )},
        if (!is.null(self$`version`)) {
        sprintf(
        '"version":
          "%s"
                ',
        self$`version`
        )}
      )
      jsoncontent <- paste(jsoncontent, collapse = ",")
      paste('{', jsoncontent, '}', sep = "")
    },
    fromJSONString = function(DataLinkJson) {
      DataLinkObject <- jsonlite::fromJSON(DataLinkJson)
      self$`dataFormat` <- DataLinkObject$`dataFormat`
      self$`description` <- DataLinkObject$`description`
      self$`fileFormat` <- DataLinkObject$`fileFormat`
      self$`name` <- DataLinkObject$`name`
      self$`provenance` <- DataLinkObject$`provenance`
      self$`scientificType` <- DataLinkObject$`scientificType`
      self$`url` <- DataLinkObject$`url`
      self$`version` <- DataLinkObject$`version`
      self
    }
  )
)

