# BrAPI-Core
#
# The Breeding API (BrAPI) is a Standardized REST ful Web Service API Specification for communicating Plant Breeding Data. BrAPI allows for easy data sharing between databases and tools involved in plant breeding. <div class=\"brapi-section\"> <h2 class=\"brapi-section-title\">General Reference Documentation</h2> <div class=\"gen-info-link\"><a href=\"https://github.com/plantbreeding/API/blob/master/Specification/GeneralInfo/URL_Structure.md\">URL Structure</a></div> <div class=\"gen-info-link\"><a href=\"https://github.com/plantbreeding/API/blob/master/Specification/GeneralInfo/Response_Structure.md\">Response Structure</a></div> <div class=\"gen-info-link\"><a href=\"https://github.com/plantbreeding/API/blob/master/Specification/GeneralInfo/Date_Time_Encoding.md\">Date/Time Encoding</a></div> <div class=\"gen-info-link\"><a href=\"https://github.com/plantbreeding/API/blob/master/Specification/GeneralInfo/Location_Encoding.md\">Location Encoding</a></div> <div class=\"gen-info-link\"><a href=\"https://github.com/plantbreeding/API/blob/master/Specification/GeneralInfo/Error_Handling.md\">Error Handling</a></div> <div class=\"gen-info-link\"><a href=\"https://github.com/plantbreeding/API/blob/master/Specification/GeneralInfo/Search_Services.md\">Search Services</a></div> </div>  <div class=\"current-brapi-section brapi-section\"> <h2 class=\"brapi-section-title\">BrAPI Core</h2> <div class=\"brapi-section-description\">The BrAPI Core module contains high level entities used for organization and management. This includes Programs, Trials, Studies, Locations, People, and Lists</div> <div class=\"version-number\">V2.0</div> <div class=\"link-btn\"><a href=\"https://github.com/plantbreeding/API/tree/master/Specification/BrAPI-Core\">GitHub</a></div> <div class=\"link-btn\"><a href=\"https://app.swaggerhub.com/apis/PlantBreedingAPI/BrAPI-Core\">SwaggerHub</a></div> <div class=\"link-btn\"><a href=\"https://brapicore.docs.apiary.io\">Apiary</a></div> <div class=\"stop-float\"></div> </div>  <div class=\"brapi-section\"> <h2 class=\"brapi-section-title\">BrAPI Phenotyping</h2> <div class=\"brapi-section-description\">The BrAPI Phenotyping module contains entities related to phenotypic observations. This includes Observation Units, Observations, Observation Variables, Traits, Scales, Methods, and Images</div> <div class=\"version-number\">V2.0</div> <div class=\"link-btn\"><a href=\"https://github.com/plantbreeding/API/tree/master/Specification/BrAPI-Phenotyping\">GitHub</a></div> <div class=\"link-btn\"><a href=\"https://app.swaggerhub.com/apis/PlantBreedingAPI/BrAPI-Phenotyping\">SwaggerHub</a></div> <div class=\"link-btn\"><a href=\"https://brapiphenotyping.docs.apiary.io\">Apiary</a></div> <div class=\"stop-float\"></div> </div>  <div class=\"brapi-section\"> <h2 class=\"brapi-section-title\">BrAPI Genotyping</h2> <div class=\"brapi-section-description\">The BrAPI Genotyping module contains entities related to genotyping analysis. This includes Samples, Markers, Variant Sets, Variants, Call Sets, Calls, References, Reads, and Vendor Orders</div> <div class=\"version-number\">V2.0</div> <div class=\"link-btn\"><a href=\"https://github.com/plantbreeding/API/tree/master/Specification/BrAPI-Genotyping\">GitHub</a></div> <div class=\"link-btn\"><a href=\"https://app.swaggerhub.com/apis/PlantBreedingAPI/BrAPI-Genotyping\">SwaggerHub</a></div> <div class=\"link-btn\"><a href=\"https://brapigenotyping.docs.apiary.io\">Apiary</a></div> <div class=\"stop-float\"></div> </div>  <div class=\"brapi-section\"> <h2 class=\"brapi-section-title\">BrAPI Germplasm</h2> <div class=\"brapi-section-description\">The BrAPI Germplasm module contains entities related to germplasm management. This includes Germplasm, Germplasm Attributes, Seed Lots, Crosses, Pedigree, and Progeny</div> <div class=\"version-number\">V2.0</div> <div class=\"link-btn\"><a href=\"https://github.com/plantbreeding/API/tree/master/Specification/BrAPI-Germplasm\">GitHub</a></div> <div class=\"link-btn\"><a href=\"https://app.swaggerhub.com/apis/PlantBreedingAPI/BrAPI-Germplasm\">SwaggerHub</a></div> <div class=\"link-btn\"><a href=\"https://brapigermplasm.docs.apiary.io\">Apiary</a></div> <div class=\"stop-float\"></div> </div>  <style> .link-btn{ float: left;  margin: 2px 10px 0 0;  padding: 0 5px;  border-radius: 5px;  background-color: #ddd; } .stop-float{   clear: both; } .version-number{   float: left;    margin: 5px 10px 0 5px; } .brapi-section-title{   margin: 0 10px 0 0;   font-size: 20px; } .current-brapi-section{   font-weight: bolder;   border-radius: 5px;    background-color: #ddd; } .brapi-section{   padding: 5px 5px;  } .brapi-section-description{   margin: 5px 0 0 5px; } </style>
#
# The version of the OpenAPI document: 2.0
# 
# Generated by: https://openapi-generator.tech

#' @docType class
#' @title GenomeMap
#'
#' @description GenomeMap Class
#'
#' @format An \code{R6Class} generator object
#'
#' @field additionalInfo  named list( character ) [optional]
#'
#' @field comments  character [optional]
#'
#' @field commonCropName  character 
#'
#' @field documentationURL  character [optional]
#'
#' @field linkageGroupCount  integer [optional]
#'
#' @field mapDbId  character 
#'
#' @field mapName  character [optional]
#'
#' @field mapPUI  character [optional]
#'
#' @field markerCount  integer [optional]
#'
#' @field publishedDate  character [optional]
#'
#' @field scientificName  character [optional]
#'
#' @field type  character 
#'
#' @field unit  character [optional]
#'
#' @importFrom R6 R6Class
#' @importFrom jsonlite fromJSON toJSON
#' @export
GenomeMap <- R6::R6Class(
  'GenomeMap',
  public = list(
    `additionalInfo` = NULL,
    `comments` = NULL,
    `commonCropName` = NULL,
    `documentationURL` = NULL,
    `linkageGroupCount` = NULL,
    `mapDbId` = NULL,
    `mapName` = NULL,
    `mapPUI` = NULL,
    `markerCount` = NULL,
    `publishedDate` = NULL,
    `scientificName` = NULL,
    `type` = NULL,
    `unit` = NULL,
    initialize = function(
        `commonCropName`, `mapDbId`, `type`, `additionalInfo`=NULL, `comments`=NULL, `documentationURL`=NULL, `linkageGroupCount`=NULL, `mapName`=NULL, `mapPUI`=NULL, `markerCount`=NULL, `publishedDate`=NULL, `scientificName`=NULL, `unit`=NULL, ...
    ) {
      local.optional.var <- list(...)
      if (!missing(`commonCropName`)) {
        stopifnot(is.character(`commonCropName`), length(`commonCropName`) == 1)
        self$`commonCropName` <- `commonCropName`
      }
      if (!missing(`mapDbId`)) {
        stopifnot(is.character(`mapDbId`), length(`mapDbId`) == 1)
        self$`mapDbId` <- `mapDbId`
      }
      if (!missing(`type`)) {
        stopifnot(is.character(`type`), length(`type`) == 1)
        self$`type` <- `type`
      }
      if (!is.null(`additionalInfo`)) {
        stopifnot(is.vector(`additionalInfo`), length(`additionalInfo`) != 0)
        sapply(`additionalInfo`, function(x) stopifnot(is.character(x)))
        self$`additionalInfo` <- `additionalInfo`
      }
      if (!is.null(`comments`)) {
        stopifnot(is.character(`comments`), length(`comments`) == 1)
        self$`comments` <- `comments`
      }
      if (!is.null(`documentationURL`)) {
        stopifnot(is.character(`documentationURL`), length(`documentationURL`) == 1)
        self$`documentationURL` <- `documentationURL`
      }
      if (!is.null(`linkageGroupCount`)) {
        stopifnot(is.numeric(`linkageGroupCount`), length(`linkageGroupCount`) == 1)
        self$`linkageGroupCount` <- `linkageGroupCount`
      }
      if (!is.null(`mapName`)) {
        stopifnot(is.character(`mapName`), length(`mapName`) == 1)
        self$`mapName` <- `mapName`
      }
      if (!is.null(`mapPUI`)) {
        stopifnot(is.character(`mapPUI`), length(`mapPUI`) == 1)
        self$`mapPUI` <- `mapPUI`
      }
      if (!is.null(`markerCount`)) {
        stopifnot(is.numeric(`markerCount`), length(`markerCount`) == 1)
        self$`markerCount` <- `markerCount`
      }
      if (!is.null(`publishedDate`)) {
        stopifnot(is.character(`publishedDate`), length(`publishedDate`) == 1)
        self$`publishedDate` <- `publishedDate`
      }
      if (!is.null(`scientificName`)) {
        stopifnot(is.character(`scientificName`), length(`scientificName`) == 1)
        self$`scientificName` <- `scientificName`
      }
      if (!is.null(`unit`)) {
        stopifnot(is.character(`unit`), length(`unit`) == 1)
        self$`unit` <- `unit`
      }
    },
    toJSON = function() {
      GenomeMapObject <- list()
      if (!is.null(self$`additionalInfo`)) {
        GenomeMapObject[['additionalInfo']] <-
          self$`additionalInfo`
      }
      if (!is.null(self$`comments`)) {
        GenomeMapObject[['comments']] <-
          self$`comments`
      }
      if (!is.null(self$`commonCropName`)) {
        GenomeMapObject[['commonCropName']] <-
          self$`commonCropName`
      }
      if (!is.null(self$`documentationURL`)) {
        GenomeMapObject[['documentationURL']] <-
          self$`documentationURL`
      }
      if (!is.null(self$`linkageGroupCount`)) {
        GenomeMapObject[['linkageGroupCount']] <-
          self$`linkageGroupCount`
      }
      if (!is.null(self$`mapDbId`)) {
        GenomeMapObject[['mapDbId']] <-
          self$`mapDbId`
      }
      if (!is.null(self$`mapName`)) {
        GenomeMapObject[['mapName']] <-
          self$`mapName`
      }
      if (!is.null(self$`mapPUI`)) {
        GenomeMapObject[['mapPUI']] <-
          self$`mapPUI`
      }
      if (!is.null(self$`markerCount`)) {
        GenomeMapObject[['markerCount']] <-
          self$`markerCount`
      }
      if (!is.null(self$`publishedDate`)) {
        GenomeMapObject[['publishedDate']] <-
          self$`publishedDate`
      }
      if (!is.null(self$`scientificName`)) {
        GenomeMapObject[['scientificName']] <-
          self$`scientificName`
      }
      if (!is.null(self$`type`)) {
        GenomeMapObject[['type']] <-
          self$`type`
      }
      if (!is.null(self$`unit`)) {
        GenomeMapObject[['unit']] <-
          self$`unit`
      }

      GenomeMapObject
    },
    fromJSON = function(GenomeMapJson) {
      GenomeMapObject <- jsonlite::fromJSON(GenomeMapJson)
      if (!is.null(GenomeMapObject$`additionalInfo`)) {
        self$`additionalInfo` <- ApiClient$new()$deserializeObj(GenomeMapObject$`additionalInfo`, "map(character)", loadNamespace("openapi"))
      }
      if (!is.null(GenomeMapObject$`comments`)) {
        self$`comments` <- GenomeMapObject$`comments`
      }
      if (!is.null(GenomeMapObject$`commonCropName`)) {
        self$`commonCropName` <- GenomeMapObject$`commonCropName`
      }
      if (!is.null(GenomeMapObject$`documentationURL`)) {
        self$`documentationURL` <- GenomeMapObject$`documentationURL`
      }
      if (!is.null(GenomeMapObject$`linkageGroupCount`)) {
        self$`linkageGroupCount` <- GenomeMapObject$`linkageGroupCount`
      }
      if (!is.null(GenomeMapObject$`mapDbId`)) {
        self$`mapDbId` <- GenomeMapObject$`mapDbId`
      }
      if (!is.null(GenomeMapObject$`mapName`)) {
        self$`mapName` <- GenomeMapObject$`mapName`
      }
      if (!is.null(GenomeMapObject$`mapPUI`)) {
        self$`mapPUI` <- GenomeMapObject$`mapPUI`
      }
      if (!is.null(GenomeMapObject$`markerCount`)) {
        self$`markerCount` <- GenomeMapObject$`markerCount`
      }
      if (!is.null(GenomeMapObject$`publishedDate`)) {
        self$`publishedDate` <- GenomeMapObject$`publishedDate`
      }
      if (!is.null(GenomeMapObject$`scientificName`)) {
        self$`scientificName` <- GenomeMapObject$`scientificName`
      }
      if (!is.null(GenomeMapObject$`type`)) {
        self$`type` <- GenomeMapObject$`type`
      }
      if (!is.null(GenomeMapObject$`unit`)) {
        self$`unit` <- GenomeMapObject$`unit`
      }
      self
    },
    toJSONString = function() {
      jsoncontent <- c(
        if (!is.null(self$`additionalInfo`)) {
        sprintf(
        '"additionalInfo":
          "%s"
        ',
        jsonlite::toJSON(lapply(self$`additionalInfo`, function(x){ x }), auto_unbox = TRUE, digits=NA)
        )},
        if (!is.null(self$`comments`)) {
        sprintf(
        '"comments":
          "%s"
                ',
        self$`comments`
        )},
        if (!is.null(self$`commonCropName`)) {
        sprintf(
        '"commonCropName":
          "%s"
                ',
        self$`commonCropName`
        )},
        if (!is.null(self$`documentationURL`)) {
        sprintf(
        '"documentationURL":
          "%s"
                ',
        self$`documentationURL`
        )},
        if (!is.null(self$`linkageGroupCount`)) {
        sprintf(
        '"linkageGroupCount":
          %d
                ',
        self$`linkageGroupCount`
        )},
        if (!is.null(self$`mapDbId`)) {
        sprintf(
        '"mapDbId":
          "%s"
                ',
        self$`mapDbId`
        )},
        if (!is.null(self$`mapName`)) {
        sprintf(
        '"mapName":
          "%s"
                ',
        self$`mapName`
        )},
        if (!is.null(self$`mapPUI`)) {
        sprintf(
        '"mapPUI":
          "%s"
                ',
        self$`mapPUI`
        )},
        if (!is.null(self$`markerCount`)) {
        sprintf(
        '"markerCount":
          %d
                ',
        self$`markerCount`
        )},
        if (!is.null(self$`publishedDate`)) {
        sprintf(
        '"publishedDate":
          "%s"
                ',
        self$`publishedDate`
        )},
        if (!is.null(self$`scientificName`)) {
        sprintf(
        '"scientificName":
          "%s"
                ',
        self$`scientificName`
        )},
        if (!is.null(self$`type`)) {
        sprintf(
        '"type":
          "%s"
                ',
        self$`type`
        )},
        if (!is.null(self$`unit`)) {
        sprintf(
        '"unit":
          "%s"
                ',
        self$`unit`
        )}
      )
      jsoncontent <- paste(jsoncontent, collapse = ",")
      paste('{', jsoncontent, '}', sep = "")
    },
    fromJSONString = function(GenomeMapJson) {
      GenomeMapObject <- jsonlite::fromJSON(GenomeMapJson)
      self$`additionalInfo` <- ApiClient$new()$deserializeObj(GenomeMapObject$`additionalInfo`, "map(character)", loadNamespace("openapi"))
      self$`comments` <- GenomeMapObject$`comments`
      self$`commonCropName` <- GenomeMapObject$`commonCropName`
      self$`documentationURL` <- GenomeMapObject$`documentationURL`
      self$`linkageGroupCount` <- GenomeMapObject$`linkageGroupCount`
      self$`mapDbId` <- GenomeMapObject$`mapDbId`
      self$`mapName` <- GenomeMapObject$`mapName`
      self$`mapPUI` <- GenomeMapObject$`mapPUI`
      self$`markerCount` <- GenomeMapObject$`markerCount`
      self$`publishedDate` <- GenomeMapObject$`publishedDate`
      self$`scientificName` <- GenomeMapObject$`scientificName`
      self$`type` <- GenomeMapObject$`type`
      self$`unit` <- GenomeMapObject$`unit`
      self
    }
  )
)

