# BrAPI-Core
#
# The Breeding API (BrAPI) is a Standardized REST ful Web Service API Specification for communicating Plant Breeding Data. BrAPI allows for easy data sharing between databases and tools involved in plant breeding. <div class=\"brapi-section\"> <h2 class=\"brapi-section-title\">General Reference Documentation</h2> <div class=\"gen-info-link\"><a href=\"https://github.com/plantbreeding/API/blob/master/Specification/GeneralInfo/URL_Structure.md\">URL Structure</a></div> <div class=\"gen-info-link\"><a href=\"https://github.com/plantbreeding/API/blob/master/Specification/GeneralInfo/Response_Structure.md\">Response Structure</a></div> <div class=\"gen-info-link\"><a href=\"https://github.com/plantbreeding/API/blob/master/Specification/GeneralInfo/Date_Time_Encoding.md\">Date/Time Encoding</a></div> <div class=\"gen-info-link\"><a href=\"https://github.com/plantbreeding/API/blob/master/Specification/GeneralInfo/Location_Encoding.md\">Location Encoding</a></div> <div class=\"gen-info-link\"><a href=\"https://github.com/plantbreeding/API/blob/master/Specification/GeneralInfo/Error_Handling.md\">Error Handling</a></div> <div class=\"gen-info-link\"><a href=\"https://github.com/plantbreeding/API/blob/master/Specification/GeneralInfo/Search_Services.md\">Search Services</a></div> </div>  <div class=\"current-brapi-section brapi-section\"> <h2 class=\"brapi-section-title\">BrAPI Core</h2> <div class=\"brapi-section-description\">The BrAPI Core module contains high level entities used for organization and management. This includes Programs, Trials, Studies, Locations, People, and Lists</div> <div class=\"version-number\">V2.0</div> <div class=\"link-btn\"><a href=\"https://github.com/plantbreeding/API/tree/master/Specification/BrAPI-Core\">GitHub</a></div> <div class=\"link-btn\"><a href=\"https://app.swaggerhub.com/apis/PlantBreedingAPI/BrAPI-Core\">SwaggerHub</a></div> <div class=\"link-btn\"><a href=\"https://brapicore.docs.apiary.io\">Apiary</a></div> <div class=\"stop-float\"></div> </div>  <div class=\"brapi-section\"> <h2 class=\"brapi-section-title\">BrAPI Phenotyping</h2> <div class=\"brapi-section-description\">The BrAPI Phenotyping module contains entities related to phenotypic observations. This includes Observation Units, Observations, Observation Variables, Traits, Scales, Methods, and Images</div> <div class=\"version-number\">V2.0</div> <div class=\"link-btn\"><a href=\"https://github.com/plantbreeding/API/tree/master/Specification/BrAPI-Phenotyping\">GitHub</a></div> <div class=\"link-btn\"><a href=\"https://app.swaggerhub.com/apis/PlantBreedingAPI/BrAPI-Phenotyping\">SwaggerHub</a></div> <div class=\"link-btn\"><a href=\"https://brapiphenotyping.docs.apiary.io\">Apiary</a></div> <div class=\"stop-float\"></div> </div>  <div class=\"brapi-section\"> <h2 class=\"brapi-section-title\">BrAPI Genotyping</h2> <div class=\"brapi-section-description\">The BrAPI Genotyping module contains entities related to genotyping analysis. This includes Samples, Markers, Variant Sets, Variants, Call Sets, Calls, References, Reads, and Vendor Orders</div> <div class=\"version-number\">V2.0</div> <div class=\"link-btn\"><a href=\"https://github.com/plantbreeding/API/tree/master/Specification/BrAPI-Genotyping\">GitHub</a></div> <div class=\"link-btn\"><a href=\"https://app.swaggerhub.com/apis/PlantBreedingAPI/BrAPI-Genotyping\">SwaggerHub</a></div> <div class=\"link-btn\"><a href=\"https://brapigenotyping.docs.apiary.io\">Apiary</a></div> <div class=\"stop-float\"></div> </div>  <div class=\"brapi-section\"> <h2 class=\"brapi-section-title\">BrAPI Germplasm</h2> <div class=\"brapi-section-description\">The BrAPI Germplasm module contains entities related to germplasm management. This includes Germplasm, Germplasm Attributes, Seed Lots, Crosses, Pedigree, and Progeny</div> <div class=\"version-number\">V2.0</div> <div class=\"link-btn\"><a href=\"https://github.com/plantbreeding/API/tree/master/Specification/BrAPI-Germplasm\">GitHub</a></div> <div class=\"link-btn\"><a href=\"https://app.swaggerhub.com/apis/PlantBreedingAPI/BrAPI-Germplasm\">SwaggerHub</a></div> <div class=\"link-btn\"><a href=\"https://brapigermplasm.docs.apiary.io\">Apiary</a></div> <div class=\"stop-float\"></div> </div>  <style> .link-btn{ float: left;  margin: 2px 10px 0 0;  padding: 0 5px;  border-radius: 5px;  background-color: #ddd; } .stop-float{   clear: both; } .version-number{   float: left;    margin: 5px 10px 0 5px; } .brapi-section-title{   margin: 0 10px 0 0;   font-size: 20px; } .current-brapi-section{   font-weight: bolder;   border-radius: 5px;    background-color: #ddd; } .brapi-section{   padding: 5px 5px;  } .brapi-section-description{   margin: 5px 0 0 5px; } </style>
#
# The version of the OpenAPI document: 2.0
# 
# Generated by: https://openapi-generator.tech

#' @docType class
#' @title Germplasm operations
#' @description openapi.Germplasm
#' @format An \code{R6Class} generator object
#' @field apiClient Handles the client-server communication.
#'
#' @section Methods:
#' \describe{
#' \strong{ BreedingmethodsBreedingMethodDbIdGet } \emph{ Get the details of a specific Breeding Method }
#' Get the details of a specific Breeding Method used to produce Germplasm
#'
#' \itemize{
#' \item \emph{ @param } breeding.method.db.id character
#' \item \emph{ @param } authorization character
#' \item \emph{ @returnType } \link{BreedingMethodSingleResponse} \cr
#'
#'
#' \item status code : 200 | OK
#'
#' \item return type : BreedingMethodSingleResponse 
#' \item response headers :
#'
#' \tabular{ll}{
#' }
#' \item status code : 400 | Bad Request
#'
#' \item return type : character 
#' \item response headers :
#'
#' \tabular{ll}{
#' }
#' \item status code : 401 | Unauthorized
#'
#' \item return type : character 
#' \item response headers :
#'
#' \tabular{ll}{
#' }
#' \item status code : 403 | Forbidden
#'
#' \item return type : character 
#' \item response headers :
#'
#' \tabular{ll}{
#' }
#' \item status code : 404 | Not Found
#'
#' \item return type : character 
#' \item response headers :
#'
#' \tabular{ll}{
#' }
#' }
#'
#' \strong{ BreedingmethodsGet } \emph{ Get the Breeding Methods }
#' Get the list of germplasm breeding methods available in a system.
#'
#' \itemize{
#' \item \emph{ @param } page integer
#' \item \emph{ @param } page.size integer
#' \item \emph{ @param } authorization character
#' \item \emph{ @returnType } \link{BreedingMethodListResponse} \cr
#'
#'
#' \item status code : 200 | OK
#'
#' \item return type : BreedingMethodListResponse 
#' \item response headers :
#'
#' \tabular{ll}{
#' }
#' \item status code : 400 | Bad Request
#'
#' \item return type : character 
#' \item response headers :
#'
#' \tabular{ll}{
#' }
#' \item status code : 401 | Unauthorized
#'
#' \item return type : character 
#' \item response headers :
#'
#' \tabular{ll}{
#' }
#' \item status code : 403 | Forbidden
#'
#' \item return type : character 
#' \item response headers :
#'
#' \tabular{ll}{
#' }
#' }
#'
#' \strong{ GermplasmGermplasmDbIdGet } \emph{ Get the details of a specific Germplasm }
#' Germplasm Details by germplasmDbId was merged with Germplasm Multi Crop Passport Data. The MCPD fields are optional and marked with the prefix [MCPD].
#'
#' \itemize{
#' \item \emph{ @param } germplasm.db.id character
#' \item \emph{ @param } authorization character
#' \item \emph{ @returnType } \link{GermplasmSingleResponse} \cr
#'
#'
#' \item status code : 200 | OK
#'
#' \item return type : GermplasmSingleResponse 
#' \item response headers :
#'
#' \tabular{ll}{
#' }
#' \item status code : 400 | Bad Request
#'
#' \item return type : character 
#' \item response headers :
#'
#' \tabular{ll}{
#' }
#' \item status code : 401 | Unauthorized
#'
#' \item return type : character 
#' \item response headers :
#'
#' \tabular{ll}{
#' }
#' \item status code : 403 | Forbidden
#'
#' \item return type : character 
#' \item response headers :
#'
#' \tabular{ll}{
#' }
#' \item status code : 404 | Not Found
#'
#' \item return type : character 
#' \item response headers :
#'
#' \tabular{ll}{
#' }
#' }
#'
#' \strong{ GermplasmGermplasmDbIdMcpdGet } \emph{ Get the details of a specific Germplasm in MCPD format }
#' Get all MCPD details of a germplasm  &lt;a target&#x3D;\&quot;_blank\&quot; href&#x3D;\&quot;https://www.bioversityInternational.org/fileadmin/user_upload/online_library/publications/pdfs/FAOBIOVERSITY_MULTI-CROP_PASSPORT_DESCRIPTORS_V.2.1_2015_2020.pdf\&quot;&gt; MCPD v2.1 spec can be found here &lt;/a&gt;  Implementation Notes  - When the MCPD spec identifies a field which can have multiple values returned, the JSON response should be an array instead of a semi-colon separated string.
#'
#' \itemize{
#' \item \emph{ @param } germplasm.db.id character
#' \item \emph{ @param } authorization character
#' \item \emph{ @returnType } \link{GermplasmMCPDResponse} \cr
#'
#'
#' \item status code : 200 | OK
#'
#' \item return type : GermplasmMCPDResponse 
#' \item response headers :
#'
#' \tabular{ll}{
#' }
#' \item status code : 400 | Bad Request
#'
#' \item return type : character 
#' \item response headers :
#'
#' \tabular{ll}{
#' }
#' \item status code : 401 | Unauthorized
#'
#' \item return type : character 
#' \item response headers :
#'
#' \tabular{ll}{
#' }
#' \item status code : 403 | Forbidden
#'
#' \item return type : character 
#' \item response headers :
#'
#' \tabular{ll}{
#' }
#' \item status code : 404 | Not Found
#'
#' \item return type : character 
#' \item response headers :
#'
#' \tabular{ll}{
#' }
#' }
#'
#' \strong{ GermplasmGermplasmDbIdPedigreeGet } \emph{ Get the pedigree details of a specific Germplasm }
#' Get the parentage information of a specific Germplasm
#'
#' \itemize{
#' \item \emph{ @param } germplasm.db.id character
#' \item \emph{ @param } notation character
#' \item \emph{ @param } include.siblings character
#' \item \emph{ @param } authorization character
#' \item \emph{ @returnType } \link{GermplasmPedigreeResponse} \cr
#'
#'
#' \item status code : 200 | OK
#'
#' \item return type : GermplasmPedigreeResponse 
#' \item response headers :
#'
#' \tabular{ll}{
#' }
#' \item status code : 400 | Bad Request
#'
#' \item return type : character 
#' \item response headers :
#'
#' \tabular{ll}{
#' }
#' \item status code : 401 | Unauthorized
#'
#' \item return type : character 
#' \item response headers :
#'
#' \tabular{ll}{
#' }
#' \item status code : 403 | Forbidden
#'
#' \item return type : character 
#' \item response headers :
#'
#' \tabular{ll}{
#' }
#' \item status code : 404 | Not Found
#'
#' \item return type : character 
#' \item response headers :
#'
#' \tabular{ll}{
#' }
#' }
#'
#' \strong{ GermplasmGermplasmDbIdProgenyGet } \emph{ Get the progeny details of a specific Germplasm }
#' Get the germplasmDbIds for all the Progeny of a particular germplasm.  Implementation Notes  - Regarding the &#39;&#39;parentType&#39;&#39; field in the progeny object. Given a germplasm A having a progeny B and C, &#39;&#39;parentType&#39;&#39; for progeny B refers to the &#39;&#39;parentType&#39;&#39; of A toward B.
#'
#' \itemize{
#' \item \emph{ @param } germplasm.db.id character
#' \item \emph{ @param } authorization character
#' \item \emph{ @returnType } \link{GermplasmProgenyResponse} \cr
#'
#'
#' \item status code : 200 | OK
#'
#' \item return type : GermplasmProgenyResponse 
#' \item response headers :
#'
#' \tabular{ll}{
#' }
#' \item status code : 400 | Bad Request
#'
#' \item return type : character 
#' \item response headers :
#'
#' \tabular{ll}{
#' }
#' \item status code : 401 | Unauthorized
#'
#' \item return type : character 
#' \item response headers :
#'
#' \tabular{ll}{
#' }
#' \item status code : 403 | Forbidden
#'
#' \item return type : character 
#' \item response headers :
#'
#' \tabular{ll}{
#' }
#' \item status code : 404 | Not Found
#'
#' \item return type : character 
#' \item response headers :
#'
#' \tabular{ll}{
#' }
#' }
#'
#' \strong{ GermplasmGermplasmDbIdPut } \emph{ Update the details of an existing Germplasm }
#' Germplasm Details by germplasmDbId was merged with Germplasm Multi Crop Passport Data. The MCPD fields are optional and marked with the prefix [MCPD].
#'
#' \itemize{
#' \item \emph{ @param } germplasm.db.id character
#' \item \emph{ @param } authorization character
#' \item \emph{ @param } germplasm.new.request \link{GermplasmNewRequest}
#' \item \emph{ @returnType } \link{GermplasmSingleResponse} \cr
#'
#'
#' \item status code : 200 | OK
#'
#' \item return type : GermplasmSingleResponse 
#' \item response headers :
#'
#' \tabular{ll}{
#' }
#' \item status code : 400 | Bad Request
#'
#' \item return type : character 
#' \item response headers :
#'
#' \tabular{ll}{
#' }
#' \item status code : 401 | Unauthorized
#'
#' \item return type : character 
#' \item response headers :
#'
#' \tabular{ll}{
#' }
#' \item status code : 403 | Forbidden
#'
#' \item return type : character 
#' \item response headers :
#'
#' \tabular{ll}{
#' }
#' \item status code : 404 | Not Found
#'
#' \item return type : character 
#' \item response headers :
#'
#' \tabular{ll}{
#' }
#' }
#'
#' \strong{ GermplasmGet } \emph{ Get a filtered list of Germplasm }
#' Addresses these needs  - General germplasm search mechanism that accepts POST for complex queries   - Possibility to search germplasm by more parameters than those allowed by the existing germplasm search   - Possibility to get MCPD details by PUID rather than dbId
#'
#' \itemize{
#' \item \emph{ @param } germplasm.pui character
#' \item \emph{ @param } germplasm.db.id character
#' \item \emph{ @param } germplasm.name character
#' \item \emph{ @param } common.crop.name character
#' \item \emph{ @param } accession.number character
#' \item \emph{ @param } collection character
#' \item \emph{ @param } genus character
#' \item \emph{ @param } species character
#' \item \emph{ @param } study.db.id character
#' \item \emph{ @param } synonym character
#' \item \emph{ @param } parent.db.id character
#' \item \emph{ @param } progeny.db.id character
#' \item \emph{ @param } external.reference.id character
#' \item \emph{ @param } external.reference.source character
#' \item \emph{ @param } page integer
#' \item \emph{ @param } page.size integer
#' \item \emph{ @param } authorization character
#' \item \emph{ @returnType } \link{GermplasmListResponse} \cr
#'
#'
#' \item status code : 200 | OK
#'
#' \item return type : GermplasmListResponse 
#' \item response headers :
#'
#' \tabular{ll}{
#' }
#' \item status code : 400 | Bad Request
#'
#' \item return type : character 
#' \item response headers :
#'
#' \tabular{ll}{
#' }
#' \item status code : 401 | Unauthorized
#'
#' \item return type : character 
#' \item response headers :
#'
#' \tabular{ll}{
#' }
#' \item status code : 403 | Forbidden
#'
#' \item return type : character 
#' \item response headers :
#'
#' \tabular{ll}{
#' }
#' }
#'
#' \strong{ GermplasmPost } \emph{ Create new Germplasm entities on this server }
#' Create new Germplasm entities on this server
#'
#' \itemize{
#' \item \emph{ @param } authorization character
#' \item \emph{ @param } germplasm.new.request list( \link{GermplasmNewRequest} )
#' \item \emph{ @returnType } \link{GermplasmListResponse} \cr
#'
#'
#' \item status code : 200 | OK
#'
#' \item return type : GermplasmListResponse 
#' \item response headers :
#'
#' \tabular{ll}{
#' }
#' \item status code : 400 | Bad Request
#'
#' \item return type : character 
#' \item response headers :
#'
#' \tabular{ll}{
#' }
#' \item status code : 401 | Unauthorized
#'
#' \item return type : character 
#' \item response headers :
#'
#' \tabular{ll}{
#' }
#' \item status code : 403 | Forbidden
#'
#' \item return type : character 
#' \item response headers :
#'
#' \tabular{ll}{
#' }
#' }
#'
#' \strong{ SearchGermplasmPost } \emph{ Submit a search request for Germplasm }
#' Search for a set of germplasm based on some criteria  Addresses these needs   - General germplasm search mechanism that accepts POST for complex queries   - Possibility to search germplasm by more parameters than those allowed by the existing germplasm search   - Possibility to get MCPD details by PUID rather than dbId  See Search Services for additional implementation details.
#'
#' \itemize{
#' \item \emph{ @param } authorization character
#' \item \emph{ @param } germplasm.search.request \link{GermplasmSearchRequest}
#' \item \emph{ @returnType } \link{GermplasmListResponse} \cr
#'
#'
#' \item status code : 200 | OK
#'
#' \item return type : GermplasmListResponse 
#' \item response headers :
#'
#' \tabular{ll}{
#' }
#' \item status code : 202 | Accepted
#'
#' \item return type : Accepted202SearchResponse
#' \item response headers :
#'
#' \tabular{ll}{
#' }
#' \item status code : 400 | Bad Request
#'
#' \item return type : character 
#' \item response headers :
#'
#' \tabular{ll}{
#' }
#' \item status code : 401 | Unauthorized
#'
#' \item return type : character 
#' \item response headers :
#'
#' \tabular{ll}{
#' }
#' \item status code : 403 | Forbidden
#'
#' \item return type : character 
#' \item response headers :
#'
#' \tabular{ll}{
#' }
#' }
#'
#' \strong{ SearchGermplasmSearchResultsDbIdGet } \emph{ Get the results of a Germplasm search request }
#' See Search Services for additional implementation details.  Addresses these needs:   1. General germplasm search mechanism that accepts POST for complex queries   2. possibility to search germplasm by more parameters than those allowed by the existing germplasm search   3. possibility to get MCPD details by PUID rather than dbId
#'
#' \itemize{
#' \item \emph{ @param } search.results.db.id character
#' \item \emph{ @param } page integer
#' \item \emph{ @param } page.size integer
#' \item \emph{ @param } authorization character
#' \item \emph{ @returnType } \link{GermplasmListResponse} \cr
#'
#'
#' \item status code : 200 | OK
#'
#' \item return type : GermplasmListResponse 
#' \item response headers :
#'
#' \tabular{ll}{
#' }
#' \item status code : 202 | Accepted
#'
#' \item return type : Accepted202SearchResponse
#' \item response headers :
#'
#' \tabular{ll}{
#' }
#' \item status code : 400 | Bad Request
#'
#' \item return type : character 
#' \item response headers :
#'
#' \tabular{ll}{
#' }
#' \item status code : 401 | Unauthorized
#'
#' \item return type : character 
#' \item response headers :
#'
#' \tabular{ll}{
#' }
#' \item status code : 403 | Forbidden
#'
#' \item return type : character 
#' \item response headers :
#'
#' \tabular{ll}{
#' }
#' \item status code : 404 | Not Found
#'
#' \item return type : character 
#' \item response headers :
#'
#' \tabular{ll}{
#' }
#' }
#'
#' }
#'
#'
#' @examples
#' \dontrun{
#' ####################  BreedingmethodsBreedingMethodDbIdGet  ####################
#'
#' library(openapi)
#' var.breeding.method.db.id <- 'breeding.method.db.id_example' # character | Internal database identifier for a breeding method
#' var.authorization <- 'Bearer XXXX' # character | HTTP HEADER - Token used for Authorization   <strong> Bearer {token_string} </strong>
#'
#' #Get the details of a specific Breeding Method
#' api.instance <- GermplasmApi$new()
#'
#' #Configure HTTP basic authorization: AuthorizationToken
#' # provide your username in the user-serial format
#' api.instance$apiClient$username <- '<user-serial>'; 
#' # provide your api key generated using the developer portal
#' api.instance$apiClient$password <- '<api_key>';
#'
#' result <- api.instance$BreedingmethodsBreedingMethodDbIdGet(var.breeding.method.db.id, authorization=var.authorization)
#'
#'
#' ####################  BreedingmethodsGet  ####################
#'
#' library(openapi)
#' var.page <- 0 # integer | Used to request a specific page of data to be returned.  The page indexing starts at 0 (the first page is 'page'= 0). Default is `0`.
#' var.page.size <- 1000 # integer | The size of the pages to be returned. Default is `1000`.
#' var.authorization <- 'Bearer XXXX' # character | HTTP HEADER - Token used for Authorization   <strong> Bearer {token_string} </strong>
#'
#' #Get the Breeding Methods
#' api.instance <- GermplasmApi$new()
#'
#' #Configure HTTP basic authorization: AuthorizationToken
#' # provide your username in the user-serial format
#' api.instance$apiClient$username <- '<user-serial>'; 
#' # provide your api key generated using the developer portal
#' api.instance$apiClient$password <- '<api_key>';
#'
#' result <- api.instance$BreedingmethodsGet(page=var.page, page.size=var.page.size, authorization=var.authorization)
#'
#'
#' ####################  GermplasmGermplasmDbIdGet  ####################
#'
#' library(openapi)
#' var.germplasm.db.id <- 'germplasm.db.id_example' # character | The internal id of the germplasm
#' var.authorization <- 'Bearer XXXX' # character | HTTP HEADER - Token used for Authorization   <strong> Bearer {token_string} </strong>
#'
#' #Get the details of a specific Germplasm
#' api.instance <- GermplasmApi$new()
#'
#' #Configure HTTP basic authorization: AuthorizationToken
#' # provide your username in the user-serial format
#' api.instance$apiClient$username <- '<user-serial>'; 
#' # provide your api key generated using the developer portal
#' api.instance$apiClient$password <- '<api_key>';
#'
#' result <- api.instance$GermplasmGermplasmDbIdGet(var.germplasm.db.id, authorization=var.authorization)
#'
#'
#' ####################  GermplasmGermplasmDbIdMcpdGet  ####################
#'
#' library(openapi)
#' var.germplasm.db.id <- 'germplasm.db.id_example' # character | the internal id of the germplasm
#' var.authorization <- 'Bearer XXXX' # character | HTTP HEADER - Token used for Authorization   <strong> Bearer {token_string} </strong>
#'
#' #Get the details of a specific Germplasm in MCPD format
#' api.instance <- GermplasmApi$new()
#'
#' #Configure HTTP basic authorization: AuthorizationToken
#' # provide your username in the user-serial format
#' api.instance$apiClient$username <- '<user-serial>'; 
#' # provide your api key generated using the developer portal
#' api.instance$apiClient$password <- '<api_key>';
#'
#' result <- api.instance$GermplasmGermplasmDbIdMcpdGet(var.germplasm.db.id, authorization=var.authorization)
#'
#'
#' ####################  GermplasmGermplasmDbIdPedigreeGet  ####################
#'
#' library(openapi)
#' var.germplasm.db.id <- 'germplasm.db.id_example' # character | the internal id of the germplasm
#' var.notation <- 'notation_example' # character | text representation of the pedigree
#' var.include.siblings <- 'include.siblings_example' # character | include array of siblings in response
#' var.authorization <- 'Bearer XXXX' # character | HTTP HEADER - Token used for Authorization   <strong> Bearer {token_string} </strong>
#'
#' #Get the pedigree details of a specific Germplasm
#' api.instance <- GermplasmApi$new()
#'
#' #Configure HTTP basic authorization: AuthorizationToken
#' # provide your username in the user-serial format
#' api.instance$apiClient$username <- '<user-serial>'; 
#' # provide your api key generated using the developer portal
#' api.instance$apiClient$password <- '<api_key>';
#'
#' result <- api.instance$GermplasmGermplasmDbIdPedigreeGet(var.germplasm.db.id, notation=var.notation, include.siblings=var.include.siblings, authorization=var.authorization)
#'
#'
#' ####################  GermplasmGermplasmDbIdProgenyGet  ####################
#'
#' library(openapi)
#' var.germplasm.db.id <- 'germplasm.db.id_example' # character | the internal id of the germplasm
#' var.authorization <- 'Bearer XXXX' # character | HTTP HEADER - Token used for Authorization   <strong> Bearer {token_string} </strong>
#'
#' #Get the progeny details of a specific Germplasm
#' api.instance <- GermplasmApi$new()
#'
#' #Configure HTTP basic authorization: AuthorizationToken
#' # provide your username in the user-serial format
#' api.instance$apiClient$username <- '<user-serial>'; 
#' # provide your api key generated using the developer portal
#' api.instance$apiClient$password <- '<api_key>';
#'
#' result <- api.instance$GermplasmGermplasmDbIdProgenyGet(var.germplasm.db.id, authorization=var.authorization)
#'
#'
#' ####################  GermplasmGermplasmDbIdPut  ####################
#'
#' library(openapi)
#' var.germplasm.db.id <- 'germplasm.db.id_example' # character | The internal id of the germplasm
#' var.authorization <- 'Bearer XXXX' # character | HTTP HEADER - Token used for Authorization   <strong> Bearer {token_string} </strong>
#' var.germplasm.new.request <- GermplasmNewRequest$new() # GermplasmNewRequest | 
#'
#' #Update the details of an existing Germplasm
#' api.instance <- GermplasmApi$new()
#'
#' #Configure HTTP basic authorization: AuthorizationToken
#' # provide your username in the user-serial format
#' api.instance$apiClient$username <- '<user-serial>'; 
#' # provide your api key generated using the developer portal
#' api.instance$apiClient$password <- '<api_key>';
#'
#' result <- api.instance$GermplasmGermplasmDbIdPut(var.germplasm.db.id, authorization=var.authorization, germplasm.new.request=var.germplasm.new.request)
#'
#'
#' ####################  GermplasmGet  ####################
#'
#' library(openapi)
#' var.germplasm.pui <- 'germplasm.pui_example' # character | Permanent unique identifier (DOI, URI, etc.)
#' var.germplasm.db.id <- 'germplasm.db.id_example' # character | Internal database identifier
#' var.germplasm.name <- 'germplasm.name_example' # character | Name of the germplasm
#' var.common.crop.name <- 'common.crop.name_example' # character | The common crop name related to this germplasm
#' var.accession.number <- 'accession.number_example' # character | Unique identifiers for accessions within a genebank
#' var.collection <- 'collection_example' # character | A specific panel/collection/population name this germplasm belongs to.
#' var.genus <- 'genus_example' # character | Genus name to identify germplasm
#' var.species <- 'species_example' # character | Species name to identify germplasm
#' var.study.db.id <- 'study.db.id_example' # character | Search for Germplasm that are associated with a particular Study
#' var.synonym <- 'synonym_example' # character | Alternative name or ID used to reference this germplasm
#' var.parent.db.id <- 'parent.db.id_example' # character | Search for Germplasm with this parent
#' var.progeny.db.id <- 'progeny.db.id_example' # character | Search for Germplasm with this child
#' var.external.reference.id <- 'external.reference.id_example' # character | An external reference ID. Could be a simple string or a URI. (use with `externalReferenceSource` parameter)
#' var.external.reference.source <- 'external.reference.source_example' # character | An identifier for the source system or database of an external reference (use with `externalReferenceID` parameter)
#' var.page <- 0 # integer | Used to request a specific page of data to be returned.  The page indexing starts at 0 (the first page is 'page'= 0). Default is `0`.
#' var.page.size <- 1000 # integer | The size of the pages to be returned. Default is `1000`.
#' var.authorization <- 'Bearer XXXX' # character | HTTP HEADER - Token used for Authorization   <strong> Bearer {token_string} </strong>
#'
#' #Get a filtered list of Germplasm
#' api.instance <- GermplasmApi$new()
#'
#' #Configure HTTP basic authorization: AuthorizationToken
#' # provide your username in the user-serial format
#' api.instance$apiClient$username <- '<user-serial>'; 
#' # provide your api key generated using the developer portal
#' api.instance$apiClient$password <- '<api_key>';
#'
#' result <- api.instance$GermplasmGet(germplasm.pui=var.germplasm.pui, germplasm.db.id=var.germplasm.db.id, germplasm.name=var.germplasm.name, common.crop.name=var.common.crop.name, accession.number=var.accession.number, collection=var.collection, genus=var.genus, species=var.species, study.db.id=var.study.db.id, synonym=var.synonym, parent.db.id=var.parent.db.id, progeny.db.id=var.progeny.db.id, external.reference.id=var.external.reference.id, external.reference.source=var.external.reference.source, page=var.page, page.size=var.page.size, authorization=var.authorization)
#'
#'
#' ####################  GermplasmPost  ####################
#'
#' library(openapi)
#' var.authorization <- 'Bearer XXXX' # character | HTTP HEADER - Token used for Authorization   <strong> Bearer {token_string} </strong>
#' var.germplasm.new.request <- [GermplasmNewRequest$new()] # array[GermplasmNewRequest] | 
#'
#' #Create new Germplasm entities on this server
#' api.instance <- GermplasmApi$new()
#'
#' #Configure HTTP basic authorization: AuthorizationToken
#' # provide your username in the user-serial format
#' api.instance$apiClient$username <- '<user-serial>'; 
#' # provide your api key generated using the developer portal
#' api.instance$apiClient$password <- '<api_key>';
#'
#' result <- api.instance$GermplasmPost(authorization=var.authorization, germplasm.new.request=var.germplasm.new.request)
#'
#'
#' ####################  SearchGermplasmPost  ####################
#'
#' library(openapi)
#' var.authorization <- 'Bearer XXXX' # character | HTTP HEADER - Token used for Authorization   <strong> Bearer {token_string} </strong>
#' var.germplasm.search.request <- GermplasmSearchRequest$new() # GermplasmSearchRequest | 
#'
#' #Submit a search request for Germplasm
#' api.instance <- GermplasmApi$new()
#'
#' #Configure HTTP basic authorization: AuthorizationToken
#' # provide your username in the user-serial format
#' api.instance$apiClient$username <- '<user-serial>'; 
#' # provide your api key generated using the developer portal
#' api.instance$apiClient$password <- '<api_key>';
#'
#' result <- api.instance$SearchGermplasmPost(authorization=var.authorization, germplasm.search.request=var.germplasm.search.request)
#'
#'
#' ####################  SearchGermplasmSearchResultsDbIdGet  ####################
#'
#' library(openapi)
#' var.search.results.db.id <- 'search.results.db.id_example' # character | Unique identifier which references the search results
#' var.page <- 0 # integer | Used to request a specific page of data to be returned.  The page indexing starts at 0 (the first page is 'page'= 0). Default is `0`.
#' var.page.size <- 1000 # integer | The size of the pages to be returned. Default is `1000`.
#' var.authorization <- 'Bearer XXXX' # character | HTTP HEADER - Token used for Authorization   <strong> Bearer {token_string} </strong>
#'
#' #Get the results of a Germplasm search request
#' api.instance <- GermplasmApi$new()
#'
#' #Configure HTTP basic authorization: AuthorizationToken
#' # provide your username in the user-serial format
#' api.instance$apiClient$username <- '<user-serial>'; 
#' # provide your api key generated using the developer portal
#' api.instance$apiClient$password <- '<api_key>';
#'
#' result <- api.instance$SearchGermplasmSearchResultsDbIdGet(var.search.results.db.id, page=var.page, page.size=var.page.size, authorization=var.authorization)
#'
#'
#' }
#' @importFrom R6 R6Class
#' @importFrom base64enc base64encode
#' @export
GermplasmApi <- R6::R6Class(
  'GermplasmApi',
  public = list(
    apiClient = NULL,
    initialize = function(apiClient){
      if (!missing(apiClient)) {
        self$apiClient <- apiClient
      }
      else {
        self$apiClient <- ApiClient$new()
      }
    },
    BreedingmethodsBreedingMethodDbIdGet = function(breeding.method.db.id, authorization=NULL, ...){
      apiResponse <- self$BreedingmethodsBreedingMethodDbIdGetWithHttpInfo(breeding.method.db.id, authorization, ...)
      resp <- apiResponse$response
      if (httr::status_code(resp) >= 200 && httr::status_code(resp) <= 299) {
        apiResponse$content
      } else if (httr::status_code(resp) >= 300 && httr::status_code(resp) <= 399) {
        apiResponse
      } else if (httr::status_code(resp) >= 400 && httr::status_code(resp) <= 499) {
        apiResponse
      } else if (httr::status_code(resp) >= 500 && httr::status_code(resp) <= 599) {
        apiResponse
      }
    },

    BreedingmethodsBreedingMethodDbIdGetWithHttpInfo = function(breeding.method.db.id, authorization=NULL, ...){
      args <- list(...)
      queryParams <- list()
      headerParams <- c()

      if (missing(`breeding.method.db.id`)) {
        stop("Missing required parameter `breeding.method.db.id`.")
      }

      headerParams['Authorization'] <- `authorization`

      body <- NULL
      urlPath <- "/breedingmethods/{breedingMethodDbId}"
      if (!missing(`breeding.method.db.id`)) {
        urlPath <- gsub(paste0("\\{", "breedingMethodDbId", "\\}"), URLencode(as.character(`breeding.method.db.id`), reserved = TRUE), urlPath)
      }


      resp <- self$apiClient$CallApi(url = paste0(self$apiClient$basePath, urlPath),
                                 method = "GET",
                                 queryParams = queryParams,
                                 headerParams = headerParams,
                                 body = body,
                                 ...)

      if (httr::status_code(resp) >= 200 && httr::status_code(resp) <= 299) {
        deserializedRespObj <- tryCatch(
          self$apiClient$deserialize(resp, "BreedingMethodSingleResponse", loadNamespace("openapi")),
          error = function(e){
             stop("Failed to deserialize response")
          }
        )
        ApiResponse$new(deserializedRespObj, resp)
      } else if (httr::status_code(resp) >= 300 && httr::status_code(resp) <= 399) {
        ApiResponse$new(paste("Server returned " , httr::status_code(resp) , " response status code."), resp)
      } else if (httr::status_code(resp) >= 400 && httr::status_code(resp) <= 499) {
        ApiResponse$new("API client error", resp)
      } else if (httr::status_code(resp) >= 500 && httr::status_code(resp) <= 599) {
        ApiResponse$new("API server error", resp)
      }
    },
    BreedingmethodsGet = function(page=NULL, page.size=NULL, authorization=NULL, ...){
      apiResponse <- self$BreedingmethodsGetWithHttpInfo(page, page.size, authorization, ...)
      resp <- apiResponse$response
      if (httr::status_code(resp) >= 200 && httr::status_code(resp) <= 299) {
        apiResponse$content
      } else if (httr::status_code(resp) >= 300 && httr::status_code(resp) <= 399) {
        apiResponse
      } else if (httr::status_code(resp) >= 400 && httr::status_code(resp) <= 499) {
        apiResponse
      } else if (httr::status_code(resp) >= 500 && httr::status_code(resp) <= 599) {
        apiResponse
      }
    },

    BreedingmethodsGetWithHttpInfo = function(page=NULL, page.size=NULL, authorization=NULL, ...){
      args <- list(...)
      queryParams <- list()
      headerParams <- c()

      headerParams['Authorization'] <- `authorization`

      queryParams['page'] <- page

      queryParams['pageSize'] <- page.size

      body <- NULL
      urlPath <- "/breedingmethods"

      resp <- self$apiClient$CallApi(url = paste0(self$apiClient$basePath, urlPath),
                                 method = "GET",
                                 queryParams = queryParams,
                                 headerParams = headerParams,
                                 body = body,
                                 ...)

      if (httr::status_code(resp) >= 200 && httr::status_code(resp) <= 299) {
        deserializedRespObj <- tryCatch(
          self$apiClient$deserialize(resp, "BreedingMethodListResponse", loadNamespace("openapi")),
          error = function(e){
             stop("Failed to deserialize response")
          }
        )
        ApiResponse$new(deserializedRespObj, resp)
      } else if (httr::status_code(resp) >= 300 && httr::status_code(resp) <= 399) {
        ApiResponse$new(paste("Server returned " , httr::status_code(resp) , " response status code."), resp)
      } else if (httr::status_code(resp) >= 400 && httr::status_code(resp) <= 499) {
        ApiResponse$new("API client error", resp)
      } else if (httr::status_code(resp) >= 500 && httr::status_code(resp) <= 599) {
        ApiResponse$new("API server error", resp)
      }
    },
    GermplasmGermplasmDbIdGet = function(germplasm.db.id, authorization=NULL, ...){
      apiResponse <- self$GermplasmGermplasmDbIdGetWithHttpInfo(germplasm.db.id, authorization, ...)
      resp <- apiResponse$response
      if (httr::status_code(resp) >= 200 && httr::status_code(resp) <= 299) {
        apiResponse$content
      } else if (httr::status_code(resp) >= 300 && httr::status_code(resp) <= 399) {
        apiResponse
      } else if (httr::status_code(resp) >= 400 && httr::status_code(resp) <= 499) {
        apiResponse
      } else if (httr::status_code(resp) >= 500 && httr::status_code(resp) <= 599) {
        apiResponse
      }
    },

    GermplasmGermplasmDbIdGetWithHttpInfo = function(germplasm.db.id, authorization=NULL, ...){
      args <- list(...)
      queryParams <- list()
      headerParams <- c()

      if (missing(`germplasm.db.id`)) {
        stop("Missing required parameter `germplasm.db.id`.")
      }

      headerParams['Authorization'] <- `authorization`

      body <- NULL
      urlPath <- "/germplasm/{germplasmDbId}"
      if (!missing(`germplasm.db.id`)) {
        urlPath <- gsub(paste0("\\{", "germplasmDbId", "\\}"), URLencode(as.character(`germplasm.db.id`), reserved = TRUE), urlPath)
      }


      resp <- self$apiClient$CallApi(url = paste0(self$apiClient$basePath, urlPath),
                                 method = "GET",
                                 queryParams = queryParams,
                                 headerParams = headerParams,
                                 body = body,
                                 ...)

      if (httr::status_code(resp) >= 200 && httr::status_code(resp) <= 299) {
        deserializedRespObj <- tryCatch(
          self$apiClient$deserialize(resp, "GermplasmSingleResponse", loadNamespace("openapi")),
          error = function(e){
             stop("Failed to deserialize response")
          }
        )
        ApiResponse$new(deserializedRespObj, resp)
      } else if (httr::status_code(resp) >= 300 && httr::status_code(resp) <= 399) {
        ApiResponse$new(paste("Server returned " , httr::status_code(resp) , " response status code."), resp)
      } else if (httr::status_code(resp) >= 400 && httr::status_code(resp) <= 499) {
        ApiResponse$new("API client error", resp)
      } else if (httr::status_code(resp) >= 500 && httr::status_code(resp) <= 599) {
        ApiResponse$new("API server error", resp)
      }
    },
    GermplasmGermplasmDbIdMcpdGet = function(germplasm.db.id, authorization=NULL, ...){
      apiResponse <- self$GermplasmGermplasmDbIdMcpdGetWithHttpInfo(germplasm.db.id, authorization, ...)
      resp <- apiResponse$response
      if (httr::status_code(resp) >= 200 && httr::status_code(resp) <= 299) {
        apiResponse$content
      } else if (httr::status_code(resp) >= 300 && httr::status_code(resp) <= 399) {
        apiResponse
      } else if (httr::status_code(resp) >= 400 && httr::status_code(resp) <= 499) {
        apiResponse
      } else if (httr::status_code(resp) >= 500 && httr::status_code(resp) <= 599) {
        apiResponse
      }
    },

    GermplasmGermplasmDbIdMcpdGetWithHttpInfo = function(germplasm.db.id, authorization=NULL, ...){
      args <- list(...)
      queryParams <- list()
      headerParams <- c()

      if (missing(`germplasm.db.id`)) {
        stop("Missing required parameter `germplasm.db.id`.")
      }

      headerParams['Authorization'] <- `authorization`

      body <- NULL
      urlPath <- "/germplasm/{germplasmDbId}/mcpd"
      if (!missing(`germplasm.db.id`)) {
        urlPath <- gsub(paste0("\\{", "germplasmDbId", "\\}"), URLencode(as.character(`germplasm.db.id`), reserved = TRUE), urlPath)
      }


      resp <- self$apiClient$CallApi(url = paste0(self$apiClient$basePath, urlPath),
                                 method = "GET",
                                 queryParams = queryParams,
                                 headerParams = headerParams,
                                 body = body,
                                 ...)

      if (httr::status_code(resp) >= 200 && httr::status_code(resp) <= 299) {
        deserializedRespObj <- tryCatch(
          self$apiClient$deserialize(resp, "GermplasmMCPDResponse", loadNamespace("openapi")),
          error = function(e){
             stop("Failed to deserialize response")
          }
        )
        ApiResponse$new(deserializedRespObj, resp)
      } else if (httr::status_code(resp) >= 300 && httr::status_code(resp) <= 399) {
        ApiResponse$new(paste("Server returned " , httr::status_code(resp) , " response status code."), resp)
      } else if (httr::status_code(resp) >= 400 && httr::status_code(resp) <= 499) {
        ApiResponse$new("API client error", resp)
      } else if (httr::status_code(resp) >= 500 && httr::status_code(resp) <= 599) {
        ApiResponse$new("API server error", resp)
      }
    },
    GermplasmGermplasmDbIdPedigreeGet = function(germplasm.db.id, notation=NULL, include.siblings=NULL, authorization=NULL, ...){
      apiResponse <- self$GermplasmGermplasmDbIdPedigreeGetWithHttpInfo(germplasm.db.id, notation, include.siblings, authorization, ...)
      resp <- apiResponse$response
      if (httr::status_code(resp) >= 200 && httr::status_code(resp) <= 299) {
        apiResponse$content
      } else if (httr::status_code(resp) >= 300 && httr::status_code(resp) <= 399) {
        apiResponse
      } else if (httr::status_code(resp) >= 400 && httr::status_code(resp) <= 499) {
        apiResponse
      } else if (httr::status_code(resp) >= 500 && httr::status_code(resp) <= 599) {
        apiResponse
      }
    },

    GermplasmGermplasmDbIdPedigreeGetWithHttpInfo = function(germplasm.db.id, notation=NULL, include.siblings=NULL, authorization=NULL, ...){
      args <- list(...)
      queryParams <- list()
      headerParams <- c()

      if (missing(`germplasm.db.id`)) {
        stop("Missing required parameter `germplasm.db.id`.")
      }

      headerParams['Authorization'] <- `authorization`

      queryParams['notation'] <- notation

      queryParams['includeSiblings'] <- include.siblings

      body <- NULL
      urlPath <- "/germplasm/{germplasmDbId}/pedigree"
      if (!missing(`germplasm.db.id`)) {
        urlPath <- gsub(paste0("\\{", "germplasmDbId", "\\}"), URLencode(as.character(`germplasm.db.id`), reserved = TRUE), urlPath)
      }


      resp <- self$apiClient$CallApi(url = paste0(self$apiClient$basePath, urlPath),
                                 method = "GET",
                                 queryParams = queryParams,
                                 headerParams = headerParams,
                                 body = body,
                                 ...)

      if (httr::status_code(resp) >= 200 && httr::status_code(resp) <= 299) {
        deserializedRespObj <- tryCatch(
          self$apiClient$deserialize(resp, "GermplasmPedigreeResponse", loadNamespace("openapi")),
          error = function(e){
             stop("Failed to deserialize response")
          }
        )
        ApiResponse$new(deserializedRespObj, resp)
      } else if (httr::status_code(resp) >= 300 && httr::status_code(resp) <= 399) {
        ApiResponse$new(paste("Server returned " , httr::status_code(resp) , " response status code."), resp)
      } else if (httr::status_code(resp) >= 400 && httr::status_code(resp) <= 499) {
        ApiResponse$new("API client error", resp)
      } else if (httr::status_code(resp) >= 500 && httr::status_code(resp) <= 599) {
        ApiResponse$new("API server error", resp)
      }
    },
    GermplasmGermplasmDbIdProgenyGet = function(germplasm.db.id, authorization=NULL, ...){
      apiResponse <- self$GermplasmGermplasmDbIdProgenyGetWithHttpInfo(germplasm.db.id, authorization, ...)
      resp <- apiResponse$response
      if (httr::status_code(resp) >= 200 && httr::status_code(resp) <= 299) {
        apiResponse$content
      } else if (httr::status_code(resp) >= 300 && httr::status_code(resp) <= 399) {
        apiResponse
      } else if (httr::status_code(resp) >= 400 && httr::status_code(resp) <= 499) {
        apiResponse
      } else if (httr::status_code(resp) >= 500 && httr::status_code(resp) <= 599) {
        apiResponse
      }
    },

    GermplasmGermplasmDbIdProgenyGetWithHttpInfo = function(germplasm.db.id, authorization=NULL, ...){
      args <- list(...)
      queryParams <- list()
      headerParams <- c()

      if (missing(`germplasm.db.id`)) {
        stop("Missing required parameter `germplasm.db.id`.")
      }

      headerParams['Authorization'] <- `authorization`

      body <- NULL
      urlPath <- "/germplasm/{germplasmDbId}/progeny"
      if (!missing(`germplasm.db.id`)) {
        urlPath <- gsub(paste0("\\{", "germplasmDbId", "\\}"), URLencode(as.character(`germplasm.db.id`), reserved = TRUE), urlPath)
      }


      resp <- self$apiClient$CallApi(url = paste0(self$apiClient$basePath, urlPath),
                                 method = "GET",
                                 queryParams = queryParams,
                                 headerParams = headerParams,
                                 body = body,
                                 ...)

      if (httr::status_code(resp) >= 200 && httr::status_code(resp) <= 299) {
        deserializedRespObj <- tryCatch(
          self$apiClient$deserialize(resp, "GermplasmProgenyResponse", loadNamespace("openapi")),
          error = function(e){
             stop("Failed to deserialize response")
          }
        )
        ApiResponse$new(deserializedRespObj, resp)
      } else if (httr::status_code(resp) >= 300 && httr::status_code(resp) <= 399) {
        ApiResponse$new(paste("Server returned " , httr::status_code(resp) , " response status code."), resp)
      } else if (httr::status_code(resp) >= 400 && httr::status_code(resp) <= 499) {
        ApiResponse$new("API client error", resp)
      } else if (httr::status_code(resp) >= 500 && httr::status_code(resp) <= 599) {
        ApiResponse$new("API server error", resp)
      }
    },
    GermplasmGermplasmDbIdPut = function(germplasm.db.id, authorization=NULL, germplasm.new.request=NULL, ...){
      apiResponse <- self$GermplasmGermplasmDbIdPutWithHttpInfo(germplasm.db.id, authorization, germplasm.new.request, ...)
      resp <- apiResponse$response
      if (httr::status_code(resp) >= 200 && httr::status_code(resp) <= 299) {
        apiResponse$content
      } else if (httr::status_code(resp) >= 300 && httr::status_code(resp) <= 399) {
        apiResponse
      } else if (httr::status_code(resp) >= 400 && httr::status_code(resp) <= 499) {
        apiResponse
      } else if (httr::status_code(resp) >= 500 && httr::status_code(resp) <= 599) {
        apiResponse
      }
    },

    GermplasmGermplasmDbIdPutWithHttpInfo = function(germplasm.db.id, authorization=NULL, germplasm.new.request=NULL, ...){
      args <- list(...)
      queryParams <- list()
      headerParams <- c()

      if (missing(`germplasm.db.id`)) {
        stop("Missing required parameter `germplasm.db.id`.")
      }

      headerParams['Authorization'] <- `authorization`

      if (!missing(`germplasm.new.request`)) {
        body <- `germplasm.new.request`$toJSONString()
      } else {
        body <- NULL
      }

      urlPath <- "/germplasm/{germplasmDbId}"
      if (!missing(`germplasm.db.id`)) {
        urlPath <- gsub(paste0("\\{", "germplasmDbId", "\\}"), URLencode(as.character(`germplasm.db.id`), reserved = TRUE), urlPath)
      }


      resp <- self$apiClient$CallApi(url = paste0(self$apiClient$basePath, urlPath),
                                 method = "PUT",
                                 queryParams = queryParams,
                                 headerParams = headerParams,
                                 body = body,
                                 ...)

      if (httr::status_code(resp) >= 200 && httr::status_code(resp) <= 299) {
        deserializedRespObj <- tryCatch(
          self$apiClient$deserialize(resp, "GermplasmSingleResponse", loadNamespace("openapi")),
          error = function(e){
             stop("Failed to deserialize response")
          }
        )
        ApiResponse$new(deserializedRespObj, resp)
      } else if (httr::status_code(resp) >= 300 && httr::status_code(resp) <= 399) {
        ApiResponse$new(paste("Server returned " , httr::status_code(resp) , " response status code."), resp)
      } else if (httr::status_code(resp) >= 400 && httr::status_code(resp) <= 499) {
        ApiResponse$new("API client error", resp)
      } else if (httr::status_code(resp) >= 500 && httr::status_code(resp) <= 599) {
        ApiResponse$new("API server error", resp)
      }
    },
    GermplasmGet = function(germplasm.pui=NULL, germplasm.db.id=NULL, germplasm.name=NULL, common.crop.name=NULL, accession.number=NULL, collection=NULL, genus=NULL, species=NULL, study.db.id=NULL, synonym=NULL, parent.db.id=NULL, progeny.db.id=NULL, external.reference.id=NULL, external.reference.source=NULL, page=NULL, page.size=NULL, authorization=NULL, ...){
      apiResponse <- self$GermplasmGetWithHttpInfo(germplasm.pui, germplasm.db.id, germplasm.name, common.crop.name, accession.number, collection, genus, species, study.db.id, synonym, parent.db.id, progeny.db.id, external.reference.id, external.reference.source, page, page.size, authorization, ...)
      resp <- apiResponse$response
      if (httr::status_code(resp) >= 200 && httr::status_code(resp) <= 299) {
        apiResponse$content
      } else if (httr::status_code(resp) >= 300 && httr::status_code(resp) <= 399) {
        apiResponse
      } else if (httr::status_code(resp) >= 400 && httr::status_code(resp) <= 499) {
        apiResponse
      } else if (httr::status_code(resp) >= 500 && httr::status_code(resp) <= 599) {
        apiResponse
      }
    },

    GermplasmGetWithHttpInfo = function(germplasm.pui=NULL, germplasm.db.id=NULL, germplasm.name=NULL, common.crop.name=NULL, accession.number=NULL, collection=NULL, genus=NULL, species=NULL, study.db.id=NULL, synonym=NULL, parent.db.id=NULL, progeny.db.id=NULL, external.reference.id=NULL, external.reference.source=NULL, page=NULL, page.size=NULL, authorization=NULL, ...){
      args <- list(...)
      queryParams <- list()
      headerParams <- c()

      headerParams['Authorization'] <- `authorization`

      queryParams['germplasmPUI'] <- germplasm.pui

      queryParams['germplasmDbId'] <- germplasm.db.id

      queryParams['germplasmName'] <- germplasm.name

      queryParams['commonCropName'] <- common.crop.name

      queryParams['accessionNumber'] <- accession.number

      queryParams['collection'] <- collection

      queryParams['genus'] <- genus

      queryParams['species'] <- species

      queryParams['studyDbId'] <- study.db.id

      queryParams['synonym'] <- synonym

      queryParams['parentDbId'] <- parent.db.id

      queryParams['progenyDbId'] <- progeny.db.id

      queryParams['externalReferenceID'] <- external.reference.id

      queryParams['externalReferenceSource'] <- external.reference.source

      queryParams['page'] <- page

      queryParams['pageSize'] <- page.size

      body <- NULL
      urlPath <- "/germplasm"

      resp <- self$apiClient$CallApi(url = paste0(self$apiClient$basePath, urlPath),
                                 method = "GET",
                                 queryParams = queryParams,
                                 headerParams = headerParams,
                                 body = body,
                                 ...)

      if (httr::status_code(resp) >= 200 && httr::status_code(resp) <= 299) {
        deserializedRespObj <- tryCatch(
          self$apiClient$deserialize(resp, "GermplasmListResponse", loadNamespace("openapi")),
          error = function(e){
             stop("Failed to deserialize response")
          }
        )
        ApiResponse$new(deserializedRespObj, resp)
      } else if (httr::status_code(resp) >= 300 && httr::status_code(resp) <= 399) {
        ApiResponse$new(paste("Server returned " , httr::status_code(resp) , " response status code."), resp)
      } else if (httr::status_code(resp) >= 400 && httr::status_code(resp) <= 499) {
        ApiResponse$new("API client error", resp)
      } else if (httr::status_code(resp) >= 500 && httr::status_code(resp) <= 599) {
        ApiResponse$new("API server error", resp)
      }
    },
    GermplasmPost = function(authorization=NULL, germplasm.new.request=NULL, ...){
      apiResponse <- self$GermplasmPostWithHttpInfo(authorization, germplasm.new.request, ...)
      resp <- apiResponse$response
      if (httr::status_code(resp) >= 200 && httr::status_code(resp) <= 299) {
        apiResponse$content
      } else if (httr::status_code(resp) >= 300 && httr::status_code(resp) <= 399) {
        apiResponse
      } else if (httr::status_code(resp) >= 400 && httr::status_code(resp) <= 499) {
        apiResponse
      } else if (httr::status_code(resp) >= 500 && httr::status_code(resp) <= 599) {
        apiResponse
      }
    },

    GermplasmPostWithHttpInfo = function(authorization=NULL, germplasm.new.request=NULL, ...){
      args <- list(...)
      queryParams <- list()
      headerParams <- c()

      headerParams['Authorization'] <- `authorization`

      if (!missing(`germplasm.new.request`)) {
        body.items = paste(unlist(lapply(germplasm.new.request, function(param){param$toJSONString()})), collapse = ",")
        body <- paste0('[', body.items, ']')
      } else {
        body <- NULL
      }

      urlPath <- "/germplasm"

      resp <- self$apiClient$CallApi(url = paste0(self$apiClient$basePath, urlPath),
                                 method = "POST",
                                 queryParams = queryParams,
                                 headerParams = headerParams,
                                 body = body,
                                 ...)

      if (httr::status_code(resp) >= 200 && httr::status_code(resp) <= 299) {
        deserializedRespObj <- tryCatch(
          self$apiClient$deserialize(resp, "GermplasmListResponse", loadNamespace("openapi")),
          error = function(e){
             stop("Failed to deserialize response")
          }
        )
        ApiResponse$new(deserializedRespObj, resp)
      } else if (httr::status_code(resp) >= 300 && httr::status_code(resp) <= 399) {
        ApiResponse$new(paste("Server returned " , httr::status_code(resp) , " response status code."), resp)
      } else if (httr::status_code(resp) >= 400 && httr::status_code(resp) <= 499) {
        ApiResponse$new("API client error", resp)
      } else if (httr::status_code(resp) >= 500 && httr::status_code(resp) <= 599) {
        ApiResponse$new("API server error", resp)
      }
    },
    SearchGermplasmPost = function(authorization=NULL, germplasm.search.request=NULL, ...){
      apiResponse <- self$SearchGermplasmPostWithHttpInfo(authorization, germplasm.search.request, ...)
      resp <- apiResponse$response
      if (httr::status_code(resp) >= 200 && httr::status_code(resp) <= 299) {
        apiResponse$content
      } else if (httr::status_code(resp) >= 300 && httr::status_code(resp) <= 399) {
        apiResponse
      } else if (httr::status_code(resp) >= 400 && httr::status_code(resp) <= 499) {
        apiResponse
      } else if (httr::status_code(resp) >= 500 && httr::status_code(resp) <= 599) {
        apiResponse
      }
    },

    SearchGermplasmPostWithHttpInfo = function(authorization=NULL, germplasm.search.request=NULL, ...){
      args <- list(...)
      queryParams <- list()
      headerParams <- c()

      headerParams['Authorization'] <- `authorization`

      if (!missing(`germplasm.search.request`)) {
        body <- `germplasm.search.request`$toJSONString()
      } else {
        body <- NULL
      }

      urlPath <- "/search/germplasm"

      resp <- self$apiClient$CallApi(url = paste0(self$apiClient$basePath, urlPath),
                                 method = "POST",
                                 queryParams = queryParams,
                                 headerParams = headerParams,
                                 body = body,
                                 ...)

      if (httr::status_code(resp) >= 200 && httr::status_code(resp) <= 299) {
        deserializedRespObj <- tryCatch(
          self$apiClient$deserialize(resp, "GermplasmListResponse", loadNamespace("openapi")),
          error = function(e){
             stop("Failed to deserialize response")
          }
        )
        ApiResponse$new(deserializedRespObj, resp)
      } else if (httr::status_code(resp) >= 300 && httr::status_code(resp) <= 399) {
        ApiResponse$new(paste("Server returned " , httr::status_code(resp) , " response status code."), resp)
      } else if (httr::status_code(resp) >= 400 && httr::status_code(resp) <= 499) {
        ApiResponse$new("API client error", resp)
      } else if (httr::status_code(resp) >= 500 && httr::status_code(resp) <= 599) {
        ApiResponse$new("API server error", resp)
      }
    },
    SearchGermplasmSearchResultsDbIdGet = function(search.results.db.id, page=NULL, page.size=NULL, authorization=NULL, ...){
      apiResponse <- self$SearchGermplasmSearchResultsDbIdGetWithHttpInfo(search.results.db.id, page, page.size, authorization, ...)
      resp <- apiResponse$response
      if (httr::status_code(resp) >= 200 && httr::status_code(resp) <= 299) {
        apiResponse$content
      } else if (httr::status_code(resp) >= 300 && httr::status_code(resp) <= 399) {
        apiResponse
      } else if (httr::status_code(resp) >= 400 && httr::status_code(resp) <= 499) {
        apiResponse
      } else if (httr::status_code(resp) >= 500 && httr::status_code(resp) <= 599) {
        apiResponse
      }
    },

    SearchGermplasmSearchResultsDbIdGetWithHttpInfo = function(search.results.db.id, page=NULL, page.size=NULL, authorization=NULL, ...){
      args <- list(...)
      queryParams <- list()
      headerParams <- c()

      if (missing(`search.results.db.id`)) {
        stop("Missing required parameter `search.results.db.id`.")
      }

      headerParams['Authorization'] <- `authorization`

      queryParams['page'] <- page

      queryParams['pageSize'] <- page.size

      body <- NULL
      urlPath <- "/search/germplasm/{searchResultsDbId}"
      if (!missing(`search.results.db.id`)) {
        urlPath <- gsub(paste0("\\{", "searchResultsDbId", "\\}"), URLencode(as.character(`search.results.db.id`), reserved = TRUE), urlPath)
      }


      resp <- self$apiClient$CallApi(url = paste0(self$apiClient$basePath, urlPath),
                                 method = "GET",
                                 queryParams = queryParams,
                                 headerParams = headerParams,
                                 body = body,
                                 ...)

      if (httr::status_code(resp) >= 200 && httr::status_code(resp) <= 299) {
        deserializedRespObj <- tryCatch(
          self$apiClient$deserialize(resp, "GermplasmListResponse", loadNamespace("openapi")),
          error = function(e){
             stop("Failed to deserialize response")
          }
        )
        ApiResponse$new(deserializedRespObj, resp)
      } else if (httr::status_code(resp) >= 300 && httr::status_code(resp) <= 399) {
        ApiResponse$new(paste("Server returned " , httr::status_code(resp) , " response status code."), resp)
      } else if (httr::status_code(resp) >= 400 && httr::status_code(resp) <= 499) {
        ApiResponse$new("API client error", resp)
      } else if (httr::status_code(resp) >= 500 && httr::status_code(resp) <= 599) {
        ApiResponse$new("API server error", resp)
      }
    }
  )
)
