# BrAPI-Core
#
# The Breeding API (BrAPI) is a Standardized REST ful Web Service API Specification for communicating Plant Breeding Data. BrAPI allows for easy data sharing between databases and tools involved in plant breeding. <div class=\"brapi-section\"> <h2 class=\"brapi-section-title\">General Reference Documentation</h2> <div class=\"gen-info-link\"><a href=\"https://github.com/plantbreeding/API/blob/master/Specification/GeneralInfo/URL_Structure.md\">URL Structure</a></div> <div class=\"gen-info-link\"><a href=\"https://github.com/plantbreeding/API/blob/master/Specification/GeneralInfo/Response_Structure.md\">Response Structure</a></div> <div class=\"gen-info-link\"><a href=\"https://github.com/plantbreeding/API/blob/master/Specification/GeneralInfo/Date_Time_Encoding.md\">Date/Time Encoding</a></div> <div class=\"gen-info-link\"><a href=\"https://github.com/plantbreeding/API/blob/master/Specification/GeneralInfo/Location_Encoding.md\">Location Encoding</a></div> <div class=\"gen-info-link\"><a href=\"https://github.com/plantbreeding/API/blob/master/Specification/GeneralInfo/Error_Handling.md\">Error Handling</a></div> <div class=\"gen-info-link\"><a href=\"https://github.com/plantbreeding/API/blob/master/Specification/GeneralInfo/Search_Services.md\">Search Services</a></div> </div>  <div class=\"current-brapi-section brapi-section\"> <h2 class=\"brapi-section-title\">BrAPI Core</h2> <div class=\"brapi-section-description\">The BrAPI Core module contains high level entities used for organization and management. This includes Programs, Trials, Studies, Locations, People, and Lists</div> <div class=\"version-number\">V2.0</div> <div class=\"link-btn\"><a href=\"https://github.com/plantbreeding/API/tree/master/Specification/BrAPI-Core\">GitHub</a></div> <div class=\"link-btn\"><a href=\"https://app.swaggerhub.com/apis/PlantBreedingAPI/BrAPI-Core\">SwaggerHub</a></div> <div class=\"link-btn\"><a href=\"https://brapicore.docs.apiary.io\">Apiary</a></div> <div class=\"stop-float\"></div> </div>  <div class=\"brapi-section\"> <h2 class=\"brapi-section-title\">BrAPI Phenotyping</h2> <div class=\"brapi-section-description\">The BrAPI Phenotyping module contains entities related to phenotypic observations. This includes Observation Units, Observations, Observation Variables, Traits, Scales, Methods, and Images</div> <div class=\"version-number\">V2.0</div> <div class=\"link-btn\"><a href=\"https://github.com/plantbreeding/API/tree/master/Specification/BrAPI-Phenotyping\">GitHub</a></div> <div class=\"link-btn\"><a href=\"https://app.swaggerhub.com/apis/PlantBreedingAPI/BrAPI-Phenotyping\">SwaggerHub</a></div> <div class=\"link-btn\"><a href=\"https://brapiphenotyping.docs.apiary.io\">Apiary</a></div> <div class=\"stop-float\"></div> </div>  <div class=\"brapi-section\"> <h2 class=\"brapi-section-title\">BrAPI Genotyping</h2> <div class=\"brapi-section-description\">The BrAPI Genotyping module contains entities related to genotyping analysis. This includes Samples, Markers, Variant Sets, Variants, Call Sets, Calls, References, Reads, and Vendor Orders</div> <div class=\"version-number\">V2.0</div> <div class=\"link-btn\"><a href=\"https://github.com/plantbreeding/API/tree/master/Specification/BrAPI-Genotyping\">GitHub</a></div> <div class=\"link-btn\"><a href=\"https://app.swaggerhub.com/apis/PlantBreedingAPI/BrAPI-Genotyping\">SwaggerHub</a></div> <div class=\"link-btn\"><a href=\"https://brapigenotyping.docs.apiary.io\">Apiary</a></div> <div class=\"stop-float\"></div> </div>  <div class=\"brapi-section\"> <h2 class=\"brapi-section-title\">BrAPI Germplasm</h2> <div class=\"brapi-section-description\">The BrAPI Germplasm module contains entities related to germplasm management. This includes Germplasm, Germplasm Attributes, Seed Lots, Crosses, Pedigree, and Progeny</div> <div class=\"version-number\">V2.0</div> <div class=\"link-btn\"><a href=\"https://github.com/plantbreeding/API/tree/master/Specification/BrAPI-Germplasm\">GitHub</a></div> <div class=\"link-btn\"><a href=\"https://app.swaggerhub.com/apis/PlantBreedingAPI/BrAPI-Germplasm\">SwaggerHub</a></div> <div class=\"link-btn\"><a href=\"https://brapigermplasm.docs.apiary.io\">Apiary</a></div> <div class=\"stop-float\"></div> </div>  <style> .link-btn{ float: left;  margin: 2px 10px 0 0;  padding: 0 5px;  border-radius: 5px;  background-color: #ddd; } .stop-float{   clear: both; } .version-number{   float: left;    margin: 5px 10px 0 5px; } .brapi-section-title{   margin: 0 10px 0 0;   font-size: 20px; } .current-brapi-section{   font-weight: bolder;   border-radius: 5px;    background-color: #ddd; } .brapi-section{   padding: 5px 5px;  } .brapi-section-description{   margin: 5px 0 0 5px; } </style>
#
# The version of the OpenAPI document: 2.0
# 
# Generated by: https://openapi-generator.tech

#' @docType class
#' @title GermplasmSearchRequest
#'
#' @description GermplasmSearchRequest Class
#'
#' @format An \code{R6Class} generator object
#'
#' @field page  integer [optional]
#'
#' @field pageSize  integer [optional]
#'
#' @field commonCropNames  list( character ) [optional]
#'
#' @field germplasmDbIds  list( character ) [optional]
#'
#' @field germplasmNames  list( character ) [optional]
#'
#' @field studyDbIds  list( character ) [optional]
#'
#' @field studyNames  list( character ) [optional]
#'
#' @field externalReferenceIDs  list( character ) [optional]
#'
#' @field externalReferenceSources  list( character ) [optional]
#'
#' @field accessionNumbers  list( character ) [optional]
#'
#' @field collections  list( character ) [optional]
#'
#' @field genus  list( character ) [optional]
#'
#' @field germplasmPUIs  list( character ) [optional]
#'
#' @field parentDbIds  list( character ) [optional]
#'
#' @field progenyDbIds  list( character ) [optional]
#'
#' @field species  list( character ) [optional]
#'
#' @field synonyms  list( character ) [optional]
#'
#' @importFrom R6 R6Class
#' @importFrom jsonlite fromJSON toJSON
#' @export
GermplasmSearchRequest <- R6::R6Class(
  'GermplasmSearchRequest',
  public = list(
    `page` = NULL,
    `pageSize` = NULL,
    `commonCropNames` = NULL,
    `germplasmDbIds` = NULL,
    `germplasmNames` = NULL,
    `studyDbIds` = NULL,
    `studyNames` = NULL,
    `externalReferenceIDs` = NULL,
    `externalReferenceSources` = NULL,
    `accessionNumbers` = NULL,
    `collections` = NULL,
    `genus` = NULL,
    `germplasmPUIs` = NULL,
    `parentDbIds` = NULL,
    `progenyDbIds` = NULL,
    `species` = NULL,
    `synonyms` = NULL,
    initialize = function(
        `page`=NULL, `pageSize`=NULL, `commonCropNames`=NULL, `germplasmDbIds`=NULL, `germplasmNames`=NULL, `studyDbIds`=NULL, `studyNames`=NULL, `externalReferenceIDs`=NULL, `externalReferenceSources`=NULL, `accessionNumbers`=NULL, `collections`=NULL, `genus`=NULL, `germplasmPUIs`=NULL, `parentDbIds`=NULL, `progenyDbIds`=NULL, `species`=NULL, `synonyms`=NULL, ...
    ) {
      local.optional.var <- list(...)
      if (!is.null(`page`)) {
        stopifnot(is.numeric(`page`), length(`page`) == 1)
        self$`page` <- `page`
      }
      if (!is.null(`pageSize`)) {
        stopifnot(is.numeric(`pageSize`), length(`pageSize`) == 1)
        self$`pageSize` <- `pageSize`
      }
      if (!is.null(`commonCropNames`)) {
        stopifnot(is.vector(`commonCropNames`), length(`commonCropNames`) != 0)
        sapply(`commonCropNames`, function(x) stopifnot(is.character(x)))
        self$`commonCropNames` <- `commonCropNames`
      }
      if (!is.null(`germplasmDbIds`)) {
        stopifnot(is.vector(`germplasmDbIds`), length(`germplasmDbIds`) != 0)
        sapply(`germplasmDbIds`, function(x) stopifnot(is.character(x)))
        self$`germplasmDbIds` <- `germplasmDbIds`
      }
      if (!is.null(`germplasmNames`)) {
        stopifnot(is.vector(`germplasmNames`), length(`germplasmNames`) != 0)
        sapply(`germplasmNames`, function(x) stopifnot(is.character(x)))
        self$`germplasmNames` <- `germplasmNames`
      }
      if (!is.null(`studyDbIds`)) {
        stopifnot(is.vector(`studyDbIds`), length(`studyDbIds`) != 0)
        sapply(`studyDbIds`, function(x) stopifnot(is.character(x)))
        self$`studyDbIds` <- `studyDbIds`
      }
      if (!is.null(`studyNames`)) {
        stopifnot(is.vector(`studyNames`), length(`studyNames`) != 0)
        sapply(`studyNames`, function(x) stopifnot(is.character(x)))
        self$`studyNames` <- `studyNames`
      }
      if (!is.null(`externalReferenceIDs`)) {
        stopifnot(is.vector(`externalReferenceIDs`), length(`externalReferenceIDs`) != 0)
        sapply(`externalReferenceIDs`, function(x) stopifnot(is.character(x)))
        self$`externalReferenceIDs` <- `externalReferenceIDs`
      }
      if (!is.null(`externalReferenceSources`)) {
        stopifnot(is.vector(`externalReferenceSources`), length(`externalReferenceSources`) != 0)
        sapply(`externalReferenceSources`, function(x) stopifnot(is.character(x)))
        self$`externalReferenceSources` <- `externalReferenceSources`
      }
      if (!is.null(`accessionNumbers`)) {
        stopifnot(is.vector(`accessionNumbers`), length(`accessionNumbers`) != 0)
        sapply(`accessionNumbers`, function(x) stopifnot(is.character(x)))
        self$`accessionNumbers` <- `accessionNumbers`
      }
      if (!is.null(`collections`)) {
        stopifnot(is.vector(`collections`), length(`collections`) != 0)
        sapply(`collections`, function(x) stopifnot(is.character(x)))
        self$`collections` <- `collections`
      }
      if (!is.null(`genus`)) {
        stopifnot(is.vector(`genus`), length(`genus`) != 0)
        sapply(`genus`, function(x) stopifnot(is.character(x)))
        self$`genus` <- `genus`
      }
      if (!is.null(`germplasmPUIs`)) {
        stopifnot(is.vector(`germplasmPUIs`), length(`germplasmPUIs`) != 0)
        sapply(`germplasmPUIs`, function(x) stopifnot(is.character(x)))
        self$`germplasmPUIs` <- `germplasmPUIs`
      }
      if (!is.null(`parentDbIds`)) {
        stopifnot(is.vector(`parentDbIds`), length(`parentDbIds`) != 0)
        sapply(`parentDbIds`, function(x) stopifnot(is.character(x)))
        self$`parentDbIds` <- `parentDbIds`
      }
      if (!is.null(`progenyDbIds`)) {
        stopifnot(is.vector(`progenyDbIds`), length(`progenyDbIds`) != 0)
        sapply(`progenyDbIds`, function(x) stopifnot(is.character(x)))
        self$`progenyDbIds` <- `progenyDbIds`
      }
      if (!is.null(`species`)) {
        stopifnot(is.vector(`species`), length(`species`) != 0)
        sapply(`species`, function(x) stopifnot(is.character(x)))
        self$`species` <- `species`
      }
      if (!is.null(`synonyms`)) {
        stopifnot(is.vector(`synonyms`), length(`synonyms`) != 0)
        sapply(`synonyms`, function(x) stopifnot(is.character(x)))
        self$`synonyms` <- `synonyms`
      }
    },
    toJSON = function() {
      GermplasmSearchRequestObject <- list()
      if (!is.null(self$`page`)) {
        GermplasmSearchRequestObject[['page']] <-
          self$`page`
      }
      if (!is.null(self$`pageSize`)) {
        GermplasmSearchRequestObject[['pageSize']] <-
          self$`pageSize`
      }
      if (!is.null(self$`commonCropNames`)) {
        GermplasmSearchRequestObject[['commonCropNames']] <-
          self$`commonCropNames`
      }
      if (!is.null(self$`germplasmDbIds`)) {
        GermplasmSearchRequestObject[['germplasmDbIds']] <-
          self$`germplasmDbIds`
      }
      if (!is.null(self$`germplasmNames`)) {
        GermplasmSearchRequestObject[['germplasmNames']] <-
          self$`germplasmNames`
      }
      if (!is.null(self$`studyDbIds`)) {
        GermplasmSearchRequestObject[['studyDbIds']] <-
          self$`studyDbIds`
      }
      if (!is.null(self$`studyNames`)) {
        GermplasmSearchRequestObject[['studyNames']] <-
          self$`studyNames`
      }
      if (!is.null(self$`externalReferenceIDs`)) {
        GermplasmSearchRequestObject[['externalReferenceIDs']] <-
          self$`externalReferenceIDs`
      }
      if (!is.null(self$`externalReferenceSources`)) {
        GermplasmSearchRequestObject[['externalReferenceSources']] <-
          self$`externalReferenceSources`
      }
      if (!is.null(self$`accessionNumbers`)) {
        GermplasmSearchRequestObject[['accessionNumbers']] <-
          self$`accessionNumbers`
      }
      if (!is.null(self$`collections`)) {
        GermplasmSearchRequestObject[['collections']] <-
          self$`collections`
      }
      if (!is.null(self$`genus`)) {
        GermplasmSearchRequestObject[['genus']] <-
          self$`genus`
      }
      if (!is.null(self$`germplasmPUIs`)) {
        GermplasmSearchRequestObject[['germplasmPUIs']] <-
          self$`germplasmPUIs`
      }
      if (!is.null(self$`parentDbIds`)) {
        GermplasmSearchRequestObject[['parentDbIds']] <-
          self$`parentDbIds`
      }
      if (!is.null(self$`progenyDbIds`)) {
        GermplasmSearchRequestObject[['progenyDbIds']] <-
          self$`progenyDbIds`
      }
      if (!is.null(self$`species`)) {
        GermplasmSearchRequestObject[['species']] <-
          self$`species`
      }
      if (!is.null(self$`synonyms`)) {
        GermplasmSearchRequestObject[['synonyms']] <-
          self$`synonyms`
      }

      GermplasmSearchRequestObject
    },
    fromJSON = function(GermplasmSearchRequestJson) {
      GermplasmSearchRequestObject <- jsonlite::fromJSON(GermplasmSearchRequestJson)
      if (!is.null(GermplasmSearchRequestObject$`page`)) {
        self$`page` <- GermplasmSearchRequestObject$`page`
      }
      if (!is.null(GermplasmSearchRequestObject$`pageSize`)) {
        self$`pageSize` <- GermplasmSearchRequestObject$`pageSize`
      }
      if (!is.null(GermplasmSearchRequestObject$`commonCropNames`)) {
        self$`commonCropNames` <- ApiClient$new()$deserializeObj(GermplasmSearchRequestObject$`commonCropNames`, "array[character]", loadNamespace("openapi"))
      }
      if (!is.null(GermplasmSearchRequestObject$`germplasmDbIds`)) {
        self$`germplasmDbIds` <- ApiClient$new()$deserializeObj(GermplasmSearchRequestObject$`germplasmDbIds`, "array[character]", loadNamespace("openapi"))
      }
      if (!is.null(GermplasmSearchRequestObject$`germplasmNames`)) {
        self$`germplasmNames` <- ApiClient$new()$deserializeObj(GermplasmSearchRequestObject$`germplasmNames`, "array[character]", loadNamespace("openapi"))
      }
      if (!is.null(GermplasmSearchRequestObject$`studyDbIds`)) {
        self$`studyDbIds` <- ApiClient$new()$deserializeObj(GermplasmSearchRequestObject$`studyDbIds`, "array[character]", loadNamespace("openapi"))
      }
      if (!is.null(GermplasmSearchRequestObject$`studyNames`)) {
        self$`studyNames` <- ApiClient$new()$deserializeObj(GermplasmSearchRequestObject$`studyNames`, "array[character]", loadNamespace("openapi"))
      }
      if (!is.null(GermplasmSearchRequestObject$`externalReferenceIDs`)) {
        self$`externalReferenceIDs` <- ApiClient$new()$deserializeObj(GermplasmSearchRequestObject$`externalReferenceIDs`, "array[character]", loadNamespace("openapi"))
      }
      if (!is.null(GermplasmSearchRequestObject$`externalReferenceSources`)) {
        self$`externalReferenceSources` <- ApiClient$new()$deserializeObj(GermplasmSearchRequestObject$`externalReferenceSources`, "array[character]", loadNamespace("openapi"))
      }
      if (!is.null(GermplasmSearchRequestObject$`accessionNumbers`)) {
        self$`accessionNumbers` <- ApiClient$new()$deserializeObj(GermplasmSearchRequestObject$`accessionNumbers`, "array[character]", loadNamespace("openapi"))
      }
      if (!is.null(GermplasmSearchRequestObject$`collections`)) {
        self$`collections` <- ApiClient$new()$deserializeObj(GermplasmSearchRequestObject$`collections`, "array[character]", loadNamespace("openapi"))
      }
      if (!is.null(GermplasmSearchRequestObject$`genus`)) {
        self$`genus` <- ApiClient$new()$deserializeObj(GermplasmSearchRequestObject$`genus`, "array[character]", loadNamespace("openapi"))
      }
      if (!is.null(GermplasmSearchRequestObject$`germplasmPUIs`)) {
        self$`germplasmPUIs` <- ApiClient$new()$deserializeObj(GermplasmSearchRequestObject$`germplasmPUIs`, "array[character]", loadNamespace("openapi"))
      }
      if (!is.null(GermplasmSearchRequestObject$`parentDbIds`)) {
        self$`parentDbIds` <- ApiClient$new()$deserializeObj(GermplasmSearchRequestObject$`parentDbIds`, "array[character]", loadNamespace("openapi"))
      }
      if (!is.null(GermplasmSearchRequestObject$`progenyDbIds`)) {
        self$`progenyDbIds` <- ApiClient$new()$deserializeObj(GermplasmSearchRequestObject$`progenyDbIds`, "array[character]", loadNamespace("openapi"))
      }
      if (!is.null(GermplasmSearchRequestObject$`species`)) {
        self$`species` <- ApiClient$new()$deserializeObj(GermplasmSearchRequestObject$`species`, "array[character]", loadNamespace("openapi"))
      }
      if (!is.null(GermplasmSearchRequestObject$`synonyms`)) {
        self$`synonyms` <- ApiClient$new()$deserializeObj(GermplasmSearchRequestObject$`synonyms`, "array[character]", loadNamespace("openapi"))
      }
      self
    },
    toJSONString = function() {
      jsoncontent <- c(
        if (!is.null(self$`page`)) {
        sprintf(
        '"page":
          %d
                ',
        self$`page`
        )},
        if (!is.null(self$`pageSize`)) {
        sprintf(
        '"pageSize":
          %d
                ',
        self$`pageSize`
        )},
        if (!is.null(self$`commonCropNames`)) {
        sprintf(
        '"commonCropNames":
           [%s]
        ',
        paste(unlist(lapply(self$`commonCropNames`, function(x) paste0('"', x, '"'))), collapse=",")
        )},
        if (!is.null(self$`germplasmDbIds`)) {
        sprintf(
        '"germplasmDbIds":
           [%s]
        ',
        paste(unlist(lapply(self$`germplasmDbIds`, function(x) paste0('"', x, '"'))), collapse=",")
        )},
        if (!is.null(self$`germplasmNames`)) {
        sprintf(
        '"germplasmNames":
           [%s]
        ',
        paste(unlist(lapply(self$`germplasmNames`, function(x) paste0('"', x, '"'))), collapse=",")
        )},
        if (!is.null(self$`studyDbIds`)) {
        sprintf(
        '"studyDbIds":
           [%s]
        ',
        paste(unlist(lapply(self$`studyDbIds`, function(x) paste0('"', x, '"'))), collapse=",")
        )},
        if (!is.null(self$`studyNames`)) {
        sprintf(
        '"studyNames":
           [%s]
        ',
        paste(unlist(lapply(self$`studyNames`, function(x) paste0('"', x, '"'))), collapse=",")
        )},
        if (!is.null(self$`externalReferenceIDs`)) {
        sprintf(
        '"externalReferenceIDs":
           [%s]
        ',
        paste(unlist(lapply(self$`externalReferenceIDs`, function(x) paste0('"', x, '"'))), collapse=",")
        )},
        if (!is.null(self$`externalReferenceSources`)) {
        sprintf(
        '"externalReferenceSources":
           [%s]
        ',
        paste(unlist(lapply(self$`externalReferenceSources`, function(x) paste0('"', x, '"'))), collapse=",")
        )},
        if (!is.null(self$`accessionNumbers`)) {
        sprintf(
        '"accessionNumbers":
           [%s]
        ',
        paste(unlist(lapply(self$`accessionNumbers`, function(x) paste0('"', x, '"'))), collapse=",")
        )},
        if (!is.null(self$`collections`)) {
        sprintf(
        '"collections":
           [%s]
        ',
        paste(unlist(lapply(self$`collections`, function(x) paste0('"', x, '"'))), collapse=",")
        )},
        if (!is.null(self$`genus`)) {
        sprintf(
        '"genus":
           [%s]
        ',
        paste(unlist(lapply(self$`genus`, function(x) paste0('"', x, '"'))), collapse=",")
        )},
        if (!is.null(self$`germplasmPUIs`)) {
        sprintf(
        '"germplasmPUIs":
           [%s]
        ',
        paste(unlist(lapply(self$`germplasmPUIs`, function(x) paste0('"', x, '"'))), collapse=",")
        )},
        if (!is.null(self$`parentDbIds`)) {
        sprintf(
        '"parentDbIds":
           [%s]
        ',
        paste(unlist(lapply(self$`parentDbIds`, function(x) paste0('"', x, '"'))), collapse=",")
        )},
        if (!is.null(self$`progenyDbIds`)) {
        sprintf(
        '"progenyDbIds":
           [%s]
        ',
        paste(unlist(lapply(self$`progenyDbIds`, function(x) paste0('"', x, '"'))), collapse=",")
        )},
        if (!is.null(self$`species`)) {
        sprintf(
        '"species":
           [%s]
        ',
        paste(unlist(lapply(self$`species`, function(x) paste0('"', x, '"'))), collapse=",")
        )},
        if (!is.null(self$`synonyms`)) {
        sprintf(
        '"synonyms":
           [%s]
        ',
        paste(unlist(lapply(self$`synonyms`, function(x) paste0('"', x, '"'))), collapse=",")
        )}
      )
      jsoncontent <- paste(jsoncontent, collapse = ",")
      paste('{', jsoncontent, '}', sep = "")
    },
    fromJSONString = function(GermplasmSearchRequestJson) {
      GermplasmSearchRequestObject <- jsonlite::fromJSON(GermplasmSearchRequestJson)
      self$`page` <- GermplasmSearchRequestObject$`page`
      self$`pageSize` <- GermplasmSearchRequestObject$`pageSize`
      self$`commonCropNames` <- ApiClient$new()$deserializeObj(GermplasmSearchRequestObject$`commonCropNames`, "array[character]", loadNamespace("openapi"))
      self$`germplasmDbIds` <- ApiClient$new()$deserializeObj(GermplasmSearchRequestObject$`germplasmDbIds`, "array[character]", loadNamespace("openapi"))
      self$`germplasmNames` <- ApiClient$new()$deserializeObj(GermplasmSearchRequestObject$`germplasmNames`, "array[character]", loadNamespace("openapi"))
      self$`studyDbIds` <- ApiClient$new()$deserializeObj(GermplasmSearchRequestObject$`studyDbIds`, "array[character]", loadNamespace("openapi"))
      self$`studyNames` <- ApiClient$new()$deserializeObj(GermplasmSearchRequestObject$`studyNames`, "array[character]", loadNamespace("openapi"))
      self$`externalReferenceIDs` <- ApiClient$new()$deserializeObj(GermplasmSearchRequestObject$`externalReferenceIDs`, "array[character]", loadNamespace("openapi"))
      self$`externalReferenceSources` <- ApiClient$new()$deserializeObj(GermplasmSearchRequestObject$`externalReferenceSources`, "array[character]", loadNamespace("openapi"))
      self$`accessionNumbers` <- ApiClient$new()$deserializeObj(GermplasmSearchRequestObject$`accessionNumbers`, "array[character]", loadNamespace("openapi"))
      self$`collections` <- ApiClient$new()$deserializeObj(GermplasmSearchRequestObject$`collections`, "array[character]", loadNamespace("openapi"))
      self$`genus` <- ApiClient$new()$deserializeObj(GermplasmSearchRequestObject$`genus`, "array[character]", loadNamespace("openapi"))
      self$`germplasmPUIs` <- ApiClient$new()$deserializeObj(GermplasmSearchRequestObject$`germplasmPUIs`, "array[character]", loadNamespace("openapi"))
      self$`parentDbIds` <- ApiClient$new()$deserializeObj(GermplasmSearchRequestObject$`parentDbIds`, "array[character]", loadNamespace("openapi"))
      self$`progenyDbIds` <- ApiClient$new()$deserializeObj(GermplasmSearchRequestObject$`progenyDbIds`, "array[character]", loadNamespace("openapi"))
      self$`species` <- ApiClient$new()$deserializeObj(GermplasmSearchRequestObject$`species`, "array[character]", loadNamespace("openapi"))
      self$`synonyms` <- ApiClient$new()$deserializeObj(GermplasmSearchRequestObject$`synonyms`, "array[character]", loadNamespace("openapi"))
      self
    }
  )
)

