# BrAPI-Core
#
# The Breeding API (BrAPI) is a Standardized REST ful Web Service API Specification for communicating Plant Breeding Data. BrAPI allows for easy data sharing between databases and tools involved in plant breeding. <div class=\"brapi-section\"> <h2 class=\"brapi-section-title\">General Reference Documentation</h2> <div class=\"gen-info-link\"><a href=\"https://github.com/plantbreeding/API/blob/master/Specification/GeneralInfo/URL_Structure.md\">URL Structure</a></div> <div class=\"gen-info-link\"><a href=\"https://github.com/plantbreeding/API/blob/master/Specification/GeneralInfo/Response_Structure.md\">Response Structure</a></div> <div class=\"gen-info-link\"><a href=\"https://github.com/plantbreeding/API/blob/master/Specification/GeneralInfo/Date_Time_Encoding.md\">Date/Time Encoding</a></div> <div class=\"gen-info-link\"><a href=\"https://github.com/plantbreeding/API/blob/master/Specification/GeneralInfo/Location_Encoding.md\">Location Encoding</a></div> <div class=\"gen-info-link\"><a href=\"https://github.com/plantbreeding/API/blob/master/Specification/GeneralInfo/Error_Handling.md\">Error Handling</a></div> <div class=\"gen-info-link\"><a href=\"https://github.com/plantbreeding/API/blob/master/Specification/GeneralInfo/Search_Services.md\">Search Services</a></div> </div>  <div class=\"current-brapi-section brapi-section\"> <h2 class=\"brapi-section-title\">BrAPI Core</h2> <div class=\"brapi-section-description\">The BrAPI Core module contains high level entities used for organization and management. This includes Programs, Trials, Studies, Locations, People, and Lists</div> <div class=\"version-number\">V2.0</div> <div class=\"link-btn\"><a href=\"https://github.com/plantbreeding/API/tree/master/Specification/BrAPI-Core\">GitHub</a></div> <div class=\"link-btn\"><a href=\"https://app.swaggerhub.com/apis/PlantBreedingAPI/BrAPI-Core\">SwaggerHub</a></div> <div class=\"link-btn\"><a href=\"https://brapicore.docs.apiary.io\">Apiary</a></div> <div class=\"stop-float\"></div> </div>  <div class=\"brapi-section\"> <h2 class=\"brapi-section-title\">BrAPI Phenotyping</h2> <div class=\"brapi-section-description\">The BrAPI Phenotyping module contains entities related to phenotypic observations. This includes Observation Units, Observations, Observation Variables, Traits, Scales, Methods, and Images</div> <div class=\"version-number\">V2.0</div> <div class=\"link-btn\"><a href=\"https://github.com/plantbreeding/API/tree/master/Specification/BrAPI-Phenotyping\">GitHub</a></div> <div class=\"link-btn\"><a href=\"https://app.swaggerhub.com/apis/PlantBreedingAPI/BrAPI-Phenotyping\">SwaggerHub</a></div> <div class=\"link-btn\"><a href=\"https://brapiphenotyping.docs.apiary.io\">Apiary</a></div> <div class=\"stop-float\"></div> </div>  <div class=\"brapi-section\"> <h2 class=\"brapi-section-title\">BrAPI Genotyping</h2> <div class=\"brapi-section-description\">The BrAPI Genotyping module contains entities related to genotyping analysis. This includes Samples, Markers, Variant Sets, Variants, Call Sets, Calls, References, Reads, and Vendor Orders</div> <div class=\"version-number\">V2.0</div> <div class=\"link-btn\"><a href=\"https://github.com/plantbreeding/API/tree/master/Specification/BrAPI-Genotyping\">GitHub</a></div> <div class=\"link-btn\"><a href=\"https://app.swaggerhub.com/apis/PlantBreedingAPI/BrAPI-Genotyping\">SwaggerHub</a></div> <div class=\"link-btn\"><a href=\"https://brapigenotyping.docs.apiary.io\">Apiary</a></div> <div class=\"stop-float\"></div> </div>  <div class=\"brapi-section\"> <h2 class=\"brapi-section-title\">BrAPI Germplasm</h2> <div class=\"brapi-section-description\">The BrAPI Germplasm module contains entities related to germplasm management. This includes Germplasm, Germplasm Attributes, Seed Lots, Crosses, Pedigree, and Progeny</div> <div class=\"version-number\">V2.0</div> <div class=\"link-btn\"><a href=\"https://github.com/plantbreeding/API/tree/master/Specification/BrAPI-Germplasm\">GitHub</a></div> <div class=\"link-btn\"><a href=\"https://app.swaggerhub.com/apis/PlantBreedingAPI/BrAPI-Germplasm\">SwaggerHub</a></div> <div class=\"link-btn\"><a href=\"https://brapigermplasm.docs.apiary.io\">Apiary</a></div> <div class=\"stop-float\"></div> </div>  <style> .link-btn{ float: left;  margin: 2px 10px 0 0;  padding: 0 5px;  border-radius: 5px;  background-color: #ddd; } .stop-float{   clear: both; } .version-number{   float: left;    margin: 5px 10px 0 5px; } .brapi-section-title{   margin: 0 10px 0 0;   font-size: 20px; } .current-brapi-section{   font-weight: bolder;   border-radius: 5px;    background-color: #ddd; } .brapi-section{   padding: 5px 5px;  } .brapi-section-description{   margin: 5px 0 0 5px; } </style>
#
# The version of the OpenAPI document: 2.0
# 
# Generated by: https://openapi-generator.tech

#' @docType class
#' @title GermplasmMCPDCollectingInfo
#'
#' @description GermplasmMCPDCollectingInfo Class
#'
#' @format An \code{R6Class} generator object
#'
#' @field collectingDate  character [optional]
#'
#' @field collectingInstitutes  list( \link{GermplasmMCPDCollectingInfoCollectingInstitutes} ) [optional]
#'
#' @field collectingMissionIdentifier  character [optional]
#'
#' @field collectingNumber  character [optional]
#'
#' @field collectingSite  \link{GermplasmMCPDCollectingInfoCollectingSite} [optional]
#'
#' @importFrom R6 R6Class
#' @importFrom jsonlite fromJSON toJSON
#' @export
GermplasmMCPDCollectingInfo <- R6::R6Class(
  'GermplasmMCPDCollectingInfo',
  public = list(
    `collectingDate` = NULL,
    `collectingInstitutes` = NULL,
    `collectingMissionIdentifier` = NULL,
    `collectingNumber` = NULL,
    `collectingSite` = NULL,
    initialize = function(
        `collectingDate`=NULL, `collectingInstitutes`=NULL, `collectingMissionIdentifier`=NULL, `collectingNumber`=NULL, `collectingSite`=NULL, ...
    ) {
      local.optional.var <- list(...)
      if (!is.null(`collectingDate`)) {
        stopifnot(is.character(`collectingDate`), length(`collectingDate`) == 1)
        self$`collectingDate` <- `collectingDate`
      }
      if (!is.null(`collectingInstitutes`)) {
        stopifnot(is.vector(`collectingInstitutes`), length(`collectingInstitutes`) != 0)
        sapply(`collectingInstitutes`, function(x) stopifnot(R6::is.R6(x)))
        self$`collectingInstitutes` <- `collectingInstitutes`
      }
      if (!is.null(`collectingMissionIdentifier`)) {
        stopifnot(is.character(`collectingMissionIdentifier`), length(`collectingMissionIdentifier`) == 1)
        self$`collectingMissionIdentifier` <- `collectingMissionIdentifier`
      }
      if (!is.null(`collectingNumber`)) {
        stopifnot(is.character(`collectingNumber`), length(`collectingNumber`) == 1)
        self$`collectingNumber` <- `collectingNumber`
      }
      if (!is.null(`collectingSite`)) {
        stopifnot(R6::is.R6(`collectingSite`))
        self$`collectingSite` <- `collectingSite`
      }
    },
    toJSON = function() {
      GermplasmMCPDCollectingInfoObject <- list()
      if (!is.null(self$`collectingDate`)) {
        GermplasmMCPDCollectingInfoObject[['collectingDate']] <-
          self$`collectingDate`
      }
      if (!is.null(self$`collectingInstitutes`)) {
        GermplasmMCPDCollectingInfoObject[['collectingInstitutes']] <-
          lapply(self$`collectingInstitutes`, function(x) x$toJSON())
      }
      if (!is.null(self$`collectingMissionIdentifier`)) {
        GermplasmMCPDCollectingInfoObject[['collectingMissionIdentifier']] <-
          self$`collectingMissionIdentifier`
      }
      if (!is.null(self$`collectingNumber`)) {
        GermplasmMCPDCollectingInfoObject[['collectingNumber']] <-
          self$`collectingNumber`
      }
      if (!is.null(self$`collectingSite`)) {
        GermplasmMCPDCollectingInfoObject[['collectingSite']] <-
          self$`collectingSite`$toJSON()
      }

      GermplasmMCPDCollectingInfoObject
    },
    fromJSON = function(GermplasmMCPDCollectingInfoJson) {
      GermplasmMCPDCollectingInfoObject <- jsonlite::fromJSON(GermplasmMCPDCollectingInfoJson)
      if (!is.null(GermplasmMCPDCollectingInfoObject$`collectingDate`)) {
        self$`collectingDate` <- GermplasmMCPDCollectingInfoObject$`collectingDate`
      }
      if (!is.null(GermplasmMCPDCollectingInfoObject$`collectingInstitutes`)) {
        self$`collectingInstitutes` <- ApiClient$new()$deserializeObj(GermplasmMCPDCollectingInfoObject$`collectingInstitutes`, "array[GermplasmMCPDCollectingInfoCollectingInstitutes]", loadNamespace("openapi"))
      }
      if (!is.null(GermplasmMCPDCollectingInfoObject$`collectingMissionIdentifier`)) {
        self$`collectingMissionIdentifier` <- GermplasmMCPDCollectingInfoObject$`collectingMissionIdentifier`
      }
      if (!is.null(GermplasmMCPDCollectingInfoObject$`collectingNumber`)) {
        self$`collectingNumber` <- GermplasmMCPDCollectingInfoObject$`collectingNumber`
      }
      if (!is.null(GermplasmMCPDCollectingInfoObject$`collectingSite`)) {
        collectingSiteObject <- GermplasmMCPDCollectingInfoCollectingSite$new()
        collectingSiteObject$fromJSON(jsonlite::toJSON(GermplasmMCPDCollectingInfoObject$collectingSite, auto_unbox = TRUE, digits = NA))
        self$`collectingSite` <- collectingSiteObject
      }
      self
    },
    toJSONString = function() {
      jsoncontent <- c(
        if (!is.null(self$`collectingDate`)) {
        sprintf(
        '"collectingDate":
          "%s"
                ',
        self$`collectingDate`
        )},
        if (!is.null(self$`collectingInstitutes`)) {
        sprintf(
        '"collectingInstitutes":
        [%s]
',
        paste(sapply(self$`collectingInstitutes`, function(x) jsonlite::toJSON(x$toJSON(), auto_unbox=TRUE, digits = NA)), collapse=",")
        )},
        if (!is.null(self$`collectingMissionIdentifier`)) {
        sprintf(
        '"collectingMissionIdentifier":
          "%s"
                ',
        self$`collectingMissionIdentifier`
        )},
        if (!is.null(self$`collectingNumber`)) {
        sprintf(
        '"collectingNumber":
          "%s"
                ',
        self$`collectingNumber`
        )},
        if (!is.null(self$`collectingSite`)) {
        sprintf(
        '"collectingSite":
        %s
        ',
        jsonlite::toJSON(self$`collectingSite`$toJSON(), auto_unbox=TRUE, digits = NA)
        )}
      )
      jsoncontent <- paste(jsoncontent, collapse = ",")
      paste('{', jsoncontent, '}', sep = "")
    },
    fromJSONString = function(GermplasmMCPDCollectingInfoJson) {
      GermplasmMCPDCollectingInfoObject <- jsonlite::fromJSON(GermplasmMCPDCollectingInfoJson)
      self$`collectingDate` <- GermplasmMCPDCollectingInfoObject$`collectingDate`
      self$`collectingInstitutes` <- ApiClient$new()$deserializeObj(GermplasmMCPDCollectingInfoObject$`collectingInstitutes`, "array[GermplasmMCPDCollectingInfoCollectingInstitutes]", loadNamespace("openapi"))
      self$`collectingMissionIdentifier` <- GermplasmMCPDCollectingInfoObject$`collectingMissionIdentifier`
      self$`collectingNumber` <- GermplasmMCPDCollectingInfoObject$`collectingNumber`
      self$`collectingSite` <- GermplasmMCPDCollectingInfoCollectingSite$new()$fromJSON(jsonlite::toJSON(GermplasmMCPDCollectingInfoObject$collectingSite, auto_unbox = TRUE, digits = NA))
      self
    }
  )
)

