# BrAPI-Core
#
# The Breeding API (BrAPI) is a Standardized REST ful Web Service API Specification for communicating Plant Breeding Data. BrAPI allows for easy data sharing between databases and tools involved in plant breeding. <div class=\"brapi-section\"> <h2 class=\"brapi-section-title\">General Reference Documentation</h2> <div class=\"gen-info-link\"><a href=\"https://github.com/plantbreeding/API/blob/master/Specification/GeneralInfo/URL_Structure.md\">URL Structure</a></div> <div class=\"gen-info-link\"><a href=\"https://github.com/plantbreeding/API/blob/master/Specification/GeneralInfo/Response_Structure.md\">Response Structure</a></div> <div class=\"gen-info-link\"><a href=\"https://github.com/plantbreeding/API/blob/master/Specification/GeneralInfo/Date_Time_Encoding.md\">Date/Time Encoding</a></div> <div class=\"gen-info-link\"><a href=\"https://github.com/plantbreeding/API/blob/master/Specification/GeneralInfo/Location_Encoding.md\">Location Encoding</a></div> <div class=\"gen-info-link\"><a href=\"https://github.com/plantbreeding/API/blob/master/Specification/GeneralInfo/Error_Handling.md\">Error Handling</a></div> <div class=\"gen-info-link\"><a href=\"https://github.com/plantbreeding/API/blob/master/Specification/GeneralInfo/Search_Services.md\">Search Services</a></div> </div>  <div class=\"current-brapi-section brapi-section\"> <h2 class=\"brapi-section-title\">BrAPI Core</h2> <div class=\"brapi-section-description\">The BrAPI Core module contains high level entities used for organization and management. This includes Programs, Trials, Studies, Locations, People, and Lists</div> <div class=\"version-number\">V2.0</div> <div class=\"link-btn\"><a href=\"https://github.com/plantbreeding/API/tree/master/Specification/BrAPI-Core\">GitHub</a></div> <div class=\"link-btn\"><a href=\"https://app.swaggerhub.com/apis/PlantBreedingAPI/BrAPI-Core\">SwaggerHub</a></div> <div class=\"link-btn\"><a href=\"https://brapicore.docs.apiary.io\">Apiary</a></div> <div class=\"stop-float\"></div> </div>  <div class=\"brapi-section\"> <h2 class=\"brapi-section-title\">BrAPI Phenotyping</h2> <div class=\"brapi-section-description\">The BrAPI Phenotyping module contains entities related to phenotypic observations. This includes Observation Units, Observations, Observation Variables, Traits, Scales, Methods, and Images</div> <div class=\"version-number\">V2.0</div> <div class=\"link-btn\"><a href=\"https://github.com/plantbreeding/API/tree/master/Specification/BrAPI-Phenotyping\">GitHub</a></div> <div class=\"link-btn\"><a href=\"https://app.swaggerhub.com/apis/PlantBreedingAPI/BrAPI-Phenotyping\">SwaggerHub</a></div> <div class=\"link-btn\"><a href=\"https://brapiphenotyping.docs.apiary.io\">Apiary</a></div> <div class=\"stop-float\"></div> </div>  <div class=\"brapi-section\"> <h2 class=\"brapi-section-title\">BrAPI Genotyping</h2> <div class=\"brapi-section-description\">The BrAPI Genotyping module contains entities related to genotyping analysis. This includes Samples, Markers, Variant Sets, Variants, Call Sets, Calls, References, Reads, and Vendor Orders</div> <div class=\"version-number\">V2.0</div> <div class=\"link-btn\"><a href=\"https://github.com/plantbreeding/API/tree/master/Specification/BrAPI-Genotyping\">GitHub</a></div> <div class=\"link-btn\"><a href=\"https://app.swaggerhub.com/apis/PlantBreedingAPI/BrAPI-Genotyping\">SwaggerHub</a></div> <div class=\"link-btn\"><a href=\"https://brapigenotyping.docs.apiary.io\">Apiary</a></div> <div class=\"stop-float\"></div> </div>  <div class=\"brapi-section\"> <h2 class=\"brapi-section-title\">BrAPI Germplasm</h2> <div class=\"brapi-section-description\">The BrAPI Germplasm module contains entities related to germplasm management. This includes Germplasm, Germplasm Attributes, Seed Lots, Crosses, Pedigree, and Progeny</div> <div class=\"version-number\">V2.0</div> <div class=\"link-btn\"><a href=\"https://github.com/plantbreeding/API/tree/master/Specification/BrAPI-Germplasm\">GitHub</a></div> <div class=\"link-btn\"><a href=\"https://app.swaggerhub.com/apis/PlantBreedingAPI/BrAPI-Germplasm\">SwaggerHub</a></div> <div class=\"link-btn\"><a href=\"https://brapigermplasm.docs.apiary.io\">Apiary</a></div> <div class=\"stop-float\"></div> </div>  <style> .link-btn{ float: left;  margin: 2px 10px 0 0;  padding: 0 5px;  border-radius: 5px;  background-color: #ddd; } .stop-float{   clear: both; } .version-number{   float: left;    margin: 5px 10px 0 5px; } .brapi-section-title{   margin: 0 10px 0 0;   font-size: 20px; } .current-brapi-section{   font-weight: bolder;   border-radius: 5px;    background-color: #ddd; } .brapi-section{   padding: 5px 5px;  } .brapi-section-description{   margin: 5px 0 0 5px; } </style>
#
# The version of the OpenAPI document: 2.0
# 
# Generated by: https://openapi-generator.tech

#' @docType class
#' @title GermplasmAttributeSearchRequestAllOf
#'
#' @description GermplasmAttributeSearchRequestAllOf Class
#'
#' @format An \code{R6Class} generator object
#'
#' @field attributeDbIds  list( character ) [optional]
#'
#' @field attributeNames  list( character ) [optional]
#'
#' @field dataTypes  list( \link{TraitDataType} ) [optional]
#'
#' @field methodDbIds  list( character ) [optional]
#'
#' @field ontologyDbIds  list( character ) [optional]
#'
#' @field scaleDbIds  list( character ) [optional]
#'
#' @field studyDbId  list( character ) [optional]
#'
#' @field traitClasses  list( character ) [optional]
#'
#' @field traitDbIds  list( character ) [optional]
#'
#' @importFrom R6 R6Class
#' @importFrom jsonlite fromJSON toJSON
#' @export
GermplasmAttributeSearchRequestAllOf <- R6::R6Class(
  'GermplasmAttributeSearchRequestAllOf',
  public = list(
    `attributeDbIds` = NULL,
    `attributeNames` = NULL,
    `dataTypes` = NULL,
    `methodDbIds` = NULL,
    `ontologyDbIds` = NULL,
    `scaleDbIds` = NULL,
    `studyDbId` = NULL,
    `traitClasses` = NULL,
    `traitDbIds` = NULL,
    initialize = function(
        `attributeDbIds`=NULL, `attributeNames`=NULL, `dataTypes`=NULL, `methodDbIds`=NULL, `ontologyDbIds`=NULL, `scaleDbIds`=NULL, `studyDbId`=NULL, `traitClasses`=NULL, `traitDbIds`=NULL, ...
    ) {
      local.optional.var <- list(...)
      if (!is.null(`attributeDbIds`)) {
        stopifnot(is.vector(`attributeDbIds`), length(`attributeDbIds`) != 0)
        sapply(`attributeDbIds`, function(x) stopifnot(is.character(x)))
        self$`attributeDbIds` <- `attributeDbIds`
      }
      if (!is.null(`attributeNames`)) {
        stopifnot(is.vector(`attributeNames`), length(`attributeNames`) != 0)
        sapply(`attributeNames`, function(x) stopifnot(is.character(x)))
        self$`attributeNames` <- `attributeNames`
      }
      if (!is.null(`dataTypes`)) {
        stopifnot(is.vector(`dataTypes`), length(`dataTypes`) != 0)
        sapply(`dataTypes`, function(x) stopifnot(R6::is.R6(x)))
        self$`dataTypes` <- `dataTypes`
      }
      if (!is.null(`methodDbIds`)) {
        stopifnot(is.vector(`methodDbIds`), length(`methodDbIds`) != 0)
        sapply(`methodDbIds`, function(x) stopifnot(is.character(x)))
        self$`methodDbIds` <- `methodDbIds`
      }
      if (!is.null(`ontologyDbIds`)) {
        stopifnot(is.vector(`ontologyDbIds`), length(`ontologyDbIds`) != 0)
        sapply(`ontologyDbIds`, function(x) stopifnot(is.character(x)))
        self$`ontologyDbIds` <- `ontologyDbIds`
      }
      if (!is.null(`scaleDbIds`)) {
        stopifnot(is.vector(`scaleDbIds`), length(`scaleDbIds`) != 0)
        sapply(`scaleDbIds`, function(x) stopifnot(is.character(x)))
        self$`scaleDbIds` <- `scaleDbIds`
      }
      if (!is.null(`studyDbId`)) {
        stopifnot(is.vector(`studyDbId`), length(`studyDbId`) != 0)
        sapply(`studyDbId`, function(x) stopifnot(is.character(x)))
        self$`studyDbId` <- `studyDbId`
      }
      if (!is.null(`traitClasses`)) {
        stopifnot(is.vector(`traitClasses`), length(`traitClasses`) != 0)
        sapply(`traitClasses`, function(x) stopifnot(is.character(x)))
        self$`traitClasses` <- `traitClasses`
      }
      if (!is.null(`traitDbIds`)) {
        stopifnot(is.vector(`traitDbIds`), length(`traitDbIds`) != 0)
        sapply(`traitDbIds`, function(x) stopifnot(is.character(x)))
        self$`traitDbIds` <- `traitDbIds`
      }
    },
    toJSON = function() {
      GermplasmAttributeSearchRequestAllOfObject <- list()
      if (!is.null(self$`attributeDbIds`)) {
        GermplasmAttributeSearchRequestAllOfObject[['attributeDbIds']] <-
          self$`attributeDbIds`
      }
      if (!is.null(self$`attributeNames`)) {
        GermplasmAttributeSearchRequestAllOfObject[['attributeNames']] <-
          self$`attributeNames`
      }
      if (!is.null(self$`dataTypes`)) {
        GermplasmAttributeSearchRequestAllOfObject[['dataTypes']] <-
          lapply(self$`dataTypes`, function(x) x$toJSON())
      }
      if (!is.null(self$`methodDbIds`)) {
        GermplasmAttributeSearchRequestAllOfObject[['methodDbIds']] <-
          self$`methodDbIds`
      }
      if (!is.null(self$`ontologyDbIds`)) {
        GermplasmAttributeSearchRequestAllOfObject[['ontologyDbIds']] <-
          self$`ontologyDbIds`
      }
      if (!is.null(self$`scaleDbIds`)) {
        GermplasmAttributeSearchRequestAllOfObject[['scaleDbIds']] <-
          self$`scaleDbIds`
      }
      if (!is.null(self$`studyDbId`)) {
        GermplasmAttributeSearchRequestAllOfObject[['studyDbId']] <-
          self$`studyDbId`
      }
      if (!is.null(self$`traitClasses`)) {
        GermplasmAttributeSearchRequestAllOfObject[['traitClasses']] <-
          self$`traitClasses`
      }
      if (!is.null(self$`traitDbIds`)) {
        GermplasmAttributeSearchRequestAllOfObject[['traitDbIds']] <-
          self$`traitDbIds`
      }

      GermplasmAttributeSearchRequestAllOfObject
    },
    fromJSON = function(GermplasmAttributeSearchRequestAllOfJson) {
      GermplasmAttributeSearchRequestAllOfObject <- jsonlite::fromJSON(GermplasmAttributeSearchRequestAllOfJson)
      if (!is.null(GermplasmAttributeSearchRequestAllOfObject$`attributeDbIds`)) {
        self$`attributeDbIds` <- ApiClient$new()$deserializeObj(GermplasmAttributeSearchRequestAllOfObject$`attributeDbIds`, "array[character]", loadNamespace("openapi"))
      }
      if (!is.null(GermplasmAttributeSearchRequestAllOfObject$`attributeNames`)) {
        self$`attributeNames` <- ApiClient$new()$deserializeObj(GermplasmAttributeSearchRequestAllOfObject$`attributeNames`, "array[character]", loadNamespace("openapi"))
      }
      if (!is.null(GermplasmAttributeSearchRequestAllOfObject$`dataTypes`)) {
        self$`dataTypes` <- ApiClient$new()$deserializeObj(GermplasmAttributeSearchRequestAllOfObject$`dataTypes`, "array[TraitDataType]", loadNamespace("openapi"))
      }
      if (!is.null(GermplasmAttributeSearchRequestAllOfObject$`methodDbIds`)) {
        self$`methodDbIds` <- ApiClient$new()$deserializeObj(GermplasmAttributeSearchRequestAllOfObject$`methodDbIds`, "array[character]", loadNamespace("openapi"))
      }
      if (!is.null(GermplasmAttributeSearchRequestAllOfObject$`ontologyDbIds`)) {
        self$`ontologyDbIds` <- ApiClient$new()$deserializeObj(GermplasmAttributeSearchRequestAllOfObject$`ontologyDbIds`, "array[character]", loadNamespace("openapi"))
      }
      if (!is.null(GermplasmAttributeSearchRequestAllOfObject$`scaleDbIds`)) {
        self$`scaleDbIds` <- ApiClient$new()$deserializeObj(GermplasmAttributeSearchRequestAllOfObject$`scaleDbIds`, "array[character]", loadNamespace("openapi"))
      }
      if (!is.null(GermplasmAttributeSearchRequestAllOfObject$`studyDbId`)) {
        self$`studyDbId` <- ApiClient$new()$deserializeObj(GermplasmAttributeSearchRequestAllOfObject$`studyDbId`, "array[character]", loadNamespace("openapi"))
      }
      if (!is.null(GermplasmAttributeSearchRequestAllOfObject$`traitClasses`)) {
        self$`traitClasses` <- ApiClient$new()$deserializeObj(GermplasmAttributeSearchRequestAllOfObject$`traitClasses`, "array[character]", loadNamespace("openapi"))
      }
      if (!is.null(GermplasmAttributeSearchRequestAllOfObject$`traitDbIds`)) {
        self$`traitDbIds` <- ApiClient$new()$deserializeObj(GermplasmAttributeSearchRequestAllOfObject$`traitDbIds`, "array[character]", loadNamespace("openapi"))
      }
      self
    },
    toJSONString = function() {
      jsoncontent <- c(
        if (!is.null(self$`attributeDbIds`)) {
        sprintf(
        '"attributeDbIds":
           [%s]
        ',
        paste(unlist(lapply(self$`attributeDbIds`, function(x) paste0('"', x, '"'))), collapse=",")
        )},
        if (!is.null(self$`attributeNames`)) {
        sprintf(
        '"attributeNames":
           [%s]
        ',
        paste(unlist(lapply(self$`attributeNames`, function(x) paste0('"', x, '"'))), collapse=",")
        )},
        if (!is.null(self$`dataTypes`)) {
        sprintf(
        '"dataTypes":
        [%s]
',
        paste(sapply(self$`dataTypes`, function(x) jsonlite::toJSON(x$toJSON(), auto_unbox=TRUE, digits = NA)), collapse=",")
        )},
        if (!is.null(self$`methodDbIds`)) {
        sprintf(
        '"methodDbIds":
           [%s]
        ',
        paste(unlist(lapply(self$`methodDbIds`, function(x) paste0('"', x, '"'))), collapse=",")
        )},
        if (!is.null(self$`ontologyDbIds`)) {
        sprintf(
        '"ontologyDbIds":
           [%s]
        ',
        paste(unlist(lapply(self$`ontologyDbIds`, function(x) paste0('"', x, '"'))), collapse=",")
        )},
        if (!is.null(self$`scaleDbIds`)) {
        sprintf(
        '"scaleDbIds":
           [%s]
        ',
        paste(unlist(lapply(self$`scaleDbIds`, function(x) paste0('"', x, '"'))), collapse=",")
        )},
        if (!is.null(self$`studyDbId`)) {
        sprintf(
        '"studyDbId":
           [%s]
        ',
        paste(unlist(lapply(self$`studyDbId`, function(x) paste0('"', x, '"'))), collapse=",")
        )},
        if (!is.null(self$`traitClasses`)) {
        sprintf(
        '"traitClasses":
           [%s]
        ',
        paste(unlist(lapply(self$`traitClasses`, function(x) paste0('"', x, '"'))), collapse=",")
        )},
        if (!is.null(self$`traitDbIds`)) {
        sprintf(
        '"traitDbIds":
           [%s]
        ',
        paste(unlist(lapply(self$`traitDbIds`, function(x) paste0('"', x, '"'))), collapse=",")
        )}
      )
      jsoncontent <- paste(jsoncontent, collapse = ",")
      paste('{', jsoncontent, '}', sep = "")
    },
    fromJSONString = function(GermplasmAttributeSearchRequestAllOfJson) {
      GermplasmAttributeSearchRequestAllOfObject <- jsonlite::fromJSON(GermplasmAttributeSearchRequestAllOfJson)
      self$`attributeDbIds` <- ApiClient$new()$deserializeObj(GermplasmAttributeSearchRequestAllOfObject$`attributeDbIds`, "array[character]", loadNamespace("openapi"))
      self$`attributeNames` <- ApiClient$new()$deserializeObj(GermplasmAttributeSearchRequestAllOfObject$`attributeNames`, "array[character]", loadNamespace("openapi"))
      self$`dataTypes` <- ApiClient$new()$deserializeObj(GermplasmAttributeSearchRequestAllOfObject$`dataTypes`, "array[TraitDataType]", loadNamespace("openapi"))
      self$`methodDbIds` <- ApiClient$new()$deserializeObj(GermplasmAttributeSearchRequestAllOfObject$`methodDbIds`, "array[character]", loadNamespace("openapi"))
      self$`ontologyDbIds` <- ApiClient$new()$deserializeObj(GermplasmAttributeSearchRequestAllOfObject$`ontologyDbIds`, "array[character]", loadNamespace("openapi"))
      self$`scaleDbIds` <- ApiClient$new()$deserializeObj(GermplasmAttributeSearchRequestAllOfObject$`scaleDbIds`, "array[character]", loadNamespace("openapi"))
      self$`studyDbId` <- ApiClient$new()$deserializeObj(GermplasmAttributeSearchRequestAllOfObject$`studyDbId`, "array[character]", loadNamespace("openapi"))
      self$`traitClasses` <- ApiClient$new()$deserializeObj(GermplasmAttributeSearchRequestAllOfObject$`traitClasses`, "array[character]", loadNamespace("openapi"))
      self$`traitDbIds` <- ApiClient$new()$deserializeObj(GermplasmAttributeSearchRequestAllOfObject$`traitDbIds`, "array[character]", loadNamespace("openapi"))
      self
    }
  )
)

