# BrAPI-Core
#
# The Breeding API (BrAPI) is a Standardized REST ful Web Service API Specification for communicating Plant Breeding Data. BrAPI allows for easy data sharing between databases and tools involved in plant breeding. <div class=\"brapi-section\"> <h2 class=\"brapi-section-title\">General Reference Documentation</h2> <div class=\"gen-info-link\"><a href=\"https://github.com/plantbreeding/API/blob/master/Specification/GeneralInfo/URL_Structure.md\">URL Structure</a></div> <div class=\"gen-info-link\"><a href=\"https://github.com/plantbreeding/API/blob/master/Specification/GeneralInfo/Response_Structure.md\">Response Structure</a></div> <div class=\"gen-info-link\"><a href=\"https://github.com/plantbreeding/API/blob/master/Specification/GeneralInfo/Date_Time_Encoding.md\">Date/Time Encoding</a></div> <div class=\"gen-info-link\"><a href=\"https://github.com/plantbreeding/API/blob/master/Specification/GeneralInfo/Location_Encoding.md\">Location Encoding</a></div> <div class=\"gen-info-link\"><a href=\"https://github.com/plantbreeding/API/blob/master/Specification/GeneralInfo/Error_Handling.md\">Error Handling</a></div> <div class=\"gen-info-link\"><a href=\"https://github.com/plantbreeding/API/blob/master/Specification/GeneralInfo/Search_Services.md\">Search Services</a></div> </div>  <div class=\"current-brapi-section brapi-section\"> <h2 class=\"brapi-section-title\">BrAPI Core</h2> <div class=\"brapi-section-description\">The BrAPI Core module contains high level entities used for organization and management. This includes Programs, Trials, Studies, Locations, People, and Lists</div> <div class=\"version-number\">V2.0</div> <div class=\"link-btn\"><a href=\"https://github.com/plantbreeding/API/tree/master/Specification/BrAPI-Core\">GitHub</a></div> <div class=\"link-btn\"><a href=\"https://app.swaggerhub.com/apis/PlantBreedingAPI/BrAPI-Core\">SwaggerHub</a></div> <div class=\"link-btn\"><a href=\"https://brapicore.docs.apiary.io\">Apiary</a></div> <div class=\"stop-float\"></div> </div>  <div class=\"brapi-section\"> <h2 class=\"brapi-section-title\">BrAPI Phenotyping</h2> <div class=\"brapi-section-description\">The BrAPI Phenotyping module contains entities related to phenotypic observations. This includes Observation Units, Observations, Observation Variables, Traits, Scales, Methods, and Images</div> <div class=\"version-number\">V2.0</div> <div class=\"link-btn\"><a href=\"https://github.com/plantbreeding/API/tree/master/Specification/BrAPI-Phenotyping\">GitHub</a></div> <div class=\"link-btn\"><a href=\"https://app.swaggerhub.com/apis/PlantBreedingAPI/BrAPI-Phenotyping\">SwaggerHub</a></div> <div class=\"link-btn\"><a href=\"https://brapiphenotyping.docs.apiary.io\">Apiary</a></div> <div class=\"stop-float\"></div> </div>  <div class=\"brapi-section\"> <h2 class=\"brapi-section-title\">BrAPI Genotyping</h2> <div class=\"brapi-section-description\">The BrAPI Genotyping module contains entities related to genotyping analysis. This includes Samples, Markers, Variant Sets, Variants, Call Sets, Calls, References, Reads, and Vendor Orders</div> <div class=\"version-number\">V2.0</div> <div class=\"link-btn\"><a href=\"https://github.com/plantbreeding/API/tree/master/Specification/BrAPI-Genotyping\">GitHub</a></div> <div class=\"link-btn\"><a href=\"https://app.swaggerhub.com/apis/PlantBreedingAPI/BrAPI-Genotyping\">SwaggerHub</a></div> <div class=\"link-btn\"><a href=\"https://brapigenotyping.docs.apiary.io\">Apiary</a></div> <div class=\"stop-float\"></div> </div>  <div class=\"brapi-section\"> <h2 class=\"brapi-section-title\">BrAPI Germplasm</h2> <div class=\"brapi-section-description\">The BrAPI Germplasm module contains entities related to germplasm management. This includes Germplasm, Germplasm Attributes, Seed Lots, Crosses, Pedigree, and Progeny</div> <div class=\"version-number\">V2.0</div> <div class=\"link-btn\"><a href=\"https://github.com/plantbreeding/API/tree/master/Specification/BrAPI-Germplasm\">GitHub</a></div> <div class=\"link-btn\"><a href=\"https://app.swaggerhub.com/apis/PlantBreedingAPI/BrAPI-Germplasm\">SwaggerHub</a></div> <div class=\"link-btn\"><a href=\"https://brapigermplasm.docs.apiary.io\">Apiary</a></div> <div class=\"stop-float\"></div> </div>  <style> .link-btn{ float: left;  margin: 2px 10px 0 0;  padding: 0 5px;  border-radius: 5px;  background-color: #ddd; } .stop-float{   clear: both; } .version-number{   float: left;    margin: 5px 10px 0 5px; } .brapi-section-title{   margin: 0 10px 0 0;   font-size: 20px; } .current-brapi-section{   font-weight: bolder;   border-radius: 5px;    background-color: #ddd; } .brapi-section{   padding: 5px 5px;  } .brapi-section-description{   margin: 5px 0 0 5px; } </style>
#
# The version of the OpenAPI document: 2.0
# 
# Generated by: https://openapi-generator.tech

#' @docType class
#' @title Germplasm
#'
#' @description Germplasm Class
#'
#' @format An \code{R6Class} generator object
#'
#' @field accessionNumber  character [optional]
#'
#' @field acquisitionDate  character [optional]
#'
#' @field additionalInfo  named list( character ) [optional]
#'
#' @field biologicalStatusOfAccessionCode  character [optional]
#'
#' @field biologicalStatusOfAccessionDescription  character [optional]
#'
#' @field breedingMethodDbId  character [optional]
#'
#' @field collection  character [optional]
#'
#' @field commonCropName  character 
#'
#' @field countryOfOriginCode  character [optional]
#'
#' @field defaultDisplayName  character [optional]
#'
#' @field documentationURL  character [optional]
#'
#' @field donors  list( \link{GermplasmNewRequestDonors} ) [optional]
#'
#' @field externalReferences ExternalReferences list( object ) [optional]
#'
#' @field genus  character [optional]
#'
#' @field germplasmName  character 
#'
#' @field germplasmOrigin  list( \link{GermplasmOrigin} ) [optional]
#'
#' @field germplasmPUI  character 
#'
#' @field germplasmPreprocessing  character [optional]
#'
#' @field instituteCode  character [optional]
#'
#' @field instituteName  character [optional]
#'
#' @field pedigree  character [optional]
#'
#' @field seedSource  character [optional]
#'
#' @field seedSourceDescription  character [optional]
#'
#' @field species  character [optional]
#'
#' @field speciesAuthority  character [optional]
#'
#' @field storageTypes  list( \link{GermplasmNewRequestStorageTypes} ) [optional]
#'
#' @field subtaxa  character [optional]
#'
#' @field subtaxaAuthority  character [optional]
#'
#' @field synonyms  list( \link{GermplasmNewRequestSynonyms} ) [optional]
#'
#' @field taxonIds  list( \link{TaxonID} ) [optional]
#'
#' @field germplasmDbId  character 
#'
#' @importFrom R6 R6Class
#' @importFrom jsonlite fromJSON toJSON
#' @export
Germplasm <- R6::R6Class(
  'Germplasm',
  public = list(
    `accessionNumber` = NULL,
    `acquisitionDate` = NULL,
    `additionalInfo` = NULL,
    `biologicalStatusOfAccessionCode` = NULL,
    `biologicalStatusOfAccessionDescription` = NULL,
    `breedingMethodDbId` = NULL,
    `collection` = NULL,
    `commonCropName` = NULL,
    `countryOfOriginCode` = NULL,
    `defaultDisplayName` = NULL,
    `documentationURL` = NULL,
    `donors` = NULL,
    `externalReferences` = NULL,
    `genus` = NULL,
    `germplasmName` = NULL,
    `germplasmOrigin` = NULL,
    `germplasmPUI` = NULL,
    `germplasmPreprocessing` = NULL,
    `instituteCode` = NULL,
    `instituteName` = NULL,
    `pedigree` = NULL,
    `seedSource` = NULL,
    `seedSourceDescription` = NULL,
    `species` = NULL,
    `speciesAuthority` = NULL,
    `storageTypes` = NULL,
    `subtaxa` = NULL,
    `subtaxaAuthority` = NULL,
    `synonyms` = NULL,
    `taxonIds` = NULL,
    `germplasmDbId` = NULL,
    initialize = function(
        `commonCropName`, `germplasmName`, `germplasmPUI`, `germplasmDbId`, `accessionNumber`=NULL, `acquisitionDate`=NULL, `additionalInfo`=NULL, `biologicalStatusOfAccessionCode`=NULL, `biologicalStatusOfAccessionDescription`=NULL, `breedingMethodDbId`=NULL, `collection`=NULL, `countryOfOriginCode`=NULL, `defaultDisplayName`=NULL, `documentationURL`=NULL, `donors`=NULL, `externalReferences`=NULL, `genus`=NULL, `germplasmOrigin`=NULL, `germplasmPreprocessing`=NULL, `instituteCode`=NULL, `instituteName`=NULL, `pedigree`=NULL, `seedSource`=NULL, `seedSourceDescription`=NULL, `species`=NULL, `speciesAuthority`=NULL, `storageTypes`=NULL, `subtaxa`=NULL, `subtaxaAuthority`=NULL, `synonyms`=NULL, `taxonIds`=NULL, ...
    ) {
      local.optional.var <- list(...)
      if (!missing(`commonCropName`)) {
        stopifnot(is.character(`commonCropName`), length(`commonCropName`) == 1)
        self$`commonCropName` <- `commonCropName`
      }
      if (!missing(`germplasmName`)) {
        stopifnot(is.character(`germplasmName`), length(`germplasmName`) == 1)
        self$`germplasmName` <- `germplasmName`
      }
      if (!missing(`germplasmPUI`)) {
        stopifnot(is.character(`germplasmPUI`), length(`germplasmPUI`) == 1)
        self$`germplasmPUI` <- `germplasmPUI`
      }
      if (!missing(`germplasmDbId`)) {
        stopifnot(is.character(`germplasmDbId`), length(`germplasmDbId`) == 1)
        self$`germplasmDbId` <- `germplasmDbId`
      }
      if (!is.null(`accessionNumber`)) {
        stopifnot(is.character(`accessionNumber`), length(`accessionNumber`) == 1)
        self$`accessionNumber` <- `accessionNumber`
      }
      if (!is.null(`acquisitionDate`)) {
        stopifnot(is.character(`acquisitionDate`), length(`acquisitionDate`) == 1)
        self$`acquisitionDate` <- `acquisitionDate`
      }
      if (!is.null(`additionalInfo`)) {
        stopifnot(is.vector(`additionalInfo`), length(`additionalInfo`) != 0)
        sapply(`additionalInfo`, function(x) stopifnot(is.character(x)))
        self$`additionalInfo` <- `additionalInfo`
      }
      if (!is.null(`biologicalStatusOfAccessionCode`)) {
        stopifnot(is.character(`biologicalStatusOfAccessionCode`), length(`biologicalStatusOfAccessionCode`) == 1)
        self$`biologicalStatusOfAccessionCode` <- `biologicalStatusOfAccessionCode`
      }
      if (!is.null(`biologicalStatusOfAccessionDescription`)) {
        stopifnot(is.character(`biologicalStatusOfAccessionDescription`), length(`biologicalStatusOfAccessionDescription`) == 1)
        self$`biologicalStatusOfAccessionDescription` <- `biologicalStatusOfAccessionDescription`
      }
      if (!is.null(`breedingMethodDbId`)) {
        stopifnot(is.character(`breedingMethodDbId`), length(`breedingMethodDbId`) == 1)
        self$`breedingMethodDbId` <- `breedingMethodDbId`
      }
      if (!is.null(`collection`)) {
        stopifnot(is.character(`collection`), length(`collection`) == 1)
        self$`collection` <- `collection`
      }
      if (!is.null(`countryOfOriginCode`)) {
        stopifnot(is.character(`countryOfOriginCode`), length(`countryOfOriginCode`) == 1)
        self$`countryOfOriginCode` <- `countryOfOriginCode`
      }
      if (!is.null(`defaultDisplayName`)) {
        stopifnot(is.character(`defaultDisplayName`), length(`defaultDisplayName`) == 1)
        self$`defaultDisplayName` <- `defaultDisplayName`
      }
      if (!is.null(`documentationURL`)) {
        stopifnot(is.character(`documentationURL`), length(`documentationURL`) == 1)
        self$`documentationURL` <- `documentationURL`
      }
      if (!is.null(`donors`)) {
        stopifnot(is.vector(`donors`), length(`donors`) != 0)
        sapply(`donors`, function(x) stopifnot(R6::is.R6(x)))
        self$`donors` <- `donors`
      }
      if (!is.null(`externalReferences`)) {
        stopifnot(is.vector(`externalReferences`), length(`externalReferences`) != 0)
        sapply(`externalReferences`, function(x) stopifnot(is.character(x)))
        self$`externalReferences` <- `externalReferences`
      }
      if (!is.null(`genus`)) {
        stopifnot(is.character(`genus`), length(`genus`) == 1)
        self$`genus` <- `genus`
      }
      if (!is.null(`germplasmOrigin`)) {
        stopifnot(is.vector(`germplasmOrigin`), length(`germplasmOrigin`) != 0)
        sapply(`germplasmOrigin`, function(x) stopifnot(R6::is.R6(x)))
        self$`germplasmOrigin` <- `germplasmOrigin`
      }
      if (!is.null(`germplasmPreprocessing`)) {
        stopifnot(is.character(`germplasmPreprocessing`), length(`germplasmPreprocessing`) == 1)
        self$`germplasmPreprocessing` <- `germplasmPreprocessing`
      }
      if (!is.null(`instituteCode`)) {
        stopifnot(is.character(`instituteCode`), length(`instituteCode`) == 1)
        self$`instituteCode` <- `instituteCode`
      }
      if (!is.null(`instituteName`)) {
        stopifnot(is.character(`instituteName`), length(`instituteName`) == 1)
        self$`instituteName` <- `instituteName`
      }
      if (!is.null(`pedigree`)) {
        stopifnot(is.character(`pedigree`), length(`pedigree`) == 1)
        self$`pedigree` <- `pedigree`
      }
      if (!is.null(`seedSource`)) {
        stopifnot(is.character(`seedSource`), length(`seedSource`) == 1)
        self$`seedSource` <- `seedSource`
      }
      if (!is.null(`seedSourceDescription`)) {
        stopifnot(is.character(`seedSourceDescription`), length(`seedSourceDescription`) == 1)
        self$`seedSourceDescription` <- `seedSourceDescription`
      }
      if (!is.null(`species`)) {
        stopifnot(is.character(`species`), length(`species`) == 1)
        self$`species` <- `species`
      }
      if (!is.null(`speciesAuthority`)) {
        stopifnot(is.character(`speciesAuthority`), length(`speciesAuthority`) == 1)
        self$`speciesAuthority` <- `speciesAuthority`
      }
      if (!is.null(`storageTypes`)) {
        stopifnot(is.vector(`storageTypes`), length(`storageTypes`) != 0)
        sapply(`storageTypes`, function(x) stopifnot(R6::is.R6(x)))
        self$`storageTypes` <- `storageTypes`
      }
      if (!is.null(`subtaxa`)) {
        stopifnot(is.character(`subtaxa`), length(`subtaxa`) == 1)
        self$`subtaxa` <- `subtaxa`
      }
      if (!is.null(`subtaxaAuthority`)) {
        stopifnot(is.character(`subtaxaAuthority`), length(`subtaxaAuthority`) == 1)
        self$`subtaxaAuthority` <- `subtaxaAuthority`
      }
      if (!is.null(`synonyms`)) {
        stopifnot(is.vector(`synonyms`), length(`synonyms`) != 0)
        sapply(`synonyms`, function(x) stopifnot(R6::is.R6(x)))
        self$`synonyms` <- `synonyms`
      }
      if (!is.null(`taxonIds`)) {
        stopifnot(is.vector(`taxonIds`), length(`taxonIds`) != 0)
        sapply(`taxonIds`, function(x) stopifnot(R6::is.R6(x)))
        self$`taxonIds` <- `taxonIds`
      }
    },
    toJSON = function() {
      GermplasmObject <- list()
      if (!is.null(self$`accessionNumber`)) {
        GermplasmObject[['accessionNumber']] <-
          self$`accessionNumber`
      }
      if (!is.null(self$`acquisitionDate`)) {
        GermplasmObject[['acquisitionDate']] <-
          self$`acquisitionDate`
      }
      if (!is.null(self$`additionalInfo`)) {
        GermplasmObject[['additionalInfo']] <-
          self$`additionalInfo`
      }
      if (!is.null(self$`biologicalStatusOfAccessionCode`)) {
        GermplasmObject[['biologicalStatusOfAccessionCode']] <-
          self$`biologicalStatusOfAccessionCode`
      }
      if (!is.null(self$`biologicalStatusOfAccessionDescription`)) {
        GermplasmObject[['biologicalStatusOfAccessionDescription']] <-
          self$`biologicalStatusOfAccessionDescription`
      }
      if (!is.null(self$`breedingMethodDbId`)) {
        GermplasmObject[['breedingMethodDbId']] <-
          self$`breedingMethodDbId`
      }
      if (!is.null(self$`collection`)) {
        GermplasmObject[['collection']] <-
          self$`collection`
      }
      if (!is.null(self$`commonCropName`)) {
        GermplasmObject[['commonCropName']] <-
          self$`commonCropName`
      }
      if (!is.null(self$`countryOfOriginCode`)) {
        GermplasmObject[['countryOfOriginCode']] <-
          self$`countryOfOriginCode`
      }
      if (!is.null(self$`defaultDisplayName`)) {
        GermplasmObject[['defaultDisplayName']] <-
          self$`defaultDisplayName`
      }
      if (!is.null(self$`documentationURL`)) {
        GermplasmObject[['documentationURL']] <-
          self$`documentationURL`
      }
      if (!is.null(self$`donors`)) {
        GermplasmObject[['donors']] <-
          lapply(self$`donors`, function(x) x$toJSON())
      }
      if (!is.null(self$`externalReferences`)) {
        GermplasmObject[['externalReferences']] <-
          self$`externalReferences`
      }
      if (!is.null(self$`genus`)) {
        GermplasmObject[['genus']] <-
          self$`genus`
      }
      if (!is.null(self$`germplasmName`)) {
        GermplasmObject[['germplasmName']] <-
          self$`germplasmName`
      }
      if (!is.null(self$`germplasmOrigin`)) {
        GermplasmObject[['germplasmOrigin']] <-
          lapply(self$`germplasmOrigin`, function(x) x$toJSON())
      }
      if (!is.null(self$`germplasmPUI`)) {
        GermplasmObject[['germplasmPUI']] <-
          self$`germplasmPUI`
      }
      if (!is.null(self$`germplasmPreprocessing`)) {
        GermplasmObject[['germplasmPreprocessing']] <-
          self$`germplasmPreprocessing`
      }
      if (!is.null(self$`instituteCode`)) {
        GermplasmObject[['instituteCode']] <-
          self$`instituteCode`
      }
      if (!is.null(self$`instituteName`)) {
        GermplasmObject[['instituteName']] <-
          self$`instituteName`
      }
      if (!is.null(self$`pedigree`)) {
        GermplasmObject[['pedigree']] <-
          self$`pedigree`
      }
      if (!is.null(self$`seedSource`)) {
        GermplasmObject[['seedSource']] <-
          self$`seedSource`
      }
      if (!is.null(self$`seedSourceDescription`)) {
        GermplasmObject[['seedSourceDescription']] <-
          self$`seedSourceDescription`
      }
      if (!is.null(self$`species`)) {
        GermplasmObject[['species']] <-
          self$`species`
      }
      if (!is.null(self$`speciesAuthority`)) {
        GermplasmObject[['speciesAuthority']] <-
          self$`speciesAuthority`
      }
      if (!is.null(self$`storageTypes`)) {
        GermplasmObject[['storageTypes']] <-
          lapply(self$`storageTypes`, function(x) x$toJSON())
      }
      if (!is.null(self$`subtaxa`)) {
        GermplasmObject[['subtaxa']] <-
          self$`subtaxa`
      }
      if (!is.null(self$`subtaxaAuthority`)) {
        GermplasmObject[['subtaxaAuthority']] <-
          self$`subtaxaAuthority`
      }
      if (!is.null(self$`synonyms`)) {
        GermplasmObject[['synonyms']] <-
          lapply(self$`synonyms`, function(x) x$toJSON())
      }
      if (!is.null(self$`taxonIds`)) {
        GermplasmObject[['taxonIds']] <-
          lapply(self$`taxonIds`, function(x) x$toJSON())
      }
      if (!is.null(self$`germplasmDbId`)) {
        GermplasmObject[['germplasmDbId']] <-
          self$`germplasmDbId`
      }

      GermplasmObject
    },
    fromJSON = function(GermplasmJson) {
      GermplasmObject <- jsonlite::fromJSON(GermplasmJson)
      if (!is.null(GermplasmObject$`accessionNumber`)) {
        self$`accessionNumber` <- GermplasmObject$`accessionNumber`
      }
      if (!is.null(GermplasmObject$`acquisitionDate`)) {
        self$`acquisitionDate` <- GermplasmObject$`acquisitionDate`
      }
      if (!is.null(GermplasmObject$`additionalInfo`)) {
        self$`additionalInfo` <- ApiClient$new()$deserializeObj(GermplasmObject$`additionalInfo`, "map(character)", loadNamespace("openapi"))
      }
      if (!is.null(GermplasmObject$`biologicalStatusOfAccessionCode`)) {
        self$`biologicalStatusOfAccessionCode` <- GermplasmObject$`biologicalStatusOfAccessionCode`
      }
      if (!is.null(GermplasmObject$`biologicalStatusOfAccessionDescription`)) {
        self$`biologicalStatusOfAccessionDescription` <- GermplasmObject$`biologicalStatusOfAccessionDescription`
      }
      if (!is.null(GermplasmObject$`breedingMethodDbId`)) {
        self$`breedingMethodDbId` <- GermplasmObject$`breedingMethodDbId`
      }
      if (!is.null(GermplasmObject$`collection`)) {
        self$`collection` <- GermplasmObject$`collection`
      }
      if (!is.null(GermplasmObject$`commonCropName`)) {
        self$`commonCropName` <- GermplasmObject$`commonCropName`
      }
      if (!is.null(GermplasmObject$`countryOfOriginCode`)) {
        self$`countryOfOriginCode` <- GermplasmObject$`countryOfOriginCode`
      }
      if (!is.null(GermplasmObject$`defaultDisplayName`)) {
        self$`defaultDisplayName` <- GermplasmObject$`defaultDisplayName`
      }
      if (!is.null(GermplasmObject$`documentationURL`)) {
        self$`documentationURL` <- GermplasmObject$`documentationURL`
      }
      if (!is.null(GermplasmObject$`donors`)) {
        self$`donors` <- ApiClient$new()$deserializeObj(GermplasmObject$`donors`, "array[GermplasmNewRequestDonors]", loadNamespace("openapi"))
      }
      if (!is.null(GermplasmObject$`externalReferences`)) {
        self$`externalReferences` <- ApiClient$new()$deserializeObj(GermplasmObject$`externalReferences`, "array[object]", loadNamespace("openapi"))
      }
      if (!is.null(GermplasmObject$`genus`)) {
        self$`genus` <- GermplasmObject$`genus`
      }
      if (!is.null(GermplasmObject$`germplasmName`)) {
        self$`germplasmName` <- GermplasmObject$`germplasmName`
      }
      if (!is.null(GermplasmObject$`germplasmOrigin`)) {
        self$`germplasmOrigin` <- ApiClient$new()$deserializeObj(GermplasmObject$`germplasmOrigin`, "array[GermplasmOrigin]", loadNamespace("openapi"))
      }
      if (!is.null(GermplasmObject$`germplasmPUI`)) {
        self$`germplasmPUI` <- GermplasmObject$`germplasmPUI`
      }
      if (!is.null(GermplasmObject$`germplasmPreprocessing`)) {
        self$`germplasmPreprocessing` <- GermplasmObject$`germplasmPreprocessing`
      }
      if (!is.null(GermplasmObject$`instituteCode`)) {
        self$`instituteCode` <- GermplasmObject$`instituteCode`
      }
      if (!is.null(GermplasmObject$`instituteName`)) {
        self$`instituteName` <- GermplasmObject$`instituteName`
      }
      if (!is.null(GermplasmObject$`pedigree`)) {
        self$`pedigree` <- GermplasmObject$`pedigree`
      }
      if (!is.null(GermplasmObject$`seedSource`)) {
        self$`seedSource` <- GermplasmObject$`seedSource`
      }
      if (!is.null(GermplasmObject$`seedSourceDescription`)) {
        self$`seedSourceDescription` <- GermplasmObject$`seedSourceDescription`
      }
      if (!is.null(GermplasmObject$`species`)) {
        self$`species` <- GermplasmObject$`species`
      }
      if (!is.null(GermplasmObject$`speciesAuthority`)) {
        self$`speciesAuthority` <- GermplasmObject$`speciesAuthority`
      }
      if (!is.null(GermplasmObject$`storageTypes`)) {
        self$`storageTypes` <- ApiClient$new()$deserializeObj(GermplasmObject$`storageTypes`, "array[GermplasmNewRequestStorageTypes]", loadNamespace("openapi"))
      }
      if (!is.null(GermplasmObject$`subtaxa`)) {
        self$`subtaxa` <- GermplasmObject$`subtaxa`
      }
      if (!is.null(GermplasmObject$`subtaxaAuthority`)) {
        self$`subtaxaAuthority` <- GermplasmObject$`subtaxaAuthority`
      }
      if (!is.null(GermplasmObject$`synonyms`)) {
        self$`synonyms` <- ApiClient$new()$deserializeObj(GermplasmObject$`synonyms`, "array[GermplasmNewRequestSynonyms]", loadNamespace("openapi"))
      }
      if (!is.null(GermplasmObject$`taxonIds`)) {
        self$`taxonIds` <- ApiClient$new()$deserializeObj(GermplasmObject$`taxonIds`, "array[TaxonID]", loadNamespace("openapi"))
      }
      if (!is.null(GermplasmObject$`germplasmDbId`)) {
        self$`germplasmDbId` <- GermplasmObject$`germplasmDbId`
      }
      self
    },
    toJSONString = function() {
      jsoncontent <- c(
        if (!is.null(self$`accessionNumber`)) {
        sprintf(
        '"accessionNumber":
          "%s"
                ',
        self$`accessionNumber`
        )},
        if (!is.null(self$`acquisitionDate`)) {
        sprintf(
        '"acquisitionDate":
          "%s"
                ',
        self$`acquisitionDate`
        )},
        if (!is.null(self$`additionalInfo`)) {
        sprintf(
        '"additionalInfo":
          "%s"
        ',
        jsonlite::toJSON(lapply(self$`additionalInfo`, function(x){ x }), auto_unbox = TRUE, digits=NA)
        )},
        if (!is.null(self$`biologicalStatusOfAccessionCode`)) {
        sprintf(
        '"biologicalStatusOfAccessionCode":
          "%s"
                ',
        self$`biologicalStatusOfAccessionCode`
        )},
        if (!is.null(self$`biologicalStatusOfAccessionDescription`)) {
        sprintf(
        '"biologicalStatusOfAccessionDescription":
          "%s"
                ',
        self$`biologicalStatusOfAccessionDescription`
        )},
        if (!is.null(self$`breedingMethodDbId`)) {
        sprintf(
        '"breedingMethodDbId":
          "%s"
                ',
        self$`breedingMethodDbId`
        )},
        if (!is.null(self$`collection`)) {
        sprintf(
        '"collection":
          "%s"
                ',
        self$`collection`
        )},
        if (!is.null(self$`commonCropName`)) {
        sprintf(
        '"commonCropName":
          "%s"
                ',
        self$`commonCropName`
        )},
        if (!is.null(self$`countryOfOriginCode`)) {
        sprintf(
        '"countryOfOriginCode":
          "%s"
                ',
        self$`countryOfOriginCode`
        )},
        if (!is.null(self$`defaultDisplayName`)) {
        sprintf(
        '"defaultDisplayName":
          "%s"
                ',
        self$`defaultDisplayName`
        )},
        if (!is.null(self$`documentationURL`)) {
        sprintf(
        '"documentationURL":
          "%s"
                ',
        self$`documentationURL`
        )},
        if (!is.null(self$`donors`)) {
        sprintf(
        '"donors":
        [%s]
',
        paste(sapply(self$`donors`, function(x) jsonlite::toJSON(x$toJSON(), auto_unbox=TRUE, digits = NA)), collapse=",")
        )},
        if (!is.null(self$`externalReferences`)) {
        sprintf(
        '"externalReferences":
           [%s]
        ',
        paste(unlist(lapply(self$`externalReferences`, function(x) paste0('"', x, '"'))), collapse=",")
        )},
        if (!is.null(self$`genus`)) {
        sprintf(
        '"genus":
          "%s"
                ',
        self$`genus`
        )},
        if (!is.null(self$`germplasmName`)) {
        sprintf(
        '"germplasmName":
          "%s"
                ',
        self$`germplasmName`
        )},
        if (!is.null(self$`germplasmOrigin`)) {
        sprintf(
        '"germplasmOrigin":
        [%s]
',
        paste(sapply(self$`germplasmOrigin`, function(x) jsonlite::toJSON(x$toJSON(), auto_unbox=TRUE, digits = NA)), collapse=",")
        )},
        if (!is.null(self$`germplasmPUI`)) {
        sprintf(
        '"germplasmPUI":
          "%s"
                ',
        self$`germplasmPUI`
        )},
        if (!is.null(self$`germplasmPreprocessing`)) {
        sprintf(
        '"germplasmPreprocessing":
          "%s"
                ',
        self$`germplasmPreprocessing`
        )},
        if (!is.null(self$`instituteCode`)) {
        sprintf(
        '"instituteCode":
          "%s"
                ',
        self$`instituteCode`
        )},
        if (!is.null(self$`instituteName`)) {
        sprintf(
        '"instituteName":
          "%s"
                ',
        self$`instituteName`
        )},
        if (!is.null(self$`pedigree`)) {
        sprintf(
        '"pedigree":
          "%s"
                ',
        self$`pedigree`
        )},
        if (!is.null(self$`seedSource`)) {
        sprintf(
        '"seedSource":
          "%s"
                ',
        self$`seedSource`
        )},
        if (!is.null(self$`seedSourceDescription`)) {
        sprintf(
        '"seedSourceDescription":
          "%s"
                ',
        self$`seedSourceDescription`
        )},
        if (!is.null(self$`species`)) {
        sprintf(
        '"species":
          "%s"
                ',
        self$`species`
        )},
        if (!is.null(self$`speciesAuthority`)) {
        sprintf(
        '"speciesAuthority":
          "%s"
                ',
        self$`speciesAuthority`
        )},
        if (!is.null(self$`storageTypes`)) {
        sprintf(
        '"storageTypes":
        [%s]
',
        paste(sapply(self$`storageTypes`, function(x) jsonlite::toJSON(x$toJSON(), auto_unbox=TRUE, digits = NA)), collapse=",")
        )},
        if (!is.null(self$`subtaxa`)) {
        sprintf(
        '"subtaxa":
          "%s"
                ',
        self$`subtaxa`
        )},
        if (!is.null(self$`subtaxaAuthority`)) {
        sprintf(
        '"subtaxaAuthority":
          "%s"
                ',
        self$`subtaxaAuthority`
        )},
        if (!is.null(self$`synonyms`)) {
        sprintf(
        '"synonyms":
        [%s]
',
        paste(sapply(self$`synonyms`, function(x) jsonlite::toJSON(x$toJSON(), auto_unbox=TRUE, digits = NA)), collapse=",")
        )},
        if (!is.null(self$`taxonIds`)) {
        sprintf(
        '"taxonIds":
        [%s]
',
        paste(sapply(self$`taxonIds`, function(x) jsonlite::toJSON(x$toJSON(), auto_unbox=TRUE, digits = NA)), collapse=",")
        )},
        if (!is.null(self$`germplasmDbId`)) {
        sprintf(
        '"germplasmDbId":
          "%s"
                ',
        self$`germplasmDbId`
        )}
      )
      jsoncontent <- paste(jsoncontent, collapse = ",")
      paste('{', jsoncontent, '}', sep = "")
    },
    fromJSONString = function(GermplasmJson) {
      GermplasmObject <- jsonlite::fromJSON(GermplasmJson)
      self$`accessionNumber` <- GermplasmObject$`accessionNumber`
      self$`acquisitionDate` <- GermplasmObject$`acquisitionDate`
      self$`additionalInfo` <- ApiClient$new()$deserializeObj(GermplasmObject$`additionalInfo`, "map(character)", loadNamespace("openapi"))
      self$`biologicalStatusOfAccessionCode` <- GermplasmObject$`biologicalStatusOfAccessionCode`
      self$`biologicalStatusOfAccessionDescription` <- GermplasmObject$`biologicalStatusOfAccessionDescription`
      self$`breedingMethodDbId` <- GermplasmObject$`breedingMethodDbId`
      self$`collection` <- GermplasmObject$`collection`
      self$`commonCropName` <- GermplasmObject$`commonCropName`
      self$`countryOfOriginCode` <- GermplasmObject$`countryOfOriginCode`
      self$`defaultDisplayName` <- GermplasmObject$`defaultDisplayName`
      self$`documentationURL` <- GermplasmObject$`documentationURL`
      self$`donors` <- ApiClient$new()$deserializeObj(GermplasmObject$`donors`, "array[GermplasmNewRequestDonors]", loadNamespace("openapi"))
      self$`externalReferences` <- ApiClient$new()$deserializeObj(GermplasmObject$`externalReferences`, "array[object]", loadNamespace("openapi"))
      self$`genus` <- GermplasmObject$`genus`
      self$`germplasmName` <- GermplasmObject$`germplasmName`
      self$`germplasmOrigin` <- ApiClient$new()$deserializeObj(GermplasmObject$`germplasmOrigin`, "array[GermplasmOrigin]", loadNamespace("openapi"))
      self$`germplasmPUI` <- GermplasmObject$`germplasmPUI`
      self$`germplasmPreprocessing` <- GermplasmObject$`germplasmPreprocessing`
      self$`instituteCode` <- GermplasmObject$`instituteCode`
      self$`instituteName` <- GermplasmObject$`instituteName`
      self$`pedigree` <- GermplasmObject$`pedigree`
      self$`seedSource` <- GermplasmObject$`seedSource`
      self$`seedSourceDescription` <- GermplasmObject$`seedSourceDescription`
      self$`species` <- GermplasmObject$`species`
      self$`speciesAuthority` <- GermplasmObject$`speciesAuthority`
      self$`storageTypes` <- ApiClient$new()$deserializeObj(GermplasmObject$`storageTypes`, "array[GermplasmNewRequestStorageTypes]", loadNamespace("openapi"))
      self$`subtaxa` <- GermplasmObject$`subtaxa`
      self$`subtaxaAuthority` <- GermplasmObject$`subtaxaAuthority`
      self$`synonyms` <- ApiClient$new()$deserializeObj(GermplasmObject$`synonyms`, "array[GermplasmNewRequestSynonyms]", loadNamespace("openapi"))
      self$`taxonIds` <- ApiClient$new()$deserializeObj(GermplasmObject$`taxonIds`, "array[TaxonID]", loadNamespace("openapi"))
      self$`germplasmDbId` <- GermplasmObject$`germplasmDbId`
      self
    }
  )
)

