# BrAPI-Core
#
# The Breeding API (BrAPI) is a Standardized REST ful Web Service API Specification for communicating Plant Breeding Data. BrAPI allows for easy data sharing between databases and tools involved in plant breeding. <div class=\"brapi-section\"> <h2 class=\"brapi-section-title\">General Reference Documentation</h2> <div class=\"gen-info-link\"><a href=\"https://github.com/plantbreeding/API/blob/master/Specification/GeneralInfo/URL_Structure.md\">URL Structure</a></div> <div class=\"gen-info-link\"><a href=\"https://github.com/plantbreeding/API/blob/master/Specification/GeneralInfo/Response_Structure.md\">Response Structure</a></div> <div class=\"gen-info-link\"><a href=\"https://github.com/plantbreeding/API/blob/master/Specification/GeneralInfo/Date_Time_Encoding.md\">Date/Time Encoding</a></div> <div class=\"gen-info-link\"><a href=\"https://github.com/plantbreeding/API/blob/master/Specification/GeneralInfo/Location_Encoding.md\">Location Encoding</a></div> <div class=\"gen-info-link\"><a href=\"https://github.com/plantbreeding/API/blob/master/Specification/GeneralInfo/Error_Handling.md\">Error Handling</a></div> <div class=\"gen-info-link\"><a href=\"https://github.com/plantbreeding/API/blob/master/Specification/GeneralInfo/Search_Services.md\">Search Services</a></div> </div>  <div class=\"current-brapi-section brapi-section\"> <h2 class=\"brapi-section-title\">BrAPI Core</h2> <div class=\"brapi-section-description\">The BrAPI Core module contains high level entities used for organization and management. This includes Programs, Trials, Studies, Locations, People, and Lists</div> <div class=\"version-number\">V2.0</div> <div class=\"link-btn\"><a href=\"https://github.com/plantbreeding/API/tree/master/Specification/BrAPI-Core\">GitHub</a></div> <div class=\"link-btn\"><a href=\"https://app.swaggerhub.com/apis/PlantBreedingAPI/BrAPI-Core\">SwaggerHub</a></div> <div class=\"link-btn\"><a href=\"https://brapicore.docs.apiary.io\">Apiary</a></div> <div class=\"stop-float\"></div> </div>  <div class=\"brapi-section\"> <h2 class=\"brapi-section-title\">BrAPI Phenotyping</h2> <div class=\"brapi-section-description\">The BrAPI Phenotyping module contains entities related to phenotypic observations. This includes Observation Units, Observations, Observation Variables, Traits, Scales, Methods, and Images</div> <div class=\"version-number\">V2.0</div> <div class=\"link-btn\"><a href=\"https://github.com/plantbreeding/API/tree/master/Specification/BrAPI-Phenotyping\">GitHub</a></div> <div class=\"link-btn\"><a href=\"https://app.swaggerhub.com/apis/PlantBreedingAPI/BrAPI-Phenotyping\">SwaggerHub</a></div> <div class=\"link-btn\"><a href=\"https://brapiphenotyping.docs.apiary.io\">Apiary</a></div> <div class=\"stop-float\"></div> </div>  <div class=\"brapi-section\"> <h2 class=\"brapi-section-title\">BrAPI Genotyping</h2> <div class=\"brapi-section-description\">The BrAPI Genotyping module contains entities related to genotyping analysis. This includes Samples, Markers, Variant Sets, Variants, Call Sets, Calls, References, Reads, and Vendor Orders</div> <div class=\"version-number\">V2.0</div> <div class=\"link-btn\"><a href=\"https://github.com/plantbreeding/API/tree/master/Specification/BrAPI-Genotyping\">GitHub</a></div> <div class=\"link-btn\"><a href=\"https://app.swaggerhub.com/apis/PlantBreedingAPI/BrAPI-Genotyping\">SwaggerHub</a></div> <div class=\"link-btn\"><a href=\"https://brapigenotyping.docs.apiary.io\">Apiary</a></div> <div class=\"stop-float\"></div> </div>  <div class=\"brapi-section\"> <h2 class=\"brapi-section-title\">BrAPI Germplasm</h2> <div class=\"brapi-section-description\">The BrAPI Germplasm module contains entities related to germplasm management. This includes Germplasm, Germplasm Attributes, Seed Lots, Crosses, Pedigree, and Progeny</div> <div class=\"version-number\">V2.0</div> <div class=\"link-btn\"><a href=\"https://github.com/plantbreeding/API/tree/master/Specification/BrAPI-Germplasm\">GitHub</a></div> <div class=\"link-btn\"><a href=\"https://app.swaggerhub.com/apis/PlantBreedingAPI/BrAPI-Germplasm\">SwaggerHub</a></div> <div class=\"link-btn\"><a href=\"https://brapigermplasm.docs.apiary.io\">Apiary</a></div> <div class=\"stop-float\"></div> </div>  <style> .link-btn{ float: left;  margin: 2px 10px 0 0;  padding: 0 5px;  border-radius: 5px;  background-color: #ddd; } .stop-float{   clear: both; } .version-number{   float: left;    margin: 5px 10px 0 5px; } .brapi-section-title{   margin: 0 10px 0 0;   font-size: 20px; } .current-brapi-section{   font-weight: bolder;   border-radius: 5px;    background-color: #ddd; } .brapi-section{   padding: 5px 5px;  } .brapi-section-description{   margin: 5px 0 0 5px; } </style>
#
# The version of the OpenAPI document: 2.0
# 
# Generated by: https://openapi-generator.tech

#' @docType class
#' @title Image
#'
#' @description Image Class
#'
#' @format An \code{R6Class} generator object
#'
#' @field additionalInfo  named list( character ) [optional]
#'
#' @field copyright  character [optional]
#'
#' @field description  character [optional]
#'
#' @field descriptiveOntologyTerms  list( character ) [optional]
#'
#' @field externalReferences ExternalReferences list( object ) [optional]
#'
#' @field imageFileName  character [optional]
#'
#' @field imageFileSize  integer [optional]
#'
#' @field imageHeight  integer [optional]
#'
#' @field imageLocation  \link{GeoJSON} [optional]
#'
#' @field imageName  character [optional]
#'
#' @field imageTimeStamp  character [optional]
#'
#' @field imageURL  character [optional]
#'
#' @field imageWidth  integer [optional]
#'
#' @field mimeType  character [optional]
#'
#' @field observationDbIds  list( character ) [optional]
#'
#' @field observationUnitDbId  character [optional]
#'
#' @field imageDbId  character [optional]
#'
#' @importFrom R6 R6Class
#' @importFrom jsonlite fromJSON toJSON
#' @export
Image <- R6::R6Class(
  'Image',
  public = list(
    `additionalInfo` = NULL,
    `copyright` = NULL,
    `description` = NULL,
    `descriptiveOntologyTerms` = NULL,
    `externalReferences` = NULL,
    `imageFileName` = NULL,
    `imageFileSize` = NULL,
    `imageHeight` = NULL,
    `imageLocation` = NULL,
    `imageName` = NULL,
    `imageTimeStamp` = NULL,
    `imageURL` = NULL,
    `imageWidth` = NULL,
    `mimeType` = NULL,
    `observationDbIds` = NULL,
    `observationUnitDbId` = NULL,
    `imageDbId` = NULL,
    initialize = function(
        `additionalInfo`=NULL, `copyright`=NULL, `description`=NULL, `descriptiveOntologyTerms`=NULL, `externalReferences`=NULL, `imageFileName`=NULL, `imageFileSize`=NULL, `imageHeight`=NULL, `imageLocation`=NULL, `imageName`=NULL, `imageTimeStamp`=NULL, `imageURL`=NULL, `imageWidth`=NULL, `mimeType`=NULL, `observationDbIds`=NULL, `observationUnitDbId`=NULL, `imageDbId`=NULL, ...
    ) {
      local.optional.var <- list(...)
      if (!is.null(`additionalInfo`)) {
        stopifnot(is.vector(`additionalInfo`), length(`additionalInfo`) != 0)
        sapply(`additionalInfo`, function(x) stopifnot(is.character(x)))
        self$`additionalInfo` <- `additionalInfo`
      }
      if (!is.null(`copyright`)) {
        stopifnot(is.character(`copyright`), length(`copyright`) == 1)
        self$`copyright` <- `copyright`
      }
      if (!is.null(`description`)) {
        stopifnot(is.character(`description`), length(`description`) == 1)
        self$`description` <- `description`
      }
      if (!is.null(`descriptiveOntologyTerms`)) {
        stopifnot(is.vector(`descriptiveOntologyTerms`), length(`descriptiveOntologyTerms`) != 0)
        sapply(`descriptiveOntologyTerms`, function(x) stopifnot(is.character(x)))
        self$`descriptiveOntologyTerms` <- `descriptiveOntologyTerms`
      }
      if (!is.null(`externalReferences`)) {
        stopifnot(is.vector(`externalReferences`), length(`externalReferences`) != 0)
        sapply(`externalReferences`, function(x) stopifnot(is.character(x)))
        self$`externalReferences` <- `externalReferences`
      }
      if (!is.null(`imageFileName`)) {
        stopifnot(is.character(`imageFileName`), length(`imageFileName`) == 1)
        self$`imageFileName` <- `imageFileName`
      }
      if (!is.null(`imageFileSize`)) {
        stopifnot(is.numeric(`imageFileSize`), length(`imageFileSize`) == 1)
        self$`imageFileSize` <- `imageFileSize`
      }
      if (!is.null(`imageHeight`)) {
        stopifnot(is.numeric(`imageHeight`), length(`imageHeight`) == 1)
        self$`imageHeight` <- `imageHeight`
      }
      if (!is.null(`imageLocation`)) {
        stopifnot(R6::is.R6(`imageLocation`))
        self$`imageLocation` <- `imageLocation`
      }
      if (!is.null(`imageName`)) {
        stopifnot(is.character(`imageName`), length(`imageName`) == 1)
        self$`imageName` <- `imageName`
      }
      if (!is.null(`imageTimeStamp`)) {
        stopifnot(is.character(`imageTimeStamp`), length(`imageTimeStamp`) == 1)
        self$`imageTimeStamp` <- `imageTimeStamp`
      }
      if (!is.null(`imageURL`)) {
        stopifnot(is.character(`imageURL`), length(`imageURL`) == 1)
        self$`imageURL` <- `imageURL`
      }
      if (!is.null(`imageWidth`)) {
        stopifnot(is.numeric(`imageWidth`), length(`imageWidth`) == 1)
        self$`imageWidth` <- `imageWidth`
      }
      if (!is.null(`mimeType`)) {
        stopifnot(is.character(`mimeType`), length(`mimeType`) == 1)
        self$`mimeType` <- `mimeType`
      }
      if (!is.null(`observationDbIds`)) {
        stopifnot(is.vector(`observationDbIds`), length(`observationDbIds`) != 0)
        sapply(`observationDbIds`, function(x) stopifnot(is.character(x)))
        self$`observationDbIds` <- `observationDbIds`
      }
      if (!is.null(`observationUnitDbId`)) {
        stopifnot(is.character(`observationUnitDbId`), length(`observationUnitDbId`) == 1)
        self$`observationUnitDbId` <- `observationUnitDbId`
      }
      if (!is.null(`imageDbId`)) {
        stopifnot(is.character(`imageDbId`), length(`imageDbId`) == 1)
        self$`imageDbId` <- `imageDbId`
      }
    },
    toJSON = function() {
      ImageObject <- list()
      if (!is.null(self$`additionalInfo`)) {
        ImageObject[['additionalInfo']] <-
          self$`additionalInfo`
      }
      if (!is.null(self$`copyright`)) {
        ImageObject[['copyright']] <-
          self$`copyright`
      }
      if (!is.null(self$`description`)) {
        ImageObject[['description']] <-
          self$`description`
      }
      if (!is.null(self$`descriptiveOntologyTerms`)) {
        ImageObject[['descriptiveOntologyTerms']] <-
          self$`descriptiveOntologyTerms`
      }
      if (!is.null(self$`externalReferences`)) {
        ImageObject[['externalReferences']] <-
          self$`externalReferences`
      }
      if (!is.null(self$`imageFileName`)) {
        ImageObject[['imageFileName']] <-
          self$`imageFileName`
      }
      if (!is.null(self$`imageFileSize`)) {
        ImageObject[['imageFileSize']] <-
          self$`imageFileSize`
      }
      if (!is.null(self$`imageHeight`)) {
        ImageObject[['imageHeight']] <-
          self$`imageHeight`
      }
      if (!is.null(self$`imageLocation`)) {
        ImageObject[['imageLocation']] <-
          self$`imageLocation`$toJSON()
      }
      if (!is.null(self$`imageName`)) {
        ImageObject[['imageName']] <-
          self$`imageName`
      }
      if (!is.null(self$`imageTimeStamp`)) {
        ImageObject[['imageTimeStamp']] <-
          self$`imageTimeStamp`
      }
      if (!is.null(self$`imageURL`)) {
        ImageObject[['imageURL']] <-
          self$`imageURL`
      }
      if (!is.null(self$`imageWidth`)) {
        ImageObject[['imageWidth']] <-
          self$`imageWidth`
      }
      if (!is.null(self$`mimeType`)) {
        ImageObject[['mimeType']] <-
          self$`mimeType`
      }
      if (!is.null(self$`observationDbIds`)) {
        ImageObject[['observationDbIds']] <-
          self$`observationDbIds`
      }
      if (!is.null(self$`observationUnitDbId`)) {
        ImageObject[['observationUnitDbId']] <-
          self$`observationUnitDbId`
      }
      if (!is.null(self$`imageDbId`)) {
        ImageObject[['imageDbId']] <-
          self$`imageDbId`
      }

      ImageObject
    },
    fromJSON = function(ImageJson) {
      ImageObject <- jsonlite::fromJSON(ImageJson)
      if (!is.null(ImageObject$`additionalInfo`)) {
        self$`additionalInfo` <- ApiClient$new()$deserializeObj(ImageObject$`additionalInfo`, "map(character)", loadNamespace("openapi"))
      }
      if (!is.null(ImageObject$`copyright`)) {
        self$`copyright` <- ImageObject$`copyright`
      }
      if (!is.null(ImageObject$`description`)) {
        self$`description` <- ImageObject$`description`
      }
      if (!is.null(ImageObject$`descriptiveOntologyTerms`)) {
        self$`descriptiveOntologyTerms` <- ApiClient$new()$deserializeObj(ImageObject$`descriptiveOntologyTerms`, "array[character]", loadNamespace("openapi"))
      }
      if (!is.null(ImageObject$`externalReferences`)) {
        self$`externalReferences` <- ApiClient$new()$deserializeObj(ImageObject$`externalReferences`, "array[object]", loadNamespace("openapi"))
      }
      if (!is.null(ImageObject$`imageFileName`)) {
        self$`imageFileName` <- ImageObject$`imageFileName`
      }
      if (!is.null(ImageObject$`imageFileSize`)) {
        self$`imageFileSize` <- ImageObject$`imageFileSize`
      }
      if (!is.null(ImageObject$`imageHeight`)) {
        self$`imageHeight` <- ImageObject$`imageHeight`
      }
      if (!is.null(ImageObject$`imageLocation`)) {
        imageLocationObject <- GeoJSON$new()
        imageLocationObject$fromJSON(jsonlite::toJSON(ImageObject$imageLocation, auto_unbox = TRUE, digits = NA))
        self$`imageLocation` <- imageLocationObject
      }
      if (!is.null(ImageObject$`imageName`)) {
        self$`imageName` <- ImageObject$`imageName`
      }
      if (!is.null(ImageObject$`imageTimeStamp`)) {
        self$`imageTimeStamp` <- ImageObject$`imageTimeStamp`
      }
      if (!is.null(ImageObject$`imageURL`)) {
        self$`imageURL` <- ImageObject$`imageURL`
      }
      if (!is.null(ImageObject$`imageWidth`)) {
        self$`imageWidth` <- ImageObject$`imageWidth`
      }
      if (!is.null(ImageObject$`mimeType`)) {
        self$`mimeType` <- ImageObject$`mimeType`
      }
      if (!is.null(ImageObject$`observationDbIds`)) {
        self$`observationDbIds` <- ApiClient$new()$deserializeObj(ImageObject$`observationDbIds`, "array[character]", loadNamespace("openapi"))
      }
      if (!is.null(ImageObject$`observationUnitDbId`)) {
        self$`observationUnitDbId` <- ImageObject$`observationUnitDbId`
      }
      if (!is.null(ImageObject$`imageDbId`)) {
        self$`imageDbId` <- ImageObject$`imageDbId`
      }
      self
    },
    toJSONString = function() {
      jsoncontent <- c(
        if (!is.null(self$`additionalInfo`)) {
        sprintf(
        '"additionalInfo":
          "%s"
        ',
        jsonlite::toJSON(lapply(self$`additionalInfo`, function(x){ x }), auto_unbox = TRUE, digits=NA)
        )},
        if (!is.null(self$`copyright`)) {
        sprintf(
        '"copyright":
          "%s"
                ',
        self$`copyright`
        )},
        if (!is.null(self$`description`)) {
        sprintf(
        '"description":
          "%s"
                ',
        self$`description`
        )},
        if (!is.null(self$`descriptiveOntologyTerms`)) {
        sprintf(
        '"descriptiveOntologyTerms":
           [%s]
        ',
        paste(unlist(lapply(self$`descriptiveOntologyTerms`, function(x) paste0('"', x, '"'))), collapse=",")
        )},
        if (!is.null(self$`externalReferences`)) {
        sprintf(
        '"externalReferences":
           [%s]
        ',
        paste(unlist(lapply(self$`externalReferences`, function(x) paste0('"', x, '"'))), collapse=",")
        )},
        if (!is.null(self$`imageFileName`)) {
        sprintf(
        '"imageFileName":
          "%s"
                ',
        self$`imageFileName`
        )},
        if (!is.null(self$`imageFileSize`)) {
        sprintf(
        '"imageFileSize":
          %d
                ',
        self$`imageFileSize`
        )},
        if (!is.null(self$`imageHeight`)) {
        sprintf(
        '"imageHeight":
          %d
                ',
        self$`imageHeight`
        )},
        if (!is.null(self$`imageLocation`)) {
        sprintf(
        '"imageLocation":
        %s
        ',
        jsonlite::toJSON(self$`imageLocation`$toJSON(), auto_unbox=TRUE, digits = NA)
        )},
        if (!is.null(self$`imageName`)) {
        sprintf(
        '"imageName":
          "%s"
                ',
        self$`imageName`
        )},
        if (!is.null(self$`imageTimeStamp`)) {
        sprintf(
        '"imageTimeStamp":
          "%s"
                ',
        self$`imageTimeStamp`
        )},
        if (!is.null(self$`imageURL`)) {
        sprintf(
        '"imageURL":
          "%s"
                ',
        self$`imageURL`
        )},
        if (!is.null(self$`imageWidth`)) {
        sprintf(
        '"imageWidth":
          %d
                ',
        self$`imageWidth`
        )},
        if (!is.null(self$`mimeType`)) {
        sprintf(
        '"mimeType":
          "%s"
                ',
        self$`mimeType`
        )},
        if (!is.null(self$`observationDbIds`)) {
        sprintf(
        '"observationDbIds":
           [%s]
        ',
        paste(unlist(lapply(self$`observationDbIds`, function(x) paste0('"', x, '"'))), collapse=",")
        )},
        if (!is.null(self$`observationUnitDbId`)) {
        sprintf(
        '"observationUnitDbId":
          "%s"
                ',
        self$`observationUnitDbId`
        )},
        if (!is.null(self$`imageDbId`)) {
        sprintf(
        '"imageDbId":
          "%s"
                ',
        self$`imageDbId`
        )}
      )
      jsoncontent <- paste(jsoncontent, collapse = ",")
      paste('{', jsoncontent, '}', sep = "")
    },
    fromJSONString = function(ImageJson) {
      ImageObject <- jsonlite::fromJSON(ImageJson)
      self$`additionalInfo` <- ApiClient$new()$deserializeObj(ImageObject$`additionalInfo`, "map(character)", loadNamespace("openapi"))
      self$`copyright` <- ImageObject$`copyright`
      self$`description` <- ImageObject$`description`
      self$`descriptiveOntologyTerms` <- ApiClient$new()$deserializeObj(ImageObject$`descriptiveOntologyTerms`, "array[character]", loadNamespace("openapi"))
      self$`externalReferences` <- ApiClient$new()$deserializeObj(ImageObject$`externalReferences`, "array[object]", loadNamespace("openapi"))
      self$`imageFileName` <- ImageObject$`imageFileName`
      self$`imageFileSize` <- ImageObject$`imageFileSize`
      self$`imageHeight` <- ImageObject$`imageHeight`
      self$`imageLocation` <- GeoJSON$new()$fromJSON(jsonlite::toJSON(ImageObject$imageLocation, auto_unbox = TRUE, digits = NA))
      self$`imageName` <- ImageObject$`imageName`
      self$`imageTimeStamp` <- ImageObject$`imageTimeStamp`
      self$`imageURL` <- ImageObject$`imageURL`
      self$`imageWidth` <- ImageObject$`imageWidth`
      self$`mimeType` <- ImageObject$`mimeType`
      self$`observationDbIds` <- ApiClient$new()$deserializeObj(ImageObject$`observationDbIds`, "array[character]", loadNamespace("openapi"))
      self$`observationUnitDbId` <- ImageObject$`observationUnitDbId`
      self$`imageDbId` <- ImageObject$`imageDbId`
      self
    }
  )
)

