# BrAPI-Core
#
# The Breeding API (BrAPI) is a Standardized REST ful Web Service API Specification for communicating Plant Breeding Data. BrAPI allows for easy data sharing between databases and tools involved in plant breeding. <div class=\"brapi-section\"> <h2 class=\"brapi-section-title\">General Reference Documentation</h2> <div class=\"gen-info-link\"><a href=\"https://github.com/plantbreeding/API/blob/master/Specification/GeneralInfo/URL_Structure.md\">URL Structure</a></div> <div class=\"gen-info-link\"><a href=\"https://github.com/plantbreeding/API/blob/master/Specification/GeneralInfo/Response_Structure.md\">Response Structure</a></div> <div class=\"gen-info-link\"><a href=\"https://github.com/plantbreeding/API/blob/master/Specification/GeneralInfo/Date_Time_Encoding.md\">Date/Time Encoding</a></div> <div class=\"gen-info-link\"><a href=\"https://github.com/plantbreeding/API/blob/master/Specification/GeneralInfo/Location_Encoding.md\">Location Encoding</a></div> <div class=\"gen-info-link\"><a href=\"https://github.com/plantbreeding/API/blob/master/Specification/GeneralInfo/Error_Handling.md\">Error Handling</a></div> <div class=\"gen-info-link\"><a href=\"https://github.com/plantbreeding/API/blob/master/Specification/GeneralInfo/Search_Services.md\">Search Services</a></div> </div>  <div class=\"current-brapi-section brapi-section\"> <h2 class=\"brapi-section-title\">BrAPI Core</h2> <div class=\"brapi-section-description\">The BrAPI Core module contains high level entities used for organization and management. This includes Programs, Trials, Studies, Locations, People, and Lists</div> <div class=\"version-number\">V2.0</div> <div class=\"link-btn\"><a href=\"https://github.com/plantbreeding/API/tree/master/Specification/BrAPI-Core\">GitHub</a></div> <div class=\"link-btn\"><a href=\"https://app.swaggerhub.com/apis/PlantBreedingAPI/BrAPI-Core\">SwaggerHub</a></div> <div class=\"link-btn\"><a href=\"https://brapicore.docs.apiary.io\">Apiary</a></div> <div class=\"stop-float\"></div> </div>  <div class=\"brapi-section\"> <h2 class=\"brapi-section-title\">BrAPI Phenotyping</h2> <div class=\"brapi-section-description\">The BrAPI Phenotyping module contains entities related to phenotypic observations. This includes Observation Units, Observations, Observation Variables, Traits, Scales, Methods, and Images</div> <div class=\"version-number\">V2.0</div> <div class=\"link-btn\"><a href=\"https://github.com/plantbreeding/API/tree/master/Specification/BrAPI-Phenotyping\">GitHub</a></div> <div class=\"link-btn\"><a href=\"https://app.swaggerhub.com/apis/PlantBreedingAPI/BrAPI-Phenotyping\">SwaggerHub</a></div> <div class=\"link-btn\"><a href=\"https://brapiphenotyping.docs.apiary.io\">Apiary</a></div> <div class=\"stop-float\"></div> </div>  <div class=\"brapi-section\"> <h2 class=\"brapi-section-title\">BrAPI Genotyping</h2> <div class=\"brapi-section-description\">The BrAPI Genotyping module contains entities related to genotyping analysis. This includes Samples, Markers, Variant Sets, Variants, Call Sets, Calls, References, Reads, and Vendor Orders</div> <div class=\"version-number\">V2.0</div> <div class=\"link-btn\"><a href=\"https://github.com/plantbreeding/API/tree/master/Specification/BrAPI-Genotyping\">GitHub</a></div> <div class=\"link-btn\"><a href=\"https://app.swaggerhub.com/apis/PlantBreedingAPI/BrAPI-Genotyping\">SwaggerHub</a></div> <div class=\"link-btn\"><a href=\"https://brapigenotyping.docs.apiary.io\">Apiary</a></div> <div class=\"stop-float\"></div> </div>  <div class=\"brapi-section\"> <h2 class=\"brapi-section-title\">BrAPI Germplasm</h2> <div class=\"brapi-section-description\">The BrAPI Germplasm module contains entities related to germplasm management. This includes Germplasm, Germplasm Attributes, Seed Lots, Crosses, Pedigree, and Progeny</div> <div class=\"version-number\">V2.0</div> <div class=\"link-btn\"><a href=\"https://github.com/plantbreeding/API/tree/master/Specification/BrAPI-Germplasm\">GitHub</a></div> <div class=\"link-btn\"><a href=\"https://app.swaggerhub.com/apis/PlantBreedingAPI/BrAPI-Germplasm\">SwaggerHub</a></div> <div class=\"link-btn\"><a href=\"https://brapigermplasm.docs.apiary.io\">Apiary</a></div> <div class=\"stop-float\"></div> </div>  <style> .link-btn{ float: left;  margin: 2px 10px 0 0;  padding: 0 5px;  border-radius: 5px;  background-color: #ddd; } .stop-float{   clear: both; } .version-number{   float: left;    margin: 5px 10px 0 5px; } .brapi-section-title{   margin: 0 10px 0 0;   font-size: 20px; } .current-brapi-section{   font-weight: bolder;   border-radius: 5px;    background-color: #ddd; } .brapi-section{   padding: 5px 5px;  } .brapi-section-description{   margin: 5px 0 0 5px; } </style>
#
# The version of the OpenAPI document: 2.0
# 
# Generated by: https://openapi-generator.tech

#' @docType class
#' @title ObservationUnit
#'
#' @description ObservationUnit Class
#'
#' @format An \code{R6Class} generator object
#'
#' @field additionalInfo  named list( character ) [optional]
#'
#' @field externalReferences ExternalReferences list( object ) [optional]
#'
#' @field germplasmDbId  character [optional]
#'
#' @field germplasmName  character [optional]
#'
#' @field locationDbId  character [optional]
#'
#' @field locationName  character [optional]
#'
#' @field observationUnitName  character [optional]
#'
#' @field observationUnitPUI  character [optional]
#'
#' @field observationUnitPosition  \link{ObservationUnitPosition} [optional]
#'
#' @field programDbId  character [optional]
#'
#' @field programName  character [optional]
#'
#' @field seedLotDbId  character [optional]
#'
#' @field studyDbId  character [optional]
#'
#' @field studyName  character [optional]
#'
#' @field treatments  list( \link{ObservationTreatment} ) [optional]
#'
#' @field trialDbId  character [optional]
#'
#' @field trialName  character [optional]
#'
#' @field observationUnitDbId  character [optional]
#'
#' @field observations  list( \link{Observation} ) [optional]
#'
#' @importFrom R6 R6Class
#' @importFrom jsonlite fromJSON toJSON
#' @export
ObservationUnit <- R6::R6Class(
  'ObservationUnit',
  public = list(
    `additionalInfo` = NULL,
    `externalReferences` = NULL,
    `germplasmDbId` = NULL,
    `germplasmName` = NULL,
    `locationDbId` = NULL,
    `locationName` = NULL,
    `observationUnitName` = NULL,
    `observationUnitPUI` = NULL,
    `observationUnitPosition` = NULL,
    `programDbId` = NULL,
    `programName` = NULL,
    `seedLotDbId` = NULL,
    `studyDbId` = NULL,
    `studyName` = NULL,
    `treatments` = NULL,
    `trialDbId` = NULL,
    `trialName` = NULL,
    `observationUnitDbId` = NULL,
    `observations` = NULL,
    initialize = function(
        `additionalInfo`=NULL, `externalReferences`=NULL, `germplasmDbId`=NULL, `germplasmName`=NULL, `locationDbId`=NULL, `locationName`=NULL, `observationUnitName`=NULL, `observationUnitPUI`=NULL, `observationUnitPosition`=NULL, `programDbId`=NULL, `programName`=NULL, `seedLotDbId`=NULL, `studyDbId`=NULL, `studyName`=NULL, `treatments`=NULL, `trialDbId`=NULL, `trialName`=NULL, `observationUnitDbId`=NULL, `observations`=NULL, ...
    ) {
      local.optional.var <- list(...)
      if (!is.null(`additionalInfo`)) {
        stopifnot(is.vector(`additionalInfo`), length(`additionalInfo`) != 0)
        sapply(`additionalInfo`, function(x) stopifnot(is.character(x)))
        self$`additionalInfo` <- `additionalInfo`
      }
      if (!is.null(`externalReferences`)) {
        stopifnot(is.vector(`externalReferences`), length(`externalReferences`) != 0)
        sapply(`externalReferences`, function(x) stopifnot(is.character(x)))
        self$`externalReferences` <- `externalReferences`
      }
      if (!is.null(`germplasmDbId`)) {
        stopifnot(is.character(`germplasmDbId`), length(`germplasmDbId`) == 1)
        self$`germplasmDbId` <- `germplasmDbId`
      }
      if (!is.null(`germplasmName`)) {
        stopifnot(is.character(`germplasmName`), length(`germplasmName`) == 1)
        self$`germplasmName` <- `germplasmName`
      }
      if (!is.null(`locationDbId`)) {
        stopifnot(is.character(`locationDbId`), length(`locationDbId`) == 1)
        self$`locationDbId` <- `locationDbId`
      }
      if (!is.null(`locationName`)) {
        stopifnot(is.character(`locationName`), length(`locationName`) == 1)
        self$`locationName` <- `locationName`
      }
      if (!is.null(`observationUnitName`)) {
        stopifnot(is.character(`observationUnitName`), length(`observationUnitName`) == 1)
        self$`observationUnitName` <- `observationUnitName`
      }
      if (!is.null(`observationUnitPUI`)) {
        stopifnot(is.character(`observationUnitPUI`), length(`observationUnitPUI`) == 1)
        self$`observationUnitPUI` <- `observationUnitPUI`
      }
      if (!is.null(`observationUnitPosition`)) {
        stopifnot(R6::is.R6(`observationUnitPosition`))
        self$`observationUnitPosition` <- `observationUnitPosition`
      }
      if (!is.null(`programDbId`)) {
        stopifnot(is.character(`programDbId`), length(`programDbId`) == 1)
        self$`programDbId` <- `programDbId`
      }
      if (!is.null(`programName`)) {
        stopifnot(is.character(`programName`), length(`programName`) == 1)
        self$`programName` <- `programName`
      }
      if (!is.null(`seedLotDbId`)) {
        stopifnot(is.character(`seedLotDbId`), length(`seedLotDbId`) == 1)
        self$`seedLotDbId` <- `seedLotDbId`
      }
      if (!is.null(`studyDbId`)) {
        stopifnot(is.character(`studyDbId`), length(`studyDbId`) == 1)
        self$`studyDbId` <- `studyDbId`
      }
      if (!is.null(`studyName`)) {
        stopifnot(is.character(`studyName`), length(`studyName`) == 1)
        self$`studyName` <- `studyName`
      }
      if (!is.null(`treatments`)) {
        stopifnot(is.vector(`treatments`), length(`treatments`) != 0)
        sapply(`treatments`, function(x) stopifnot(R6::is.R6(x)))
        self$`treatments` <- `treatments`
      }
      if (!is.null(`trialDbId`)) {
        stopifnot(is.character(`trialDbId`), length(`trialDbId`) == 1)
        self$`trialDbId` <- `trialDbId`
      }
      if (!is.null(`trialName`)) {
        stopifnot(is.character(`trialName`), length(`trialName`) == 1)
        self$`trialName` <- `trialName`
      }
      if (!is.null(`observationUnitDbId`)) {
        stopifnot(is.character(`observationUnitDbId`), length(`observationUnitDbId`) == 1)
        self$`observationUnitDbId` <- `observationUnitDbId`
      }
      if (!is.null(`observations`)) {
        stopifnot(is.vector(`observations`), length(`observations`) != 0)
        sapply(`observations`, function(x) stopifnot(R6::is.R6(x)))
        self$`observations` <- `observations`
      }
    },
    toJSON = function() {
      ObservationUnitObject <- list()
      if (!is.null(self$`additionalInfo`)) {
        ObservationUnitObject[['additionalInfo']] <-
          self$`additionalInfo`
      }
      if (!is.null(self$`externalReferences`)) {
        ObservationUnitObject[['externalReferences']] <-
          self$`externalReferences`
      }
      if (!is.null(self$`germplasmDbId`)) {
        ObservationUnitObject[['germplasmDbId']] <-
          self$`germplasmDbId`
      }
      if (!is.null(self$`germplasmName`)) {
        ObservationUnitObject[['germplasmName']] <-
          self$`germplasmName`
      }
      if (!is.null(self$`locationDbId`)) {
        ObservationUnitObject[['locationDbId']] <-
          self$`locationDbId`
      }
      if (!is.null(self$`locationName`)) {
        ObservationUnitObject[['locationName']] <-
          self$`locationName`
      }
      if (!is.null(self$`observationUnitName`)) {
        ObservationUnitObject[['observationUnitName']] <-
          self$`observationUnitName`
      }
      if (!is.null(self$`observationUnitPUI`)) {
        ObservationUnitObject[['observationUnitPUI']] <-
          self$`observationUnitPUI`
      }
      if (!is.null(self$`observationUnitPosition`)) {
        ObservationUnitObject[['observationUnitPosition']] <-
          self$`observationUnitPosition`$toJSON()
      }
      if (!is.null(self$`programDbId`)) {
        ObservationUnitObject[['programDbId']] <-
          self$`programDbId`
      }
      if (!is.null(self$`programName`)) {
        ObservationUnitObject[['programName']] <-
          self$`programName`
      }
      if (!is.null(self$`seedLotDbId`)) {
        ObservationUnitObject[['seedLotDbId']] <-
          self$`seedLotDbId`
      }
      if (!is.null(self$`studyDbId`)) {
        ObservationUnitObject[['studyDbId']] <-
          self$`studyDbId`
      }
      if (!is.null(self$`studyName`)) {
        ObservationUnitObject[['studyName']] <-
          self$`studyName`
      }
      if (!is.null(self$`treatments`)) {
        ObservationUnitObject[['treatments']] <-
          lapply(self$`treatments`, function(x) x$toJSON())
      }
      if (!is.null(self$`trialDbId`)) {
        ObservationUnitObject[['trialDbId']] <-
          self$`trialDbId`
      }
      if (!is.null(self$`trialName`)) {
        ObservationUnitObject[['trialName']] <-
          self$`trialName`
      }
      if (!is.null(self$`observationUnitDbId`)) {
        ObservationUnitObject[['observationUnitDbId']] <-
          self$`observationUnitDbId`
      }
      if (!is.null(self$`observations`)) {
        ObservationUnitObject[['observations']] <-
          lapply(self$`observations`, function(x) x$toJSON())
      }

      ObservationUnitObject
    },
    fromJSON = function(ObservationUnitJson) {
      ObservationUnitObject <- jsonlite::fromJSON(ObservationUnitJson)
      if (!is.null(ObservationUnitObject$`additionalInfo`)) {
        self$`additionalInfo` <- ApiClient$new()$deserializeObj(ObservationUnitObject$`additionalInfo`, "map(character)", loadNamespace("openapi"))
      }
      if (!is.null(ObservationUnitObject$`externalReferences`)) {
        self$`externalReferences` <- ApiClient$new()$deserializeObj(ObservationUnitObject$`externalReferences`, "array[object]", loadNamespace("openapi"))
      }
      if (!is.null(ObservationUnitObject$`germplasmDbId`)) {
        self$`germplasmDbId` <- ObservationUnitObject$`germplasmDbId`
      }
      if (!is.null(ObservationUnitObject$`germplasmName`)) {
        self$`germplasmName` <- ObservationUnitObject$`germplasmName`
      }
      if (!is.null(ObservationUnitObject$`locationDbId`)) {
        self$`locationDbId` <- ObservationUnitObject$`locationDbId`
      }
      if (!is.null(ObservationUnitObject$`locationName`)) {
        self$`locationName` <- ObservationUnitObject$`locationName`
      }
      if (!is.null(ObservationUnitObject$`observationUnitName`)) {
        self$`observationUnitName` <- ObservationUnitObject$`observationUnitName`
      }
      if (!is.null(ObservationUnitObject$`observationUnitPUI`)) {
        self$`observationUnitPUI` <- ObservationUnitObject$`observationUnitPUI`
      }
      if (!is.null(ObservationUnitObject$`observationUnitPosition`)) {
        observationUnitPositionObject <- ObservationUnitPosition$new()
        observationUnitPositionObject$fromJSON(jsonlite::toJSON(ObservationUnitObject$observationUnitPosition, auto_unbox = TRUE, digits = NA))
        self$`observationUnitPosition` <- observationUnitPositionObject
      }
      if (!is.null(ObservationUnitObject$`programDbId`)) {
        self$`programDbId` <- ObservationUnitObject$`programDbId`
      }
      if (!is.null(ObservationUnitObject$`programName`)) {
        self$`programName` <- ObservationUnitObject$`programName`
      }
      if (!is.null(ObservationUnitObject$`seedLotDbId`)) {
        self$`seedLotDbId` <- ObservationUnitObject$`seedLotDbId`
      }
      if (!is.null(ObservationUnitObject$`studyDbId`)) {
        self$`studyDbId` <- ObservationUnitObject$`studyDbId`
      }
      if (!is.null(ObservationUnitObject$`studyName`)) {
        self$`studyName` <- ObservationUnitObject$`studyName`
      }
      if (!is.null(ObservationUnitObject$`treatments`)) {
        self$`treatments` <- ApiClient$new()$deserializeObj(ObservationUnitObject$`treatments`, "array[ObservationTreatment]", loadNamespace("openapi"))
      }
      if (!is.null(ObservationUnitObject$`trialDbId`)) {
        self$`trialDbId` <- ObservationUnitObject$`trialDbId`
      }
      if (!is.null(ObservationUnitObject$`trialName`)) {
        self$`trialName` <- ObservationUnitObject$`trialName`
      }
      if (!is.null(ObservationUnitObject$`observationUnitDbId`)) {
        self$`observationUnitDbId` <- ObservationUnitObject$`observationUnitDbId`
      }
      if (!is.null(ObservationUnitObject$`observations`)) {
        self$`observations` <- ApiClient$new()$deserializeObj(ObservationUnitObject$`observations`, "array[Observation]", loadNamespace("openapi"))
      }
      self
    },
    toJSONString = function() {
      jsoncontent <- c(
        if (!is.null(self$`additionalInfo`)) {
        sprintf(
        '"additionalInfo":
          "%s"
        ',
        jsonlite::toJSON(lapply(self$`additionalInfo`, function(x){ x }), auto_unbox = TRUE, digits=NA)
        )},
        if (!is.null(self$`externalReferences`)) {
        sprintf(
        '"externalReferences":
           [%s]
        ',
        paste(unlist(lapply(self$`externalReferences`, function(x) paste0('"', x, '"'))), collapse=",")
        )},
        if (!is.null(self$`germplasmDbId`)) {
        sprintf(
        '"germplasmDbId":
          "%s"
                ',
        self$`germplasmDbId`
        )},
        if (!is.null(self$`germplasmName`)) {
        sprintf(
        '"germplasmName":
          "%s"
                ',
        self$`germplasmName`
        )},
        if (!is.null(self$`locationDbId`)) {
        sprintf(
        '"locationDbId":
          "%s"
                ',
        self$`locationDbId`
        )},
        if (!is.null(self$`locationName`)) {
        sprintf(
        '"locationName":
          "%s"
                ',
        self$`locationName`
        )},
        if (!is.null(self$`observationUnitName`)) {
        sprintf(
        '"observationUnitName":
          "%s"
                ',
        self$`observationUnitName`
        )},
        if (!is.null(self$`observationUnitPUI`)) {
        sprintf(
        '"observationUnitPUI":
          "%s"
                ',
        self$`observationUnitPUI`
        )},
        if (!is.null(self$`observationUnitPosition`)) {
        sprintf(
        '"observationUnitPosition":
        %s
        ',
        jsonlite::toJSON(self$`observationUnitPosition`$toJSON(), auto_unbox=TRUE, digits = NA)
        )},
        if (!is.null(self$`programDbId`)) {
        sprintf(
        '"programDbId":
          "%s"
                ',
        self$`programDbId`
        )},
        if (!is.null(self$`programName`)) {
        sprintf(
        '"programName":
          "%s"
                ',
        self$`programName`
        )},
        if (!is.null(self$`seedLotDbId`)) {
        sprintf(
        '"seedLotDbId":
          "%s"
                ',
        self$`seedLotDbId`
        )},
        if (!is.null(self$`studyDbId`)) {
        sprintf(
        '"studyDbId":
          "%s"
                ',
        self$`studyDbId`
        )},
        if (!is.null(self$`studyName`)) {
        sprintf(
        '"studyName":
          "%s"
                ',
        self$`studyName`
        )},
        if (!is.null(self$`treatments`)) {
        sprintf(
        '"treatments":
        [%s]
',
        paste(sapply(self$`treatments`, function(x) jsonlite::toJSON(x$toJSON(), auto_unbox=TRUE, digits = NA)), collapse=",")
        )},
        if (!is.null(self$`trialDbId`)) {
        sprintf(
        '"trialDbId":
          "%s"
                ',
        self$`trialDbId`
        )},
        if (!is.null(self$`trialName`)) {
        sprintf(
        '"trialName":
          "%s"
                ',
        self$`trialName`
        )},
        if (!is.null(self$`observationUnitDbId`)) {
        sprintf(
        '"observationUnitDbId":
          "%s"
                ',
        self$`observationUnitDbId`
        )},
        if (!is.null(self$`observations`)) {
        sprintf(
        '"observations":
        [%s]
',
        paste(sapply(self$`observations`, function(x) jsonlite::toJSON(x$toJSON(), auto_unbox=TRUE, digits = NA)), collapse=",")
        )}
      )
      jsoncontent <- paste(jsoncontent, collapse = ",")
      paste('{', jsoncontent, '}', sep = "")
    },
    fromJSONString = function(ObservationUnitJson) {
      ObservationUnitObject <- jsonlite::fromJSON(ObservationUnitJson)
      self$`additionalInfo` <- ApiClient$new()$deserializeObj(ObservationUnitObject$`additionalInfo`, "map(character)", loadNamespace("openapi"))
      self$`externalReferences` <- ApiClient$new()$deserializeObj(ObservationUnitObject$`externalReferences`, "array[object]", loadNamespace("openapi"))
      self$`germplasmDbId` <- ObservationUnitObject$`germplasmDbId`
      self$`germplasmName` <- ObservationUnitObject$`germplasmName`
      self$`locationDbId` <- ObservationUnitObject$`locationDbId`
      self$`locationName` <- ObservationUnitObject$`locationName`
      self$`observationUnitName` <- ObservationUnitObject$`observationUnitName`
      self$`observationUnitPUI` <- ObservationUnitObject$`observationUnitPUI`
      self$`observationUnitPosition` <- ObservationUnitPosition$new()$fromJSON(jsonlite::toJSON(ObservationUnitObject$observationUnitPosition, auto_unbox = TRUE, digits = NA))
      self$`programDbId` <- ObservationUnitObject$`programDbId`
      self$`programName` <- ObservationUnitObject$`programName`
      self$`seedLotDbId` <- ObservationUnitObject$`seedLotDbId`
      self$`studyDbId` <- ObservationUnitObject$`studyDbId`
      self$`studyName` <- ObservationUnitObject$`studyName`
      self$`treatments` <- ApiClient$new()$deserializeObj(ObservationUnitObject$`treatments`, "array[ObservationTreatment]", loadNamespace("openapi"))
      self$`trialDbId` <- ObservationUnitObject$`trialDbId`
      self$`trialName` <- ObservationUnitObject$`trialName`
      self$`observationUnitDbId` <- ObservationUnitObject$`observationUnitDbId`
      self$`observations` <- ApiClient$new()$deserializeObj(ObservationUnitObject$`observations`, "array[Observation]", loadNamespace("openapi"))
      self
    }
  )
)

