# BrAPI-Core
#
# The Breeding API (BrAPI) is a Standardized REST ful Web Service API Specification for communicating Plant Breeding Data. BrAPI allows for easy data sharing between databases and tools involved in plant breeding. <div class=\"brapi-section\"> <h2 class=\"brapi-section-title\">General Reference Documentation</h2> <div class=\"gen-info-link\"><a href=\"https://github.com/plantbreeding/API/blob/master/Specification/GeneralInfo/URL_Structure.md\">URL Structure</a></div> <div class=\"gen-info-link\"><a href=\"https://github.com/plantbreeding/API/blob/master/Specification/GeneralInfo/Response_Structure.md\">Response Structure</a></div> <div class=\"gen-info-link\"><a href=\"https://github.com/plantbreeding/API/blob/master/Specification/GeneralInfo/Date_Time_Encoding.md\">Date/Time Encoding</a></div> <div class=\"gen-info-link\"><a href=\"https://github.com/plantbreeding/API/blob/master/Specification/GeneralInfo/Location_Encoding.md\">Location Encoding</a></div> <div class=\"gen-info-link\"><a href=\"https://github.com/plantbreeding/API/blob/master/Specification/GeneralInfo/Error_Handling.md\">Error Handling</a></div> <div class=\"gen-info-link\"><a href=\"https://github.com/plantbreeding/API/blob/master/Specification/GeneralInfo/Search_Services.md\">Search Services</a></div> </div>  <div class=\"current-brapi-section brapi-section\"> <h2 class=\"brapi-section-title\">BrAPI Core</h2> <div class=\"brapi-section-description\">The BrAPI Core module contains high level entities used for organization and management. This includes Programs, Trials, Studies, Locations, People, and Lists</div> <div class=\"version-number\">V2.0</div> <div class=\"link-btn\"><a href=\"https://github.com/plantbreeding/API/tree/master/Specification/BrAPI-Core\">GitHub</a></div> <div class=\"link-btn\"><a href=\"https://app.swaggerhub.com/apis/PlantBreedingAPI/BrAPI-Core\">SwaggerHub</a></div> <div class=\"link-btn\"><a href=\"https://brapicore.docs.apiary.io\">Apiary</a></div> <div class=\"stop-float\"></div> </div>  <div class=\"brapi-section\"> <h2 class=\"brapi-section-title\">BrAPI Phenotyping</h2> <div class=\"brapi-section-description\">The BrAPI Phenotyping module contains entities related to phenotypic observations. This includes Observation Units, Observations, Observation Variables, Traits, Scales, Methods, and Images</div> <div class=\"version-number\">V2.0</div> <div class=\"link-btn\"><a href=\"https://github.com/plantbreeding/API/tree/master/Specification/BrAPI-Phenotyping\">GitHub</a></div> <div class=\"link-btn\"><a href=\"https://app.swaggerhub.com/apis/PlantBreedingAPI/BrAPI-Phenotyping\">SwaggerHub</a></div> <div class=\"link-btn\"><a href=\"https://brapiphenotyping.docs.apiary.io\">Apiary</a></div> <div class=\"stop-float\"></div> </div>  <div class=\"brapi-section\"> <h2 class=\"brapi-section-title\">BrAPI Genotyping</h2> <div class=\"brapi-section-description\">The BrAPI Genotyping module contains entities related to genotyping analysis. This includes Samples, Markers, Variant Sets, Variants, Call Sets, Calls, References, Reads, and Vendor Orders</div> <div class=\"version-number\">V2.0</div> <div class=\"link-btn\"><a href=\"https://github.com/plantbreeding/API/tree/master/Specification/BrAPI-Genotyping\">GitHub</a></div> <div class=\"link-btn\"><a href=\"https://app.swaggerhub.com/apis/PlantBreedingAPI/BrAPI-Genotyping\">SwaggerHub</a></div> <div class=\"link-btn\"><a href=\"https://brapigenotyping.docs.apiary.io\">Apiary</a></div> <div class=\"stop-float\"></div> </div>  <div class=\"brapi-section\"> <h2 class=\"brapi-section-title\">BrAPI Germplasm</h2> <div class=\"brapi-section-description\">The BrAPI Germplasm module contains entities related to germplasm management. This includes Germplasm, Germplasm Attributes, Seed Lots, Crosses, Pedigree, and Progeny</div> <div class=\"version-number\">V2.0</div> <div class=\"link-btn\"><a href=\"https://github.com/plantbreeding/API/tree/master/Specification/BrAPI-Germplasm\">GitHub</a></div> <div class=\"link-btn\"><a href=\"https://app.swaggerhub.com/apis/PlantBreedingAPI/BrAPI-Germplasm\">SwaggerHub</a></div> <div class=\"link-btn\"><a href=\"https://brapigermplasm.docs.apiary.io\">Apiary</a></div> <div class=\"stop-float\"></div> </div>  <style> .link-btn{ float: left;  margin: 2px 10px 0 0;  padding: 0 5px;  border-radius: 5px;  background-color: #ddd; } .stop-float{   clear: both; } .version-number{   float: left;    margin: 5px 10px 0 5px; } .brapi-section-title{   margin: 0 10px 0 0;   font-size: 20px; } .current-brapi-section{   font-weight: bolder;   border-radius: 5px;    background-color: #ddd; } .brapi-section{   padding: 5px 5px;  } .brapi-section-description{   margin: 5px 0 0 5px; } </style>
#
# The version of the OpenAPI document: 2.0
# 
# Generated by: https://openapi-generator.tech

#' @docType class
#' @title ReferenceSet
#'
#' @description ReferenceSet Class
#'
#' @format An \code{R6Class} generator object
#'
#' @field additionalInfo  named list( character ) [optional]
#'
#' @field assemblyPUI  character [optional]
#'
#' @field description  character [optional]
#'
#' @field isDerived  character [optional]
#'
#' @field md5checksum  character [optional]
#'
#' @field referenceSetDbId  character [optional]
#'
#' @field referenceSetName  character [optional]
#'
#' @field sourceAccessions  list( character ) [optional]
#'
#' @field sourceURI  character [optional]
#'
#' @field species  \link{OntologyTerm} [optional]
#'
#' @importFrom R6 R6Class
#' @importFrom jsonlite fromJSON toJSON
#' @export
ReferenceSet <- R6::R6Class(
  'ReferenceSet',
  public = list(
    `additionalInfo` = NULL,
    `assemblyPUI` = NULL,
    `description` = NULL,
    `isDerived` = NULL,
    `md5checksum` = NULL,
    `referenceSetDbId` = NULL,
    `referenceSetName` = NULL,
    `sourceAccessions` = NULL,
    `sourceURI` = NULL,
    `species` = NULL,
    initialize = function(
        `additionalInfo`=NULL, `assemblyPUI`=NULL, `description`=NULL, `isDerived`=NULL, `md5checksum`=NULL, `referenceSetDbId`=NULL, `referenceSetName`=NULL, `sourceAccessions`=NULL, `sourceURI`=NULL, `species`=NULL, ...
    ) {
      local.optional.var <- list(...)
      if (!is.null(`additionalInfo`)) {
        stopifnot(is.vector(`additionalInfo`), length(`additionalInfo`) != 0)
        sapply(`additionalInfo`, function(x) stopifnot(is.character(x)))
        self$`additionalInfo` <- `additionalInfo`
      }
      if (!is.null(`assemblyPUI`)) {
        stopifnot(is.character(`assemblyPUI`), length(`assemblyPUI`) == 1)
        self$`assemblyPUI` <- `assemblyPUI`
      }
      if (!is.null(`description`)) {
        stopifnot(is.character(`description`), length(`description`) == 1)
        self$`description` <- `description`
      }
      if (!is.null(`isDerived`)) {
        self$`isDerived` <- `isDerived`
      }
      if (!is.null(`md5checksum`)) {
        stopifnot(is.character(`md5checksum`), length(`md5checksum`) == 1)
        self$`md5checksum` <- `md5checksum`
      }
      if (!is.null(`referenceSetDbId`)) {
        stopifnot(is.character(`referenceSetDbId`), length(`referenceSetDbId`) == 1)
        self$`referenceSetDbId` <- `referenceSetDbId`
      }
      if (!is.null(`referenceSetName`)) {
        stopifnot(is.character(`referenceSetName`), length(`referenceSetName`) == 1)
        self$`referenceSetName` <- `referenceSetName`
      }
      if (!is.null(`sourceAccessions`)) {
        stopifnot(is.vector(`sourceAccessions`), length(`sourceAccessions`) != 0)
        sapply(`sourceAccessions`, function(x) stopifnot(is.character(x)))
        self$`sourceAccessions` <- `sourceAccessions`
      }
      if (!is.null(`sourceURI`)) {
        stopifnot(is.character(`sourceURI`), length(`sourceURI`) == 1)
        self$`sourceURI` <- `sourceURI`
      }
      if (!is.null(`species`)) {
        stopifnot(R6::is.R6(`species`))
        self$`species` <- `species`
      }
    },
    toJSON = function() {
      ReferenceSetObject <- list()
      if (!is.null(self$`additionalInfo`)) {
        ReferenceSetObject[['additionalInfo']] <-
          self$`additionalInfo`
      }
      if (!is.null(self$`assemblyPUI`)) {
        ReferenceSetObject[['assemblyPUI']] <-
          self$`assemblyPUI`
      }
      if (!is.null(self$`description`)) {
        ReferenceSetObject[['description']] <-
          self$`description`
      }
      if (!is.null(self$`isDerived`)) {
        ReferenceSetObject[['isDerived']] <-
          self$`isDerived`
      }
      if (!is.null(self$`md5checksum`)) {
        ReferenceSetObject[['md5checksum']] <-
          self$`md5checksum`
      }
      if (!is.null(self$`referenceSetDbId`)) {
        ReferenceSetObject[['referenceSetDbId']] <-
          self$`referenceSetDbId`
      }
      if (!is.null(self$`referenceSetName`)) {
        ReferenceSetObject[['referenceSetName']] <-
          self$`referenceSetName`
      }
      if (!is.null(self$`sourceAccessions`)) {
        ReferenceSetObject[['sourceAccessions']] <-
          self$`sourceAccessions`
      }
      if (!is.null(self$`sourceURI`)) {
        ReferenceSetObject[['sourceURI']] <-
          self$`sourceURI`
      }
      if (!is.null(self$`species`)) {
        ReferenceSetObject[['species']] <-
          self$`species`$toJSON()
      }

      ReferenceSetObject
    },
    fromJSON = function(ReferenceSetJson) {
      ReferenceSetObject <- jsonlite::fromJSON(ReferenceSetJson)
      if (!is.null(ReferenceSetObject$`additionalInfo`)) {
        self$`additionalInfo` <- ApiClient$new()$deserializeObj(ReferenceSetObject$`additionalInfo`, "map(character)", loadNamespace("openapi"))
      }
      if (!is.null(ReferenceSetObject$`assemblyPUI`)) {
        self$`assemblyPUI` <- ReferenceSetObject$`assemblyPUI`
      }
      if (!is.null(ReferenceSetObject$`description`)) {
        self$`description` <- ReferenceSetObject$`description`
      }
      if (!is.null(ReferenceSetObject$`isDerived`)) {
        self$`isDerived` <- ReferenceSetObject$`isDerived`
      }
      if (!is.null(ReferenceSetObject$`md5checksum`)) {
        self$`md5checksum` <- ReferenceSetObject$`md5checksum`
      }
      if (!is.null(ReferenceSetObject$`referenceSetDbId`)) {
        self$`referenceSetDbId` <- ReferenceSetObject$`referenceSetDbId`
      }
      if (!is.null(ReferenceSetObject$`referenceSetName`)) {
        self$`referenceSetName` <- ReferenceSetObject$`referenceSetName`
      }
      if (!is.null(ReferenceSetObject$`sourceAccessions`)) {
        self$`sourceAccessions` <- ApiClient$new()$deserializeObj(ReferenceSetObject$`sourceAccessions`, "array[character]", loadNamespace("openapi"))
      }
      if (!is.null(ReferenceSetObject$`sourceURI`)) {
        self$`sourceURI` <- ReferenceSetObject$`sourceURI`
      }
      if (!is.null(ReferenceSetObject$`species`)) {
        speciesObject <- OntologyTerm$new()
        speciesObject$fromJSON(jsonlite::toJSON(ReferenceSetObject$species, auto_unbox = TRUE, digits = NA))
        self$`species` <- speciesObject
      }
      self
    },
    toJSONString = function() {
      jsoncontent <- c(
        if (!is.null(self$`additionalInfo`)) {
        sprintf(
        '"additionalInfo":
          "%s"
        ',
        jsonlite::toJSON(lapply(self$`additionalInfo`, function(x){ x }), auto_unbox = TRUE, digits=NA)
        )},
        if (!is.null(self$`assemblyPUI`)) {
        sprintf(
        '"assemblyPUI":
          "%s"
                ',
        self$`assemblyPUI`
        )},
        if (!is.null(self$`description`)) {
        sprintf(
        '"description":
          "%s"
                ',
        self$`description`
        )},
        if (!is.null(self$`isDerived`)) {
        sprintf(
        '"isDerived":
          "%s"
                ',
        self$`isDerived`
        )},
        if (!is.null(self$`md5checksum`)) {
        sprintf(
        '"md5checksum":
          "%s"
                ',
        self$`md5checksum`
        )},
        if (!is.null(self$`referenceSetDbId`)) {
        sprintf(
        '"referenceSetDbId":
          "%s"
                ',
        self$`referenceSetDbId`
        )},
        if (!is.null(self$`referenceSetName`)) {
        sprintf(
        '"referenceSetName":
          "%s"
                ',
        self$`referenceSetName`
        )},
        if (!is.null(self$`sourceAccessions`)) {
        sprintf(
        '"sourceAccessions":
           [%s]
        ',
        paste(unlist(lapply(self$`sourceAccessions`, function(x) paste0('"', x, '"'))), collapse=",")
        )},
        if (!is.null(self$`sourceURI`)) {
        sprintf(
        '"sourceURI":
          "%s"
                ',
        self$`sourceURI`
        )},
        if (!is.null(self$`species`)) {
        sprintf(
        '"species":
        %s
        ',
        jsonlite::toJSON(self$`species`$toJSON(), auto_unbox=TRUE, digits = NA)
        )}
      )
      jsoncontent <- paste(jsoncontent, collapse = ",")
      paste('{', jsoncontent, '}', sep = "")
    },
    fromJSONString = function(ReferenceSetJson) {
      ReferenceSetObject <- jsonlite::fromJSON(ReferenceSetJson)
      self$`additionalInfo` <- ApiClient$new()$deserializeObj(ReferenceSetObject$`additionalInfo`, "map(character)", loadNamespace("openapi"))
      self$`assemblyPUI` <- ReferenceSetObject$`assemblyPUI`
      self$`description` <- ReferenceSetObject$`description`
      self$`isDerived` <- ReferenceSetObject$`isDerived`
      self$`md5checksum` <- ReferenceSetObject$`md5checksum`
      self$`referenceSetDbId` <- ReferenceSetObject$`referenceSetDbId`
      self$`referenceSetName` <- ReferenceSetObject$`referenceSetName`
      self$`sourceAccessions` <- ApiClient$new()$deserializeObj(ReferenceSetObject$`sourceAccessions`, "array[character]", loadNamespace("openapi"))
      self$`sourceURI` <- ReferenceSetObject$`sourceURI`
      self$`species` <- OntologyTerm$new()$fromJSON(jsonlite::toJSON(ReferenceSetObject$species, auto_unbox = TRUE, digits = NA))
      self
    }
  )
)

