# BrAPI-Core
#
# The Breeding API (BrAPI) is a Standardized REST ful Web Service API Specification for communicating Plant Breeding Data. BrAPI allows for easy data sharing between databases and tools involved in plant breeding. <div class=\"brapi-section\"> <h2 class=\"brapi-section-title\">General Reference Documentation</h2> <div class=\"gen-info-link\"><a href=\"https://github.com/plantbreeding/API/blob/master/Specification/GeneralInfo/URL_Structure.md\">URL Structure</a></div> <div class=\"gen-info-link\"><a href=\"https://github.com/plantbreeding/API/blob/master/Specification/GeneralInfo/Response_Structure.md\">Response Structure</a></div> <div class=\"gen-info-link\"><a href=\"https://github.com/plantbreeding/API/blob/master/Specification/GeneralInfo/Date_Time_Encoding.md\">Date/Time Encoding</a></div> <div class=\"gen-info-link\"><a href=\"https://github.com/plantbreeding/API/blob/master/Specification/GeneralInfo/Location_Encoding.md\">Location Encoding</a></div> <div class=\"gen-info-link\"><a href=\"https://github.com/plantbreeding/API/blob/master/Specification/GeneralInfo/Error_Handling.md\">Error Handling</a></div> <div class=\"gen-info-link\"><a href=\"https://github.com/plantbreeding/API/blob/master/Specification/GeneralInfo/Search_Services.md\">Search Services</a></div> </div>  <div class=\"current-brapi-section brapi-section\"> <h2 class=\"brapi-section-title\">BrAPI Core</h2> <div class=\"brapi-section-description\">The BrAPI Core module contains high level entities used for organization and management. This includes Programs, Trials, Studies, Locations, People, and Lists</div> <div class=\"version-number\">V2.0</div> <div class=\"link-btn\"><a href=\"https://github.com/plantbreeding/API/tree/master/Specification/BrAPI-Core\">GitHub</a></div> <div class=\"link-btn\"><a href=\"https://app.swaggerhub.com/apis/PlantBreedingAPI/BrAPI-Core\">SwaggerHub</a></div> <div class=\"link-btn\"><a href=\"https://brapicore.docs.apiary.io\">Apiary</a></div> <div class=\"stop-float\"></div> </div>  <div class=\"brapi-section\"> <h2 class=\"brapi-section-title\">BrAPI Phenotyping</h2> <div class=\"brapi-section-description\">The BrAPI Phenotyping module contains entities related to phenotypic observations. This includes Observation Units, Observations, Observation Variables, Traits, Scales, Methods, and Images</div> <div class=\"version-number\">V2.0</div> <div class=\"link-btn\"><a href=\"https://github.com/plantbreeding/API/tree/master/Specification/BrAPI-Phenotyping\">GitHub</a></div> <div class=\"link-btn\"><a href=\"https://app.swaggerhub.com/apis/PlantBreedingAPI/BrAPI-Phenotyping\">SwaggerHub</a></div> <div class=\"link-btn\"><a href=\"https://brapiphenotyping.docs.apiary.io\">Apiary</a></div> <div class=\"stop-float\"></div> </div>  <div class=\"brapi-section\"> <h2 class=\"brapi-section-title\">BrAPI Genotyping</h2> <div class=\"brapi-section-description\">The BrAPI Genotyping module contains entities related to genotyping analysis. This includes Samples, Markers, Variant Sets, Variants, Call Sets, Calls, References, Reads, and Vendor Orders</div> <div class=\"version-number\">V2.0</div> <div class=\"link-btn\"><a href=\"https://github.com/plantbreeding/API/tree/master/Specification/BrAPI-Genotyping\">GitHub</a></div> <div class=\"link-btn\"><a href=\"https://app.swaggerhub.com/apis/PlantBreedingAPI/BrAPI-Genotyping\">SwaggerHub</a></div> <div class=\"link-btn\"><a href=\"https://brapigenotyping.docs.apiary.io\">Apiary</a></div> <div class=\"stop-float\"></div> </div>  <div class=\"brapi-section\"> <h2 class=\"brapi-section-title\">BrAPI Germplasm</h2> <div class=\"brapi-section-description\">The BrAPI Germplasm module contains entities related to germplasm management. This includes Germplasm, Germplasm Attributes, Seed Lots, Crosses, Pedigree, and Progeny</div> <div class=\"version-number\">V2.0</div> <div class=\"link-btn\"><a href=\"https://github.com/plantbreeding/API/tree/master/Specification/BrAPI-Germplasm\">GitHub</a></div> <div class=\"link-btn\"><a href=\"https://app.swaggerhub.com/apis/PlantBreedingAPI/BrAPI-Germplasm\">SwaggerHub</a></div> <div class=\"link-btn\"><a href=\"https://brapigermplasm.docs.apiary.io\">Apiary</a></div> <div class=\"stop-float\"></div> </div>  <style> .link-btn{ float: left;  margin: 2px 10px 0 0;  padding: 0 5px;  border-radius: 5px;  background-color: #ddd; } .stop-float{   clear: both; } .version-number{   float: left;    margin: 5px 10px 0 5px; } .brapi-section-title{   margin: 0 10px 0 0;   font-size: 20px; } .current-brapi-section{   font-weight: bolder;   border-radius: 5px;    background-color: #ddd; } .brapi-section{   padding: 5px 5px;  } .brapi-section-description{   margin: 5px 0 0 5px; } </style>
#
# The version of the OpenAPI document: 2.0
# 
# Generated by: https://openapi-generator.tech

#' @docType class
#' @title SampleNewRequest
#'
#' @description SampleNewRequest Class
#'
#' @format An \code{R6Class} generator object
#'
#' @field additionalInfo  named list( character ) [optional]
#'
#' @field column  integer [optional]
#'
#' @field externalReferences ExternalReferences list( object ) [optional]
#'
#' @field germplasmDbId  character [optional]
#'
#' @field observationUnitDbId  character [optional]
#'
#' @field plateDbId  character [optional]
#'
#' @field plateName  character [optional]
#'
#' @field programDbId  character [optional]
#'
#' @field row  character [optional]
#'
#' @field sampleBarcode  character [optional]
#'
#' @field sampleDescription  character [optional]
#'
#' @field sampleGroupDbId  character [optional]
#'
#' @field sampleName  character [optional]
#'
#' @field samplePUI  character [optional]
#'
#' @field sampleTimestamp  character [optional]
#'
#' @field sampleType  character [optional]
#'
#' @field studyDbId  character [optional]
#'
#' @field takenBy  character [optional]
#'
#' @field tissueType  character [optional]
#'
#' @field trialDbId  character [optional]
#'
#' @field well  character [optional]
#'
#' @importFrom R6 R6Class
#' @importFrom jsonlite fromJSON toJSON
#' @export
SampleNewRequest <- R6::R6Class(
  'SampleNewRequest',
  public = list(
    `additionalInfo` = NULL,
    `column` = NULL,
    `externalReferences` = NULL,
    `germplasmDbId` = NULL,
    `observationUnitDbId` = NULL,
    `plateDbId` = NULL,
    `plateName` = NULL,
    `programDbId` = NULL,
    `row` = NULL,
    `sampleBarcode` = NULL,
    `sampleDescription` = NULL,
    `sampleGroupDbId` = NULL,
    `sampleName` = NULL,
    `samplePUI` = NULL,
    `sampleTimestamp` = NULL,
    `sampleType` = NULL,
    `studyDbId` = NULL,
    `takenBy` = NULL,
    `tissueType` = NULL,
    `trialDbId` = NULL,
    `well` = NULL,
    initialize = function(
        `additionalInfo`=NULL, `column`=NULL, `externalReferences`=NULL, `germplasmDbId`=NULL, `observationUnitDbId`=NULL, `plateDbId`=NULL, `plateName`=NULL, `programDbId`=NULL, `row`=NULL, `sampleBarcode`=NULL, `sampleDescription`=NULL, `sampleGroupDbId`=NULL, `sampleName`=NULL, `samplePUI`=NULL, `sampleTimestamp`=NULL, `sampleType`=NULL, `studyDbId`=NULL, `takenBy`=NULL, `tissueType`=NULL, `trialDbId`=NULL, `well`=NULL, ...
    ) {
      local.optional.var <- list(...)
      if (!is.null(`additionalInfo`)) {
        stopifnot(is.vector(`additionalInfo`), length(`additionalInfo`) != 0)
        sapply(`additionalInfo`, function(x) stopifnot(is.character(x)))
        self$`additionalInfo` <- `additionalInfo`
      }
      if (!is.null(`column`)) {
        stopifnot(is.numeric(`column`), length(`column`) == 1)
        self$`column` <- `column`
      }
      if (!is.null(`externalReferences`)) {
        stopifnot(is.vector(`externalReferences`), length(`externalReferences`) != 0)
        sapply(`externalReferences`, function(x) stopifnot(is.character(x)))
        self$`externalReferences` <- `externalReferences`
      }
      if (!is.null(`germplasmDbId`)) {
        stopifnot(is.character(`germplasmDbId`), length(`germplasmDbId`) == 1)
        self$`germplasmDbId` <- `germplasmDbId`
      }
      if (!is.null(`observationUnitDbId`)) {
        stopifnot(is.character(`observationUnitDbId`), length(`observationUnitDbId`) == 1)
        self$`observationUnitDbId` <- `observationUnitDbId`
      }
      if (!is.null(`plateDbId`)) {
        stopifnot(is.character(`plateDbId`), length(`plateDbId`) == 1)
        self$`plateDbId` <- `plateDbId`
      }
      if (!is.null(`plateName`)) {
        stopifnot(is.character(`plateName`), length(`plateName`) == 1)
        self$`plateName` <- `plateName`
      }
      if (!is.null(`programDbId`)) {
        stopifnot(is.character(`programDbId`), length(`programDbId`) == 1)
        self$`programDbId` <- `programDbId`
      }
      if (!is.null(`row`)) {
        stopifnot(is.character(`row`), length(`row`) == 1)
        self$`row` <- `row`
      }
      if (!is.null(`sampleBarcode`)) {
        stopifnot(is.character(`sampleBarcode`), length(`sampleBarcode`) == 1)
        self$`sampleBarcode` <- `sampleBarcode`
      }
      if (!is.null(`sampleDescription`)) {
        stopifnot(is.character(`sampleDescription`), length(`sampleDescription`) == 1)
        self$`sampleDescription` <- `sampleDescription`
      }
      if (!is.null(`sampleGroupDbId`)) {
        stopifnot(is.character(`sampleGroupDbId`), length(`sampleGroupDbId`) == 1)
        self$`sampleGroupDbId` <- `sampleGroupDbId`
      }
      if (!is.null(`sampleName`)) {
        stopifnot(is.character(`sampleName`), length(`sampleName`) == 1)
        self$`sampleName` <- `sampleName`
      }
      if (!is.null(`samplePUI`)) {
        stopifnot(is.character(`samplePUI`), length(`samplePUI`) == 1)
        self$`samplePUI` <- `samplePUI`
      }
      if (!is.null(`sampleTimestamp`)) {
        stopifnot(is.character(`sampleTimestamp`), length(`sampleTimestamp`) == 1)
        self$`sampleTimestamp` <- `sampleTimestamp`
      }
      if (!is.null(`sampleType`)) {
        stopifnot(is.character(`sampleType`), length(`sampleType`) == 1)
        self$`sampleType` <- `sampleType`
      }
      if (!is.null(`studyDbId`)) {
        stopifnot(is.character(`studyDbId`), length(`studyDbId`) == 1)
        self$`studyDbId` <- `studyDbId`
      }
      if (!is.null(`takenBy`)) {
        stopifnot(is.character(`takenBy`), length(`takenBy`) == 1)
        self$`takenBy` <- `takenBy`
      }
      if (!is.null(`tissueType`)) {
        stopifnot(is.character(`tissueType`), length(`tissueType`) == 1)
        self$`tissueType` <- `tissueType`
      }
      if (!is.null(`trialDbId`)) {
        stopifnot(is.character(`trialDbId`), length(`trialDbId`) == 1)
        self$`trialDbId` <- `trialDbId`
      }
      if (!is.null(`well`)) {
        stopifnot(is.character(`well`), length(`well`) == 1)
        self$`well` <- `well`
      }
    },
    toJSON = function() {
      SampleNewRequestObject <- list()
      if (!is.null(self$`additionalInfo`)) {
        SampleNewRequestObject[['additionalInfo']] <-
          self$`additionalInfo`
      }
      if (!is.null(self$`column`)) {
        SampleNewRequestObject[['column']] <-
          self$`column`
      }
      if (!is.null(self$`externalReferences`)) {
        SampleNewRequestObject[['externalReferences']] <-
          self$`externalReferences`
      }
      if (!is.null(self$`germplasmDbId`)) {
        SampleNewRequestObject[['germplasmDbId']] <-
          self$`germplasmDbId`
      }
      if (!is.null(self$`observationUnitDbId`)) {
        SampleNewRequestObject[['observationUnitDbId']] <-
          self$`observationUnitDbId`
      }
      if (!is.null(self$`plateDbId`)) {
        SampleNewRequestObject[['plateDbId']] <-
          self$`plateDbId`
      }
      if (!is.null(self$`plateName`)) {
        SampleNewRequestObject[['plateName']] <-
          self$`plateName`
      }
      if (!is.null(self$`programDbId`)) {
        SampleNewRequestObject[['programDbId']] <-
          self$`programDbId`
      }
      if (!is.null(self$`row`)) {
        SampleNewRequestObject[['row']] <-
          self$`row`
      }
      if (!is.null(self$`sampleBarcode`)) {
        SampleNewRequestObject[['sampleBarcode']] <-
          self$`sampleBarcode`
      }
      if (!is.null(self$`sampleDescription`)) {
        SampleNewRequestObject[['sampleDescription']] <-
          self$`sampleDescription`
      }
      if (!is.null(self$`sampleGroupDbId`)) {
        SampleNewRequestObject[['sampleGroupDbId']] <-
          self$`sampleGroupDbId`
      }
      if (!is.null(self$`sampleName`)) {
        SampleNewRequestObject[['sampleName']] <-
          self$`sampleName`
      }
      if (!is.null(self$`samplePUI`)) {
        SampleNewRequestObject[['samplePUI']] <-
          self$`samplePUI`
      }
      if (!is.null(self$`sampleTimestamp`)) {
        SampleNewRequestObject[['sampleTimestamp']] <-
          self$`sampleTimestamp`
      }
      if (!is.null(self$`sampleType`)) {
        SampleNewRequestObject[['sampleType']] <-
          self$`sampleType`
      }
      if (!is.null(self$`studyDbId`)) {
        SampleNewRequestObject[['studyDbId']] <-
          self$`studyDbId`
      }
      if (!is.null(self$`takenBy`)) {
        SampleNewRequestObject[['takenBy']] <-
          self$`takenBy`
      }
      if (!is.null(self$`tissueType`)) {
        SampleNewRequestObject[['tissueType']] <-
          self$`tissueType`
      }
      if (!is.null(self$`trialDbId`)) {
        SampleNewRequestObject[['trialDbId']] <-
          self$`trialDbId`
      }
      if (!is.null(self$`well`)) {
        SampleNewRequestObject[['well']] <-
          self$`well`
      }

      SampleNewRequestObject
    },
    fromJSON = function(SampleNewRequestJson) {
      SampleNewRequestObject <- jsonlite::fromJSON(SampleNewRequestJson)
      if (!is.null(SampleNewRequestObject$`additionalInfo`)) {
        self$`additionalInfo` <- ApiClient$new()$deserializeObj(SampleNewRequestObject$`additionalInfo`, "map(character)", loadNamespace("openapi"))
      }
      if (!is.null(SampleNewRequestObject$`column`)) {
        self$`column` <- SampleNewRequestObject$`column`
      }
      if (!is.null(SampleNewRequestObject$`externalReferences`)) {
        self$`externalReferences` <- ApiClient$new()$deserializeObj(SampleNewRequestObject$`externalReferences`, "array[object]", loadNamespace("openapi"))
      }
      if (!is.null(SampleNewRequestObject$`germplasmDbId`)) {
        self$`germplasmDbId` <- SampleNewRequestObject$`germplasmDbId`
      }
      if (!is.null(SampleNewRequestObject$`observationUnitDbId`)) {
        self$`observationUnitDbId` <- SampleNewRequestObject$`observationUnitDbId`
      }
      if (!is.null(SampleNewRequestObject$`plateDbId`)) {
        self$`plateDbId` <- SampleNewRequestObject$`plateDbId`
      }
      if (!is.null(SampleNewRequestObject$`plateName`)) {
        self$`plateName` <- SampleNewRequestObject$`plateName`
      }
      if (!is.null(SampleNewRequestObject$`programDbId`)) {
        self$`programDbId` <- SampleNewRequestObject$`programDbId`
      }
      if (!is.null(SampleNewRequestObject$`row`)) {
        self$`row` <- SampleNewRequestObject$`row`
      }
      if (!is.null(SampleNewRequestObject$`sampleBarcode`)) {
        self$`sampleBarcode` <- SampleNewRequestObject$`sampleBarcode`
      }
      if (!is.null(SampleNewRequestObject$`sampleDescription`)) {
        self$`sampleDescription` <- SampleNewRequestObject$`sampleDescription`
      }
      if (!is.null(SampleNewRequestObject$`sampleGroupDbId`)) {
        self$`sampleGroupDbId` <- SampleNewRequestObject$`sampleGroupDbId`
      }
      if (!is.null(SampleNewRequestObject$`sampleName`)) {
        self$`sampleName` <- SampleNewRequestObject$`sampleName`
      }
      if (!is.null(SampleNewRequestObject$`samplePUI`)) {
        self$`samplePUI` <- SampleNewRequestObject$`samplePUI`
      }
      if (!is.null(SampleNewRequestObject$`sampleTimestamp`)) {
        self$`sampleTimestamp` <- SampleNewRequestObject$`sampleTimestamp`
      }
      if (!is.null(SampleNewRequestObject$`sampleType`)) {
        self$`sampleType` <- SampleNewRequestObject$`sampleType`
      }
      if (!is.null(SampleNewRequestObject$`studyDbId`)) {
        self$`studyDbId` <- SampleNewRequestObject$`studyDbId`
      }
      if (!is.null(SampleNewRequestObject$`takenBy`)) {
        self$`takenBy` <- SampleNewRequestObject$`takenBy`
      }
      if (!is.null(SampleNewRequestObject$`tissueType`)) {
        self$`tissueType` <- SampleNewRequestObject$`tissueType`
      }
      if (!is.null(SampleNewRequestObject$`trialDbId`)) {
        self$`trialDbId` <- SampleNewRequestObject$`trialDbId`
      }
      if (!is.null(SampleNewRequestObject$`well`)) {
        self$`well` <- SampleNewRequestObject$`well`
      }
      self
    },
    toJSONString = function() {
      jsoncontent <- c(
        if (!is.null(self$`additionalInfo`)) {
        sprintf(
        '"additionalInfo":
          "%s"
        ',
        jsonlite::toJSON(lapply(self$`additionalInfo`, function(x){ x }), auto_unbox = TRUE, digits=NA)
        )},
        if (!is.null(self$`column`)) {
        sprintf(
        '"column":
          %d
                ',
        self$`column`
        )},
        if (!is.null(self$`externalReferences`)) {
        sprintf(
        '"externalReferences":
           [%s]
        ',
        paste(unlist(lapply(self$`externalReferences`, function(x) paste0('"', x, '"'))), collapse=",")
        )},
        if (!is.null(self$`germplasmDbId`)) {
        sprintf(
        '"germplasmDbId":
          "%s"
                ',
        self$`germplasmDbId`
        )},
        if (!is.null(self$`observationUnitDbId`)) {
        sprintf(
        '"observationUnitDbId":
          "%s"
                ',
        self$`observationUnitDbId`
        )},
        if (!is.null(self$`plateDbId`)) {
        sprintf(
        '"plateDbId":
          "%s"
                ',
        self$`plateDbId`
        )},
        if (!is.null(self$`plateName`)) {
        sprintf(
        '"plateName":
          "%s"
                ',
        self$`plateName`
        )},
        if (!is.null(self$`programDbId`)) {
        sprintf(
        '"programDbId":
          "%s"
                ',
        self$`programDbId`
        )},
        if (!is.null(self$`row`)) {
        sprintf(
        '"row":
          "%s"
                ',
        self$`row`
        )},
        if (!is.null(self$`sampleBarcode`)) {
        sprintf(
        '"sampleBarcode":
          "%s"
                ',
        self$`sampleBarcode`
        )},
        if (!is.null(self$`sampleDescription`)) {
        sprintf(
        '"sampleDescription":
          "%s"
                ',
        self$`sampleDescription`
        )},
        if (!is.null(self$`sampleGroupDbId`)) {
        sprintf(
        '"sampleGroupDbId":
          "%s"
                ',
        self$`sampleGroupDbId`
        )},
        if (!is.null(self$`sampleName`)) {
        sprintf(
        '"sampleName":
          "%s"
                ',
        self$`sampleName`
        )},
        if (!is.null(self$`samplePUI`)) {
        sprintf(
        '"samplePUI":
          "%s"
                ',
        self$`samplePUI`
        )},
        if (!is.null(self$`sampleTimestamp`)) {
        sprintf(
        '"sampleTimestamp":
          "%s"
                ',
        self$`sampleTimestamp`
        )},
        if (!is.null(self$`sampleType`)) {
        sprintf(
        '"sampleType":
          "%s"
                ',
        self$`sampleType`
        )},
        if (!is.null(self$`studyDbId`)) {
        sprintf(
        '"studyDbId":
          "%s"
                ',
        self$`studyDbId`
        )},
        if (!is.null(self$`takenBy`)) {
        sprintf(
        '"takenBy":
          "%s"
                ',
        self$`takenBy`
        )},
        if (!is.null(self$`tissueType`)) {
        sprintf(
        '"tissueType":
          "%s"
                ',
        self$`tissueType`
        )},
        if (!is.null(self$`trialDbId`)) {
        sprintf(
        '"trialDbId":
          "%s"
                ',
        self$`trialDbId`
        )},
        if (!is.null(self$`well`)) {
        sprintf(
        '"well":
          "%s"
                ',
        self$`well`
        )}
      )
      jsoncontent <- paste(jsoncontent, collapse = ",")
      paste('{', jsoncontent, '}', sep = "")
    },
    fromJSONString = function(SampleNewRequestJson) {
      SampleNewRequestObject <- jsonlite::fromJSON(SampleNewRequestJson)
      self$`additionalInfo` <- ApiClient$new()$deserializeObj(SampleNewRequestObject$`additionalInfo`, "map(character)", loadNamespace("openapi"))
      self$`column` <- SampleNewRequestObject$`column`
      self$`externalReferences` <- ApiClient$new()$deserializeObj(SampleNewRequestObject$`externalReferences`, "array[object]", loadNamespace("openapi"))
      self$`germplasmDbId` <- SampleNewRequestObject$`germplasmDbId`
      self$`observationUnitDbId` <- SampleNewRequestObject$`observationUnitDbId`
      self$`plateDbId` <- SampleNewRequestObject$`plateDbId`
      self$`plateName` <- SampleNewRequestObject$`plateName`
      self$`programDbId` <- SampleNewRequestObject$`programDbId`
      self$`row` <- SampleNewRequestObject$`row`
      self$`sampleBarcode` <- SampleNewRequestObject$`sampleBarcode`
      self$`sampleDescription` <- SampleNewRequestObject$`sampleDescription`
      self$`sampleGroupDbId` <- SampleNewRequestObject$`sampleGroupDbId`
      self$`sampleName` <- SampleNewRequestObject$`sampleName`
      self$`samplePUI` <- SampleNewRequestObject$`samplePUI`
      self$`sampleTimestamp` <- SampleNewRequestObject$`sampleTimestamp`
      self$`sampleType` <- SampleNewRequestObject$`sampleType`
      self$`studyDbId` <- SampleNewRequestObject$`studyDbId`
      self$`takenBy` <- SampleNewRequestObject$`takenBy`
      self$`tissueType` <- SampleNewRequestObject$`tissueType`
      self$`trialDbId` <- SampleNewRequestObject$`trialDbId`
      self$`well` <- SampleNewRequestObject$`well`
      self
    }
  )
)

