# BrAPI-Core
#
# The Breeding API (BrAPI) is a Standardized REST ful Web Service API Specification for communicating Plant Breeding Data. BrAPI allows for easy data sharing between databases and tools involved in plant breeding. <div class=\"brapi-section\"> <h2 class=\"brapi-section-title\">General Reference Documentation</h2> <div class=\"gen-info-link\"><a href=\"https://github.com/plantbreeding/API/blob/master/Specification/GeneralInfo/URL_Structure.md\">URL Structure</a></div> <div class=\"gen-info-link\"><a href=\"https://github.com/plantbreeding/API/blob/master/Specification/GeneralInfo/Response_Structure.md\">Response Structure</a></div> <div class=\"gen-info-link\"><a href=\"https://github.com/plantbreeding/API/blob/master/Specification/GeneralInfo/Date_Time_Encoding.md\">Date/Time Encoding</a></div> <div class=\"gen-info-link\"><a href=\"https://github.com/plantbreeding/API/blob/master/Specification/GeneralInfo/Location_Encoding.md\">Location Encoding</a></div> <div class=\"gen-info-link\"><a href=\"https://github.com/plantbreeding/API/blob/master/Specification/GeneralInfo/Error_Handling.md\">Error Handling</a></div> <div class=\"gen-info-link\"><a href=\"https://github.com/plantbreeding/API/blob/master/Specification/GeneralInfo/Search_Services.md\">Search Services</a></div> </div>  <div class=\"current-brapi-section brapi-section\"> <h2 class=\"brapi-section-title\">BrAPI Core</h2> <div class=\"brapi-section-description\">The BrAPI Core module contains high level entities used for organization and management. This includes Programs, Trials, Studies, Locations, People, and Lists</div> <div class=\"version-number\">V2.0</div> <div class=\"link-btn\"><a href=\"https://github.com/plantbreeding/API/tree/master/Specification/BrAPI-Core\">GitHub</a></div> <div class=\"link-btn\"><a href=\"https://app.swaggerhub.com/apis/PlantBreedingAPI/BrAPI-Core\">SwaggerHub</a></div> <div class=\"link-btn\"><a href=\"https://brapicore.docs.apiary.io\">Apiary</a></div> <div class=\"stop-float\"></div> </div>  <div class=\"brapi-section\"> <h2 class=\"brapi-section-title\">BrAPI Phenotyping</h2> <div class=\"brapi-section-description\">The BrAPI Phenotyping module contains entities related to phenotypic observations. This includes Observation Units, Observations, Observation Variables, Traits, Scales, Methods, and Images</div> <div class=\"version-number\">V2.0</div> <div class=\"link-btn\"><a href=\"https://github.com/plantbreeding/API/tree/master/Specification/BrAPI-Phenotyping\">GitHub</a></div> <div class=\"link-btn\"><a href=\"https://app.swaggerhub.com/apis/PlantBreedingAPI/BrAPI-Phenotyping\">SwaggerHub</a></div> <div class=\"link-btn\"><a href=\"https://brapiphenotyping.docs.apiary.io\">Apiary</a></div> <div class=\"stop-float\"></div> </div>  <div class=\"brapi-section\"> <h2 class=\"brapi-section-title\">BrAPI Genotyping</h2> <div class=\"brapi-section-description\">The BrAPI Genotyping module contains entities related to genotyping analysis. This includes Samples, Markers, Variant Sets, Variants, Call Sets, Calls, References, Reads, and Vendor Orders</div> <div class=\"version-number\">V2.0</div> <div class=\"link-btn\"><a href=\"https://github.com/plantbreeding/API/tree/master/Specification/BrAPI-Genotyping\">GitHub</a></div> <div class=\"link-btn\"><a href=\"https://app.swaggerhub.com/apis/PlantBreedingAPI/BrAPI-Genotyping\">SwaggerHub</a></div> <div class=\"link-btn\"><a href=\"https://brapigenotyping.docs.apiary.io\">Apiary</a></div> <div class=\"stop-float\"></div> </div>  <div class=\"brapi-section\"> <h2 class=\"brapi-section-title\">BrAPI Germplasm</h2> <div class=\"brapi-section-description\">The BrAPI Germplasm module contains entities related to germplasm management. This includes Germplasm, Germplasm Attributes, Seed Lots, Crosses, Pedigree, and Progeny</div> <div class=\"version-number\">V2.0</div> <div class=\"link-btn\"><a href=\"https://github.com/plantbreeding/API/tree/master/Specification/BrAPI-Germplasm\">GitHub</a></div> <div class=\"link-btn\"><a href=\"https://app.swaggerhub.com/apis/PlantBreedingAPI/BrAPI-Germplasm\">SwaggerHub</a></div> <div class=\"link-btn\"><a href=\"https://brapigermplasm.docs.apiary.io\">Apiary</a></div> <div class=\"stop-float\"></div> </div>  <style> .link-btn{ float: left;  margin: 2px 10px 0 0;  padding: 0 5px;  border-radius: 5px;  background-color: #ddd; } .stop-float{   clear: both; } .version-number{   float: left;    margin: 5px 10px 0 5px; } .brapi-section-title{   margin: 0 10px 0 0;   font-size: 20px; } .current-brapi-section{   font-weight: bolder;   border-radius: 5px;    background-color: #ddd; } .brapi-section{   padding: 5px 5px;  } .brapi-section-description{   margin: 5px 0 0 5px; } </style>
#
# The version of the OpenAPI document: 2.0
# 
# Generated by: https://openapi-generator.tech

#' @docType class
#' @title Trait
#'
#' @description Trait Class
#'
#' @format An \code{R6Class} generator object
#'
#' @field additionalInfo  named list( character ) [optional]
#'
#' @field alternativeAbbreviations  list( character ) [optional]
#'
#' @field attribute  character [optional]
#'
#' @field entity  character [optional]
#'
#' @field externalReferences ExternalReferences list( object ) [optional]
#'
#' @field mainAbbreviation  character [optional]
#'
#' @field ontologyReference  \link{OntologyReference} [optional]
#'
#' @field status  character [optional]
#'
#' @field synonyms  list( character ) [optional]
#'
#' @field traitClass  character [optional]
#'
#' @field traitDescription  character [optional]
#'
#' @field traitName  character [optional]
#'
#' @field traitDbId  character [optional]
#'
#' @importFrom R6 R6Class
#' @importFrom jsonlite fromJSON toJSON
#' @export
Trait <- R6::R6Class(
  'Trait',
  public = list(
    `additionalInfo` = NULL,
    `alternativeAbbreviations` = NULL,
    `attribute` = NULL,
    `entity` = NULL,
    `externalReferences` = NULL,
    `mainAbbreviation` = NULL,
    `ontologyReference` = NULL,
    `status` = NULL,
    `synonyms` = NULL,
    `traitClass` = NULL,
    `traitDescription` = NULL,
    `traitName` = NULL,
    `traitDbId` = NULL,
    initialize = function(
        `additionalInfo`=NULL, `alternativeAbbreviations`=NULL, `attribute`=NULL, `entity`=NULL, `externalReferences`=NULL, `mainAbbreviation`=NULL, `ontologyReference`=NULL, `status`=NULL, `synonyms`=NULL, `traitClass`=NULL, `traitDescription`=NULL, `traitName`=NULL, `traitDbId`=NULL, ...
    ) {
      local.optional.var <- list(...)
      if (!is.null(`additionalInfo`)) {
        stopifnot(is.vector(`additionalInfo`), length(`additionalInfo`) != 0)
        sapply(`additionalInfo`, function(x) stopifnot(is.character(x)))
        self$`additionalInfo` <- `additionalInfo`
      }
      if (!is.null(`alternativeAbbreviations`)) {
        stopifnot(is.vector(`alternativeAbbreviations`), length(`alternativeAbbreviations`) != 0)
        sapply(`alternativeAbbreviations`, function(x) stopifnot(is.character(x)))
        self$`alternativeAbbreviations` <- `alternativeAbbreviations`
      }
      if (!is.null(`attribute`)) {
        stopifnot(is.character(`attribute`), length(`attribute`) == 1)
        self$`attribute` <- `attribute`
      }
      if (!is.null(`entity`)) {
        stopifnot(is.character(`entity`), length(`entity`) == 1)
        self$`entity` <- `entity`
      }
      if (!is.null(`externalReferences`)) {
        stopifnot(is.vector(`externalReferences`), length(`externalReferences`) != 0)
        sapply(`externalReferences`, function(x) stopifnot(is.character(x)))
        self$`externalReferences` <- `externalReferences`
      }
      if (!is.null(`mainAbbreviation`)) {
        stopifnot(is.character(`mainAbbreviation`), length(`mainAbbreviation`) == 1)
        self$`mainAbbreviation` <- `mainAbbreviation`
      }
      if (!is.null(`ontologyReference`)) {
        stopifnot(R6::is.R6(`ontologyReference`))
        self$`ontologyReference` <- `ontologyReference`
      }
      if (!is.null(`status`)) {
        stopifnot(is.character(`status`), length(`status`) == 1)
        self$`status` <- `status`
      }
      if (!is.null(`synonyms`)) {
        stopifnot(is.vector(`synonyms`), length(`synonyms`) != 0)
        sapply(`synonyms`, function(x) stopifnot(is.character(x)))
        self$`synonyms` <- `synonyms`
      }
      if (!is.null(`traitClass`)) {
        stopifnot(is.character(`traitClass`), length(`traitClass`) == 1)
        self$`traitClass` <- `traitClass`
      }
      if (!is.null(`traitDescription`)) {
        stopifnot(is.character(`traitDescription`), length(`traitDescription`) == 1)
        self$`traitDescription` <- `traitDescription`
      }
      if (!is.null(`traitName`)) {
        stopifnot(is.character(`traitName`), length(`traitName`) == 1)
        self$`traitName` <- `traitName`
      }
      if (!is.null(`traitDbId`)) {
        stopifnot(is.character(`traitDbId`), length(`traitDbId`) == 1)
        self$`traitDbId` <- `traitDbId`
      }
    },
    toJSON = function() {
      TraitObject <- list()
      if (!is.null(self$`additionalInfo`)) {
        TraitObject[['additionalInfo']] <-
          self$`additionalInfo`
      }
      if (!is.null(self$`alternativeAbbreviations`)) {
        TraitObject[['alternativeAbbreviations']] <-
          self$`alternativeAbbreviations`
      }
      if (!is.null(self$`attribute`)) {
        TraitObject[['attribute']] <-
          self$`attribute`
      }
      if (!is.null(self$`entity`)) {
        TraitObject[['entity']] <-
          self$`entity`
      }
      if (!is.null(self$`externalReferences`)) {
        TraitObject[['externalReferences']] <-
          self$`externalReferences`
      }
      if (!is.null(self$`mainAbbreviation`)) {
        TraitObject[['mainAbbreviation']] <-
          self$`mainAbbreviation`
      }
      if (!is.null(self$`ontologyReference`)) {
        TraitObject[['ontologyReference']] <-
          self$`ontologyReference`$toJSON()
      }
      if (!is.null(self$`status`)) {
        TraitObject[['status']] <-
          self$`status`
      }
      if (!is.null(self$`synonyms`)) {
        TraitObject[['synonyms']] <-
          self$`synonyms`
      }
      if (!is.null(self$`traitClass`)) {
        TraitObject[['traitClass']] <-
          self$`traitClass`
      }
      if (!is.null(self$`traitDescription`)) {
        TraitObject[['traitDescription']] <-
          self$`traitDescription`
      }
      if (!is.null(self$`traitName`)) {
        TraitObject[['traitName']] <-
          self$`traitName`
      }
      if (!is.null(self$`traitDbId`)) {
        TraitObject[['traitDbId']] <-
          self$`traitDbId`
      }

      TraitObject
    },
    fromJSON = function(TraitJson) {
      TraitObject <- jsonlite::fromJSON(TraitJson)
      if (!is.null(TraitObject$`additionalInfo`)) {
        self$`additionalInfo` <- ApiClient$new()$deserializeObj(TraitObject$`additionalInfo`, "map(character)", loadNamespace("openapi"))
      }
      if (!is.null(TraitObject$`alternativeAbbreviations`)) {
        self$`alternativeAbbreviations` <- ApiClient$new()$deserializeObj(TraitObject$`alternativeAbbreviations`, "array[character]", loadNamespace("openapi"))
      }
      if (!is.null(TraitObject$`attribute`)) {
        self$`attribute` <- TraitObject$`attribute`
      }
      if (!is.null(TraitObject$`entity`)) {
        self$`entity` <- TraitObject$`entity`
      }
      if (!is.null(TraitObject$`externalReferences`)) {
        self$`externalReferences` <- ApiClient$new()$deserializeObj(TraitObject$`externalReferences`, "array[object]", loadNamespace("openapi"))
      }
      if (!is.null(TraitObject$`mainAbbreviation`)) {
        self$`mainAbbreviation` <- TraitObject$`mainAbbreviation`
      }
      if (!is.null(TraitObject$`ontologyReference`)) {
        ontologyReferenceObject <- OntologyReference$new()
        ontologyReferenceObject$fromJSON(jsonlite::toJSON(TraitObject$ontologyReference, auto_unbox = TRUE, digits = NA))
        self$`ontologyReference` <- ontologyReferenceObject
      }
      if (!is.null(TraitObject$`status`)) {
        self$`status` <- TraitObject$`status`
      }
      if (!is.null(TraitObject$`synonyms`)) {
        self$`synonyms` <- ApiClient$new()$deserializeObj(TraitObject$`synonyms`, "array[character]", loadNamespace("openapi"))
      }
      if (!is.null(TraitObject$`traitClass`)) {
        self$`traitClass` <- TraitObject$`traitClass`
      }
      if (!is.null(TraitObject$`traitDescription`)) {
        self$`traitDescription` <- TraitObject$`traitDescription`
      }
      if (!is.null(TraitObject$`traitName`)) {
        self$`traitName` <- TraitObject$`traitName`
      }
      if (!is.null(TraitObject$`traitDbId`)) {
        self$`traitDbId` <- TraitObject$`traitDbId`
      }
      self
    },
    toJSONString = function() {
      jsoncontent <- c(
        if (!is.null(self$`additionalInfo`)) {
        sprintf(
        '"additionalInfo":
          "%s"
        ',
        jsonlite::toJSON(lapply(self$`additionalInfo`, function(x){ x }), auto_unbox = TRUE, digits=NA)
        )},
        if (!is.null(self$`alternativeAbbreviations`)) {
        sprintf(
        '"alternativeAbbreviations":
           [%s]
        ',
        paste(unlist(lapply(self$`alternativeAbbreviations`, function(x) paste0('"', x, '"'))), collapse=",")
        )},
        if (!is.null(self$`attribute`)) {
        sprintf(
        '"attribute":
          "%s"
                ',
        self$`attribute`
        )},
        if (!is.null(self$`entity`)) {
        sprintf(
        '"entity":
          "%s"
                ',
        self$`entity`
        )},
        if (!is.null(self$`externalReferences`)) {
        sprintf(
        '"externalReferences":
           [%s]
        ',
        paste(unlist(lapply(self$`externalReferences`, function(x) paste0('"', x, '"'))), collapse=",")
        )},
        if (!is.null(self$`mainAbbreviation`)) {
        sprintf(
        '"mainAbbreviation":
          "%s"
                ',
        self$`mainAbbreviation`
        )},
        if (!is.null(self$`ontologyReference`)) {
        sprintf(
        '"ontologyReference":
        %s
        ',
        jsonlite::toJSON(self$`ontologyReference`$toJSON(), auto_unbox=TRUE, digits = NA)
        )},
        if (!is.null(self$`status`)) {
        sprintf(
        '"status":
          "%s"
                ',
        self$`status`
        )},
        if (!is.null(self$`synonyms`)) {
        sprintf(
        '"synonyms":
           [%s]
        ',
        paste(unlist(lapply(self$`synonyms`, function(x) paste0('"', x, '"'))), collapse=",")
        )},
        if (!is.null(self$`traitClass`)) {
        sprintf(
        '"traitClass":
          "%s"
                ',
        self$`traitClass`
        )},
        if (!is.null(self$`traitDescription`)) {
        sprintf(
        '"traitDescription":
          "%s"
                ',
        self$`traitDescription`
        )},
        if (!is.null(self$`traitName`)) {
        sprintf(
        '"traitName":
          "%s"
                ',
        self$`traitName`
        )},
        if (!is.null(self$`traitDbId`)) {
        sprintf(
        '"traitDbId":
          "%s"
                ',
        self$`traitDbId`
        )}
      )
      jsoncontent <- paste(jsoncontent, collapse = ",")
      paste('{', jsoncontent, '}', sep = "")
    },
    fromJSONString = function(TraitJson) {
      TraitObject <- jsonlite::fromJSON(TraitJson)
      self$`additionalInfo` <- ApiClient$new()$deserializeObj(TraitObject$`additionalInfo`, "map(character)", loadNamespace("openapi"))
      self$`alternativeAbbreviations` <- ApiClient$new()$deserializeObj(TraitObject$`alternativeAbbreviations`, "array[character]", loadNamespace("openapi"))
      self$`attribute` <- TraitObject$`attribute`
      self$`entity` <- TraitObject$`entity`
      self$`externalReferences` <- ApiClient$new()$deserializeObj(TraitObject$`externalReferences`, "array[object]", loadNamespace("openapi"))
      self$`mainAbbreviation` <- TraitObject$`mainAbbreviation`
      self$`ontologyReference` <- OntologyReference$new()$fromJSON(jsonlite::toJSON(TraitObject$ontologyReference, auto_unbox = TRUE, digits = NA))
      self$`status` <- TraitObject$`status`
      self$`synonyms` <- ApiClient$new()$deserializeObj(TraitObject$`synonyms`, "array[character]", loadNamespace("openapi"))
      self$`traitClass` <- TraitObject$`traitClass`
      self$`traitDescription` <- TraitObject$`traitDescription`
      self$`traitName` <- TraitObject$`traitName`
      self$`traitDbId` <- TraitObject$`traitDbId`
      self
    }
  )
)

