# BrAPI-Core
#
# The Breeding API (BrAPI) is a Standardized REST ful Web Service API Specification for communicating Plant Breeding Data. BrAPI allows for easy data sharing between databases and tools involved in plant breeding. <div class=\"brapi-section\"> <h2 class=\"brapi-section-title\">General Reference Documentation</h2> <div class=\"gen-info-link\"><a href=\"https://github.com/plantbreeding/API/blob/master/Specification/GeneralInfo/URL_Structure.md\">URL Structure</a></div> <div class=\"gen-info-link\"><a href=\"https://github.com/plantbreeding/API/blob/master/Specification/GeneralInfo/Response_Structure.md\">Response Structure</a></div> <div class=\"gen-info-link\"><a href=\"https://github.com/plantbreeding/API/blob/master/Specification/GeneralInfo/Date_Time_Encoding.md\">Date/Time Encoding</a></div> <div class=\"gen-info-link\"><a href=\"https://github.com/plantbreeding/API/blob/master/Specification/GeneralInfo/Location_Encoding.md\">Location Encoding</a></div> <div class=\"gen-info-link\"><a href=\"https://github.com/plantbreeding/API/blob/master/Specification/GeneralInfo/Error_Handling.md\">Error Handling</a></div> <div class=\"gen-info-link\"><a href=\"https://github.com/plantbreeding/API/blob/master/Specification/GeneralInfo/Search_Services.md\">Search Services</a></div> </div>  <div class=\"current-brapi-section brapi-section\"> <h2 class=\"brapi-section-title\">BrAPI Core</h2> <div class=\"brapi-section-description\">The BrAPI Core module contains high level entities used for organization and management. This includes Programs, Trials, Studies, Locations, People, and Lists</div> <div class=\"version-number\">V2.0</div> <div class=\"link-btn\"><a href=\"https://github.com/plantbreeding/API/tree/master/Specification/BrAPI-Core\">GitHub</a></div> <div class=\"link-btn\"><a href=\"https://app.swaggerhub.com/apis/PlantBreedingAPI/BrAPI-Core\">SwaggerHub</a></div> <div class=\"link-btn\"><a href=\"https://brapicore.docs.apiary.io\">Apiary</a></div> <div class=\"stop-float\"></div> </div>  <div class=\"brapi-section\"> <h2 class=\"brapi-section-title\">BrAPI Phenotyping</h2> <div class=\"brapi-section-description\">The BrAPI Phenotyping module contains entities related to phenotypic observations. This includes Observation Units, Observations, Observation Variables, Traits, Scales, Methods, and Images</div> <div class=\"version-number\">V2.0</div> <div class=\"link-btn\"><a href=\"https://github.com/plantbreeding/API/tree/master/Specification/BrAPI-Phenotyping\">GitHub</a></div> <div class=\"link-btn\"><a href=\"https://app.swaggerhub.com/apis/PlantBreedingAPI/BrAPI-Phenotyping\">SwaggerHub</a></div> <div class=\"link-btn\"><a href=\"https://brapiphenotyping.docs.apiary.io\">Apiary</a></div> <div class=\"stop-float\"></div> </div>  <div class=\"brapi-section\"> <h2 class=\"brapi-section-title\">BrAPI Genotyping</h2> <div class=\"brapi-section-description\">The BrAPI Genotyping module contains entities related to genotyping analysis. This includes Samples, Markers, Variant Sets, Variants, Call Sets, Calls, References, Reads, and Vendor Orders</div> <div class=\"version-number\">V2.0</div> <div class=\"link-btn\"><a href=\"https://github.com/plantbreeding/API/tree/master/Specification/BrAPI-Genotyping\">GitHub</a></div> <div class=\"link-btn\"><a href=\"https://app.swaggerhub.com/apis/PlantBreedingAPI/BrAPI-Genotyping\">SwaggerHub</a></div> <div class=\"link-btn\"><a href=\"https://brapigenotyping.docs.apiary.io\">Apiary</a></div> <div class=\"stop-float\"></div> </div>  <div class=\"brapi-section\"> <h2 class=\"brapi-section-title\">BrAPI Germplasm</h2> <div class=\"brapi-section-description\">The BrAPI Germplasm module contains entities related to germplasm management. This includes Germplasm, Germplasm Attributes, Seed Lots, Crosses, Pedigree, and Progeny</div> <div class=\"version-number\">V2.0</div> <div class=\"link-btn\"><a href=\"https://github.com/plantbreeding/API/tree/master/Specification/BrAPI-Germplasm\">GitHub</a></div> <div class=\"link-btn\"><a href=\"https://app.swaggerhub.com/apis/PlantBreedingAPI/BrAPI-Germplasm\">SwaggerHub</a></div> <div class=\"link-btn\"><a href=\"https://brapigermplasm.docs.apiary.io\">Apiary</a></div> <div class=\"stop-float\"></div> </div>  <style> .link-btn{ float: left;  margin: 2px 10px 0 0;  padding: 0 5px;  border-radius: 5px;  background-color: #ddd; } .stop-float{   clear: both; } .version-number{   float: left;    margin: 5px 10px 0 5px; } .brapi-section-title{   margin: 0 10px 0 0;   font-size: 20px; } .current-brapi-section{   font-weight: bolder;   border-radius: 5px;    background-color: #ddd; } .brapi-section{   padding: 5px 5px;  } .brapi-section-description{   margin: 5px 0 0 5px; } </style>
#
# The version of the OpenAPI document: 2.0
# 
# Generated by: https://openapi-generator.tech

#' @docType class
#' @title Variant
#'
#' @description Variant Class
#'
#' @format An \code{R6Class} generator object
#'
#' @field additionalInfo  named list( character ) [optional]
#'
#' @field alternate_bases  list( character ) [optional]
#'
#' @field ciend  list( integer ) [optional]
#'
#' @field cipos  list( integer ) [optional]
#'
#' @field created  character [optional]
#'
#' @field end  integer [optional]
#'
#' @field filtersApplied  character [optional]
#'
#' @field filtersFailed  list( character ) [optional]
#'
#' @field filtersPassed  character [optional]
#'
#' @field referenceBases  character [optional]
#'
#' @field referenceName  character [optional]
#'
#' @field start  integer [optional]
#'
#' @field svlen  integer [optional]
#'
#' @field updated  character [optional]
#'
#' @field variantDbId  character [optional]
#'
#' @field variantNames  list( character ) [optional]
#'
#' @field variantSetDbId  list( character ) [optional]
#'
#' @field variantType  character [optional]
#'
#' @importFrom R6 R6Class
#' @importFrom jsonlite fromJSON toJSON
#' @export
Variant <- R6::R6Class(
  'Variant',
  public = list(
    `additionalInfo` = NULL,
    `alternate_bases` = NULL,
    `ciend` = NULL,
    `cipos` = NULL,
    `created` = NULL,
    `end` = NULL,
    `filtersApplied` = NULL,
    `filtersFailed` = NULL,
    `filtersPassed` = NULL,
    `referenceBases` = NULL,
    `referenceName` = NULL,
    `start` = NULL,
    `svlen` = NULL,
    `updated` = NULL,
    `variantDbId` = NULL,
    `variantNames` = NULL,
    `variantSetDbId` = NULL,
    `variantType` = NULL,
    initialize = function(
        `additionalInfo`=NULL, `alternate_bases`=NULL, `ciend`=NULL, `cipos`=NULL, `created`=NULL, `end`=NULL, `filtersApplied`=NULL, `filtersFailed`=NULL, `filtersPassed`=NULL, `referenceBases`=NULL, `referenceName`=NULL, `start`=NULL, `svlen`=NULL, `updated`=NULL, `variantDbId`=NULL, `variantNames`=NULL, `variantSetDbId`=NULL, `variantType`=NULL, ...
    ) {
      local.optional.var <- list(...)
      if (!is.null(`additionalInfo`)) {
        stopifnot(is.vector(`additionalInfo`), length(`additionalInfo`) != 0)
        sapply(`additionalInfo`, function(x) stopifnot(is.character(x)))
        self$`additionalInfo` <- `additionalInfo`
      }
      if (!is.null(`alternate_bases`)) {
        stopifnot(is.vector(`alternate_bases`), length(`alternate_bases`) != 0)
        sapply(`alternate_bases`, function(x) stopifnot(is.character(x)))
        self$`alternate_bases` <- `alternate_bases`
      }
      if (!is.null(`ciend`)) {
        stopifnot(is.vector(`ciend`), length(`ciend`) != 0)
        sapply(`ciend`, function(x) stopifnot(is.character(x)))
        self$`ciend` <- `ciend`
      }
      if (!is.null(`cipos`)) {
        stopifnot(is.vector(`cipos`), length(`cipos`) != 0)
        sapply(`cipos`, function(x) stopifnot(is.character(x)))
        self$`cipos` <- `cipos`
      }
      if (!is.null(`created`)) {
        stopifnot(is.character(`created`), length(`created`) == 1)
        self$`created` <- `created`
      }
      if (!is.null(`end`)) {
        stopifnot(is.numeric(`end`), length(`end`) == 1)
        self$`end` <- `end`
      }
      if (!is.null(`filtersApplied`)) {
        self$`filtersApplied` <- `filtersApplied`
      }
      if (!is.null(`filtersFailed`)) {
        stopifnot(is.vector(`filtersFailed`), length(`filtersFailed`) != 0)
        sapply(`filtersFailed`, function(x) stopifnot(is.character(x)))
        self$`filtersFailed` <- `filtersFailed`
      }
      if (!is.null(`filtersPassed`)) {
        self$`filtersPassed` <- `filtersPassed`
      }
      if (!is.null(`referenceBases`)) {
        stopifnot(is.character(`referenceBases`), length(`referenceBases`) == 1)
        self$`referenceBases` <- `referenceBases`
      }
      if (!is.null(`referenceName`)) {
        stopifnot(is.character(`referenceName`), length(`referenceName`) == 1)
        self$`referenceName` <- `referenceName`
      }
      if (!is.null(`start`)) {
        stopifnot(is.numeric(`start`), length(`start`) == 1)
        self$`start` <- `start`
      }
      if (!is.null(`svlen`)) {
        stopifnot(is.numeric(`svlen`), length(`svlen`) == 1)
        self$`svlen` <- `svlen`
      }
      if (!is.null(`updated`)) {
        stopifnot(is.character(`updated`), length(`updated`) == 1)
        self$`updated` <- `updated`
      }
      if (!is.null(`variantDbId`)) {
        stopifnot(is.character(`variantDbId`), length(`variantDbId`) == 1)
        self$`variantDbId` <- `variantDbId`
      }
      if (!is.null(`variantNames`)) {
        stopifnot(is.vector(`variantNames`), length(`variantNames`) != 0)
        sapply(`variantNames`, function(x) stopifnot(is.character(x)))
        self$`variantNames` <- `variantNames`
      }
      if (!is.null(`variantSetDbId`)) {
        stopifnot(is.vector(`variantSetDbId`), length(`variantSetDbId`) != 0)
        sapply(`variantSetDbId`, function(x) stopifnot(is.character(x)))
        self$`variantSetDbId` <- `variantSetDbId`
      }
      if (!is.null(`variantType`)) {
        stopifnot(is.character(`variantType`), length(`variantType`) == 1)
        self$`variantType` <- `variantType`
      }
    },
    toJSON = function() {
      VariantObject <- list()
      if (!is.null(self$`additionalInfo`)) {
        VariantObject[['additionalInfo']] <-
          self$`additionalInfo`
      }
      if (!is.null(self$`alternate_bases`)) {
        VariantObject[['alternate_bases']] <-
          self$`alternate_bases`
      }
      if (!is.null(self$`ciend`)) {
        VariantObject[['ciend']] <-
          self$`ciend`
      }
      if (!is.null(self$`cipos`)) {
        VariantObject[['cipos']] <-
          self$`cipos`
      }
      if (!is.null(self$`created`)) {
        VariantObject[['created']] <-
          self$`created`
      }
      if (!is.null(self$`end`)) {
        VariantObject[['end']] <-
          self$`end`
      }
      if (!is.null(self$`filtersApplied`)) {
        VariantObject[['filtersApplied']] <-
          self$`filtersApplied`
      }
      if (!is.null(self$`filtersFailed`)) {
        VariantObject[['filtersFailed']] <-
          self$`filtersFailed`
      }
      if (!is.null(self$`filtersPassed`)) {
        VariantObject[['filtersPassed']] <-
          self$`filtersPassed`
      }
      if (!is.null(self$`referenceBases`)) {
        VariantObject[['referenceBases']] <-
          self$`referenceBases`
      }
      if (!is.null(self$`referenceName`)) {
        VariantObject[['referenceName']] <-
          self$`referenceName`
      }
      if (!is.null(self$`start`)) {
        VariantObject[['start']] <-
          self$`start`
      }
      if (!is.null(self$`svlen`)) {
        VariantObject[['svlen']] <-
          self$`svlen`
      }
      if (!is.null(self$`updated`)) {
        VariantObject[['updated']] <-
          self$`updated`
      }
      if (!is.null(self$`variantDbId`)) {
        VariantObject[['variantDbId']] <-
          self$`variantDbId`
      }
      if (!is.null(self$`variantNames`)) {
        VariantObject[['variantNames']] <-
          self$`variantNames`
      }
      if (!is.null(self$`variantSetDbId`)) {
        VariantObject[['variantSetDbId']] <-
          self$`variantSetDbId`
      }
      if (!is.null(self$`variantType`)) {
        VariantObject[['variantType']] <-
          self$`variantType`
      }

      VariantObject
    },
    fromJSON = function(VariantJson) {
      VariantObject <- jsonlite::fromJSON(VariantJson)
      if (!is.null(VariantObject$`additionalInfo`)) {
        self$`additionalInfo` <- ApiClient$new()$deserializeObj(VariantObject$`additionalInfo`, "map(character)", loadNamespace("openapi"))
      }
      if (!is.null(VariantObject$`alternate_bases`)) {
        self$`alternate_bases` <- ApiClient$new()$deserializeObj(VariantObject$`alternate_bases`, "array[character]", loadNamespace("openapi"))
      }
      if (!is.null(VariantObject$`ciend`)) {
        self$`ciend` <- ApiClient$new()$deserializeObj(VariantObject$`ciend`, "array[integer]", loadNamespace("openapi"))
      }
      if (!is.null(VariantObject$`cipos`)) {
        self$`cipos` <- ApiClient$new()$deserializeObj(VariantObject$`cipos`, "array[integer]", loadNamespace("openapi"))
      }
      if (!is.null(VariantObject$`created`)) {
        self$`created` <- VariantObject$`created`
      }
      if (!is.null(VariantObject$`end`)) {
        self$`end` <- VariantObject$`end`
      }
      if (!is.null(VariantObject$`filtersApplied`)) {
        self$`filtersApplied` <- VariantObject$`filtersApplied`
      }
      if (!is.null(VariantObject$`filtersFailed`)) {
        self$`filtersFailed` <- ApiClient$new()$deserializeObj(VariantObject$`filtersFailed`, "array[character]", loadNamespace("openapi"))
      }
      if (!is.null(VariantObject$`filtersPassed`)) {
        self$`filtersPassed` <- VariantObject$`filtersPassed`
      }
      if (!is.null(VariantObject$`referenceBases`)) {
        self$`referenceBases` <- VariantObject$`referenceBases`
      }
      if (!is.null(VariantObject$`referenceName`)) {
        self$`referenceName` <- VariantObject$`referenceName`
      }
      if (!is.null(VariantObject$`start`)) {
        self$`start` <- VariantObject$`start`
      }
      if (!is.null(VariantObject$`svlen`)) {
        self$`svlen` <- VariantObject$`svlen`
      }
      if (!is.null(VariantObject$`updated`)) {
        self$`updated` <- VariantObject$`updated`
      }
      if (!is.null(VariantObject$`variantDbId`)) {
        self$`variantDbId` <- VariantObject$`variantDbId`
      }
      if (!is.null(VariantObject$`variantNames`)) {
        self$`variantNames` <- ApiClient$new()$deserializeObj(VariantObject$`variantNames`, "array[character]", loadNamespace("openapi"))
      }
      if (!is.null(VariantObject$`variantSetDbId`)) {
        self$`variantSetDbId` <- ApiClient$new()$deserializeObj(VariantObject$`variantSetDbId`, "array[character]", loadNamespace("openapi"))
      }
      if (!is.null(VariantObject$`variantType`)) {
        self$`variantType` <- VariantObject$`variantType`
      }
      self
    },
    toJSONString = function() {
      jsoncontent <- c(
        if (!is.null(self$`additionalInfo`)) {
        sprintf(
        '"additionalInfo":
          "%s"
        ',
        jsonlite::toJSON(lapply(self$`additionalInfo`, function(x){ x }), auto_unbox = TRUE, digits=NA)
        )},
        if (!is.null(self$`alternate_bases`)) {
        sprintf(
        '"alternate_bases":
           [%s]
        ',
        paste(unlist(lapply(self$`alternate_bases`, function(x) paste0('"', x, '"'))), collapse=",")
        )},
        if (!is.null(self$`ciend`)) {
        sprintf(
        '"ciend":
           [%s]
        ',
        paste(unlist(lapply(self$`ciend`, function(x) paste0('"', x, '"'))), collapse=",")
        )},
        if (!is.null(self$`cipos`)) {
        sprintf(
        '"cipos":
           [%s]
        ',
        paste(unlist(lapply(self$`cipos`, function(x) paste0('"', x, '"'))), collapse=",")
        )},
        if (!is.null(self$`created`)) {
        sprintf(
        '"created":
          "%s"
                ',
        self$`created`
        )},
        if (!is.null(self$`end`)) {
        sprintf(
        '"end":
          %d
                ',
        self$`end`
        )},
        if (!is.null(self$`filtersApplied`)) {
        sprintf(
        '"filtersApplied":
          "%s"
                ',
        self$`filtersApplied`
        )},
        if (!is.null(self$`filtersFailed`)) {
        sprintf(
        '"filtersFailed":
           [%s]
        ',
        paste(unlist(lapply(self$`filtersFailed`, function(x) paste0('"', x, '"'))), collapse=",")
        )},
        if (!is.null(self$`filtersPassed`)) {
        sprintf(
        '"filtersPassed":
          "%s"
                ',
        self$`filtersPassed`
        )},
        if (!is.null(self$`referenceBases`)) {
        sprintf(
        '"referenceBases":
          "%s"
                ',
        self$`referenceBases`
        )},
        if (!is.null(self$`referenceName`)) {
        sprintf(
        '"referenceName":
          "%s"
                ',
        self$`referenceName`
        )},
        if (!is.null(self$`start`)) {
        sprintf(
        '"start":
          %d
                ',
        self$`start`
        )},
        if (!is.null(self$`svlen`)) {
        sprintf(
        '"svlen":
          %d
                ',
        self$`svlen`
        )},
        if (!is.null(self$`updated`)) {
        sprintf(
        '"updated":
          "%s"
                ',
        self$`updated`
        )},
        if (!is.null(self$`variantDbId`)) {
        sprintf(
        '"variantDbId":
          "%s"
                ',
        self$`variantDbId`
        )},
        if (!is.null(self$`variantNames`)) {
        sprintf(
        '"variantNames":
           [%s]
        ',
        paste(unlist(lapply(self$`variantNames`, function(x) paste0('"', x, '"'))), collapse=",")
        )},
        if (!is.null(self$`variantSetDbId`)) {
        sprintf(
        '"variantSetDbId":
           [%s]
        ',
        paste(unlist(lapply(self$`variantSetDbId`, function(x) paste0('"', x, '"'))), collapse=",")
        )},
        if (!is.null(self$`variantType`)) {
        sprintf(
        '"variantType":
          "%s"
                ',
        self$`variantType`
        )}
      )
      jsoncontent <- paste(jsoncontent, collapse = ",")
      paste('{', jsoncontent, '}', sep = "")
    },
    fromJSONString = function(VariantJson) {
      VariantObject <- jsonlite::fromJSON(VariantJson)
      self$`additionalInfo` <- ApiClient$new()$deserializeObj(VariantObject$`additionalInfo`, "map(character)", loadNamespace("openapi"))
      self$`alternate_bases` <- ApiClient$new()$deserializeObj(VariantObject$`alternate_bases`, "array[character]", loadNamespace("openapi"))
      self$`ciend` <- ApiClient$new()$deserializeObj(VariantObject$`ciend`, "array[integer]", loadNamespace("openapi"))
      self$`cipos` <- ApiClient$new()$deserializeObj(VariantObject$`cipos`, "array[integer]", loadNamespace("openapi"))
      self$`created` <- VariantObject$`created`
      self$`end` <- VariantObject$`end`
      self$`filtersApplied` <- VariantObject$`filtersApplied`
      self$`filtersFailed` <- ApiClient$new()$deserializeObj(VariantObject$`filtersFailed`, "array[character]", loadNamespace("openapi"))
      self$`filtersPassed` <- VariantObject$`filtersPassed`
      self$`referenceBases` <- VariantObject$`referenceBases`
      self$`referenceName` <- VariantObject$`referenceName`
      self$`start` <- VariantObject$`start`
      self$`svlen` <- VariantObject$`svlen`
      self$`updated` <- VariantObject$`updated`
      self$`variantDbId` <- VariantObject$`variantDbId`
      self$`variantNames` <- ApiClient$new()$deserializeObj(VariantObject$`variantNames`, "array[character]", loadNamespace("openapi"))
      self$`variantSetDbId` <- ApiClient$new()$deserializeObj(VariantObject$`variantSetDbId`, "array[character]", loadNamespace("openapi"))
      self$`variantType` <- VariantObject$`variantType`
      self
    }
  )
)

